{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1><center><font size=\"6\"> Kaggle: Personalized Medicine - Redefining Cancer Treatment</font></center></h1>\n",
    "\n",
    "<h3><center><font size=\"4\"> https://www.kaggle.com/c/msk-redefining-cancer-treatment </font></center></h2>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table of  Contents\n",
    "1. [Initialization](#1)\n",
    "    1. [Load packages](#11)\n",
    "    1. [Define Metadata](#12)   \n",
    "1. [Load Processed Data and Validate](#2)\n",
    "1. [Data Insights](#3)\n",
    "    1. [Data Structure](#31)\n",
    "    1. [Summary Stats](#32)\n",
    "    1. [Dataset shape](#32)\n",
    "1. [Data Cleansing](#4)\n",
    "    1. [Text Processing](#41)\n",
    "    1. [Merge Two datasets](#42)\n",
    "    1. [Handeling Missing Values](#43)\n",
    "1. [Split Dataset](#5)\n",
    "    1. [Feature - Target Split](#51)\n",
    "    1. [Train - Test - Validation Split](#52)\n",
    "    1. [Distribution of data in Train, Test and Validation set](#53)\n",
    "1. [Feature Engineering](#6)\n",
    "    1. [Handeling Categorical Features](#61)\n",
    "    1. [Evaluating Transformed Features](#62)\n",
    "    1. [Merging Important Features](#63)\n",
    "1. [Modeling Workflow](#7)\n",
    "    1. [Model 1: Compare Different Classifiers and Different Encoding Technique](#71)\n",
    "    1. [Model 2: Hyperparameter Tuning](#72)\n",
    "        1. [Logistic Regression using Hyperparameter Tuning](#721)\n",
    "        1. [Random Forest using Hyperparameter Tuning](#722)\n",
    "    1. [Model 3: Stacking Model](#73)\n",
    "    1. [Model 4: Voting classifier](#74)\n",
    "1. [Conclusion](#8)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='1'>1. Initialization</a>  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id='11'>A. Load Packages</a>  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\behnazfakharfirouzeh\\anaconda3\\lib\\site-packages\\sklearn\\utils\\deprecation.py:143: FutureWarning: The sklearn.metrics.classification module is  deprecated in version 0.22 and will be removed in version 0.24. The corresponding classes / functions should instead be imported from sklearn.metrics. Anything that cannot be imported from sklearn.metrics is now part of the private API.\n",
      "  warnings.warn(message, FutureWarning)\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "# ------------------------------------------------------------------------------------------------------------------------------\n",
    "# pandas is a Python library written for data manipulation and analysis. \n",
    "# In particular, it offers data structures and operations for manipulating numerical tables and time series.\n",
    "import pandas as pd \n",
    "# NumPy is a Python library for adding support for large, multi-dimensional arrays and matrices, \n",
    "# along with a large collection of high-level mathematical functions to operate on these arrays. \n",
    "import numpy as np\n",
    "# SciPy, which stands for Scientific Python, is a scientific computation library that uses NumPy underneath. It provides more \n",
    "# utility functions for optimization, stats and signal processing.\n",
    "# Stack sparse matrices horizontally (column wise)\n",
    "from scipy.sparse import hstack\n",
    "# ------------------------------------------------------------------------------------------------------------------------------\n",
    "# Visualization\n",
    "# Matplotlib is a plotting library for the Python programming language and its numerical mathematics extension NumPy. It  \n",
    "# provides an object-oriented API for embedding plots into applications using general-purpose GUI toolkits\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "# Seaborn is a data visualization library built on top of matplotlib and closely integrated with pandas data structures.\n",
    "import seaborn as sns\n",
    "\n",
    "#  is a tool to visualize high-dimensional data. It converts similarities between data points to joint probabilities and tries \n",
    "# to minimize the Kullback-Leibler divergence between the joint probabilities of the low-dimensional embedding and the high-\n",
    "# dimensional data.\n",
    "from sklearn.manifold import TSNE\n",
    "# ------------------------------------------------------------------------------------------------------------------------------\n",
    "# NLP\n",
    "from nltk.corpus import stopwords\n",
    "# CountVectorizer is used to transform a given text into a vector on the basis of the frequency (count) of each word that occurs\n",
    "# in the entire text.\n",
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "# TfidfVectorizer converts a collection of raw documents to a matrix of TF-IDF features.\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "# ------------------------------------------------------------------------------------------------------------------------------\n",
    "# split into train and test set \n",
    "from sklearn.model_selection import train_test_split\n",
    "# StratifiedKFold, which is a cross-validation object is a variation of KFold that returns stratified folds. The folds are made \n",
    "# by preserving the percentage of samples for each class.\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "\n",
    "# You can fit a model on a training dataset and calibrate this prefit model using a hold out validation dataset. Alternately, \n",
    "# the CalibratedClassifierCV can fit multiple copies of the model using k-fold cross-validation and calibrate the probabilities \n",
    "# predicted by these models using the hold out set.\n",
    "from sklearn.calibration import CalibratedClassifierCV\n",
    "# ------------------------------------------------------------------------------------------------------------------------------\n",
    "# Classifier\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.linear_model import SGDClassifier\n",
    "from sklearn.naive_bayes import MultinomialNB\n",
    "from sklearn.naive_bayes import GaussianNB\n",
    "from sklearn.svm import SVC\n",
    "\n",
    "# Also known as one-vs-all, this strategy consists in fitting one classifier per class. For each classifier, the class is fitted\n",
    "# against all the other classes. In addition to its computational efficiency (only n_classes classifiers are needed), one \n",
    "# advantage of this approach is its interpretability. Since each class is represented by one and one classifier only, it is \n",
    "# possible to gain knowledge about the class by inspecting its corresponding classifier. This is the most commonly used strategy \n",
    "# for multiclass classification and is a fair default choice.\n",
    "from sklearn.multiclass import OneVsRestClassifier\n",
    "\n",
    "# An ensemble-learning meta-classifier for stacking.\n",
    "from mlxtend.classifier import StackingClassifier\n",
    "\n",
    "# Choosing estimators and their parameters\n",
    "from sklearn import model_selection\n",
    "\n",
    "#Refer:http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.VotingClassifier.html\n",
    "from sklearn.ensemble import VotingClassifier\n",
    "# ------------------------------------------------------------------------------------------------------------------------------\n",
    "# Performance evaluation\n",
    "from sklearn.metrics import precision_score, recall_score, f1_score, roc_auc_score\n",
    "from sklearn.metrics.classification import accuracy_score, log_loss\n",
    "from sklearn.metrics import precision_recall_curve, fbeta_score\n",
    "from sklearn.metrics import classification_report\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn.metrics import roc_curve\n",
    "# ------------------------------------------------------------------------------------------------------------------------------\n",
    "# Resampling models\n",
    "from imblearn.over_sampling import SMOTE\n",
    "# ------------------------------------------------------------------------------------------------------------------------------\n",
    "# Dimensionality Reduction\n",
    "from sklearn.decomposition import TruncatedSVD\n",
    "# ------------------------------------------------------------------------------------------------------------------------------\n",
    "# Hyperparameter Tuning\n",
    "from sklearn.model_selection import GridSearchCV\n",
    "# ------------------------------------------------------------------------------------------------------------------------------\n",
    "# Others!\n",
    "\n",
    "# The math module is a standard module in Python, which gives access to the underlying C library functions.\n",
    "import math\n",
    "\n",
    "# Normalize samples individually to unit norm.\n",
    "from sklearn.preprocessing import normalize\n",
    "\n",
    "# The Python module re provides full support for Perl-like regular expressions in Python. The re module raises the exception re. \n",
    "# error if an error occurs while compiling or using a regular expression.\n",
    "import re\n",
    "\n",
    "# The time() function returns the number of seconds passed since epoch.\n",
    "import time\n",
    "\n",
    "# The shuffle() method takes a sequence, like a list, and reorganize the order of the items.\n",
    "from sklearn.utils import shuffle\n",
    "\n",
    "# Python Counter is a container that will hold the count of each of the elements present in the container. The counter is a \n",
    "\n",
    "# sub-class available inside the dictionary class.\n",
    "\n",
    "# Defaultdict is a container like dictionaries present in the module collections. Defaultdict is a sub-class of the dict class \n",
    "#that returns a dictionary-like object. The functionality of both dictionaries and defualtdict are almost same except for the \n",
    "#fact that defualtdict never raises a KeyError. It provides a default value for the key that does not exists.\n",
    "from collections import Counter, defaultdict\n",
    "# ------------------------------------------------------------------------------------------------------------------------------\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "# This is to suppress any deprecation warnings "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This code helps to print dfs side by side\n",
    "# https://jakevdp.github.io/PythonDataScienceHandbook/03.06-concat-and-append.html\n",
    "\n",
    "class display_df_(object):\n",
    "    \"\"\"Display HTML representation of multiple objects\"\"\"\n",
    "    template = \"\"\"<div style=\"float: left; padding: 10px;\">\n",
    "    <p style='font-family:\"Courier New\", Courier, monospace'>{0}</p>{1}\n",
    "    </div>\"\"\"\n",
    "    def __init__(self, *args):\n",
    "        self.args = args\n",
    "        \n",
    "    def _repr_html_(self):\n",
    "        return '\\n'.join(self.template.format(a, eval(a)._repr_html_())\n",
    "                         for a in self.args)\n",
    "    \n",
    "    def __repr__(self):\n",
    "        return '\\n\\n'.join(a + '\\n' + repr(eval(a))\n",
    "                           for a in self.args)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id='12'>B. Define Metadata</a>  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the name of the target class column here instead of manually typing it out everywhere\n",
    "target_class_name = 'Class'\n",
    "\n",
    "input_file_name1 = 'training_variants'\n",
    "input_file_name2 = 'training_text'\n",
    "\n",
    "data_path = 'C:/Behnaz/Udemy/training/'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id='2'>2. Load Processed Data and Validate</a> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_v = pd.read_csv(data_path + input_file_name1)\n",
    "# Loading training_text dataset. This is seperated by ||\n",
    "df_t =pd.read_csv(data_path + input_file_name2,sep=\"\\|\\|\",engine=\"python\",names=[\"ID\",\"TEXT\"],skiprows=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"float: left; padding: 10px;\">\n",
       "    <p style='font-family:\"Courier New\", Courier, monospace'>df_v.head(5)</p><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>FAM58A</td>\n",
       "      <td>Truncating Mutations</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>CBL</td>\n",
       "      <td>W802*</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>CBL</td>\n",
       "      <td>Q249E</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>CBL</td>\n",
       "      <td>N454D</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>CBL</td>\n",
       "      <td>L399V</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "    </div>\n",
       "<div style=\"float: left; padding: 10px;\">\n",
       "    <p style='font-family:\"Courier New\", Courier, monospace'>df_t.head(5)</p><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>TEXT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Cyclin-dependent kinases (CDKs) regulate a var...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>Recent evidence has demonstrated that acquired...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>Oncogenic mutations in the monomeric Casitas B...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "    </div>"
      ],
      "text/plain": [
       "df_v.head(5)\n",
       "   ID    Gene             Variation  Class\n",
       "0   0  FAM58A  Truncating Mutations      1\n",
       "1   1     CBL                 W802*      2\n",
       "2   2     CBL                 Q249E      2\n",
       "3   3     CBL                 N454D      3\n",
       "4   4     CBL                 L399V      4\n",
       "\n",
       "df_t.head(5)\n",
       "   ID                                               TEXT\n",
       "0   0  Cyclin-dependent kinases (CDKs) regulate a var...\n",
       "1   1   Abstract Background  Non-small cell lung canc...\n",
       "2   2   Abstract Background  Non-small cell lung canc...\n",
       "3   3  Recent evidence has demonstrated that acquired...\n",
       "4   4  Oncogenic mutations in the monomeric Casitas B..."
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "display_df_('df_v.head(5)', 'df_t.head(5)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"float: left; padding: 10px;\">\n",
       "    <p style='font-family:\"Courier New\", Courier, monospace'>df_v.tail(5)</p><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3316</th>\n",
       "      <td>3316</td>\n",
       "      <td>RUNX1</td>\n",
       "      <td>D171N</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3317</th>\n",
       "      <td>3317</td>\n",
       "      <td>RUNX1</td>\n",
       "      <td>A122*</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3318</th>\n",
       "      <td>3318</td>\n",
       "      <td>RUNX1</td>\n",
       "      <td>Fusions</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3319</th>\n",
       "      <td>3319</td>\n",
       "      <td>RUNX1</td>\n",
       "      <td>R80C</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3320</th>\n",
       "      <td>3320</td>\n",
       "      <td>RUNX1</td>\n",
       "      <td>K83E</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "    </div>\n",
       "<div style=\"float: left; padding: 10px;\">\n",
       "    <p style='font-family:\"Courier New\", Courier, monospace'>df_t.tail(5)</p><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>TEXT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3316</th>\n",
       "      <td>3316</td>\n",
       "      <td>Introduction  Myelodysplastic syndromes (MDS) ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3317</th>\n",
       "      <td>3317</td>\n",
       "      <td>Introduction  Myelodysplastic syndromes (MDS) ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3318</th>\n",
       "      <td>3318</td>\n",
       "      <td>The Runt-related transcription factor 1 gene (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3319</th>\n",
       "      <td>3319</td>\n",
       "      <td>The RUNX1/AML1 gene is the most frequent targe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3320</th>\n",
       "      <td>3320</td>\n",
       "      <td>The most frequent mutations associated with le...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "    </div>"
      ],
      "text/plain": [
       "df_v.tail(5)\n",
       "        ID   Gene Variation  Class\n",
       "3316  3316  RUNX1     D171N      4\n",
       "3317  3317  RUNX1     A122*      1\n",
       "3318  3318  RUNX1   Fusions      1\n",
       "3319  3319  RUNX1      R80C      4\n",
       "3320  3320  RUNX1      K83E      4\n",
       "\n",
       "df_t.tail(5)\n",
       "        ID                                               TEXT\n",
       "3316  3316  Introduction  Myelodysplastic syndromes (MDS) ...\n",
       "3317  3317  Introduction  Myelodysplastic syndromes (MDS) ...\n",
       "3318  3318  The Runt-related transcription factor 1 gene (...\n",
       "3319  3319  The RUNX1/AML1 gene is the most frequent targe...\n",
       "3320  3320  The most frequent mutations associated with le..."
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "display_df_('df_v.tail(5)', 'df_t.tail(5)')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### $\\rightarrow$ No trouble with loading the data, the head and tail."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id=3 > 3. Data Insights"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id='31'>A. Data Structure</a> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 3321 entries, 0 to 3320\n",
      "Data columns (total 4 columns):\n",
      " #   Column     Non-Null Count  Dtype \n",
      "---  ------     --------------  ----- \n",
      " 0   ID         3321 non-null   int64 \n",
      " 1   Gene       3321 non-null   object\n",
      " 2   Variation  3321 non-null   object\n",
      " 3   Class      3321 non-null   int64 \n",
      "dtypes: int64(2), object(2)\n",
      "memory usage: 103.9+ KB\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "None"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Lets see the data structure\n",
    "display(df_v.info())\n",
    "# This will give you an indication of how many rows and columns are present in the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 3321 entries, 0 to 3320\n",
      "Data columns (total 2 columns):\n",
      " #   Column  Non-Null Count  Dtype \n",
      "---  ------  --------------  ----- \n",
      " 0   ID      3321 non-null   int64 \n",
      " 1   TEXT    3316 non-null   object\n",
      "dtypes: int64(1), object(1)\n",
      "memory usage: 52.0+ KB\n"
     ]
    }
   ],
   "source": [
    "# Lets see the data structure\n",
    "df_t.info()\n",
    "# This will give you an indication of how many rows and columns are present in the dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id='32'>B. Summary Stats</a> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"float: left; padding: 10px;\">\n",
       "    <p style='font-family:\"Courier New\", Courier, monospace'>df_v.describe()</p><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>3321.000000</td>\n",
       "      <td>3321.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>1660.000000</td>\n",
       "      <td>4.365854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>958.834449</td>\n",
       "      <td>2.309781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>830.000000</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>1660.000000</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>2490.000000</td>\n",
       "      <td>7.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>3320.000000</td>\n",
       "      <td>9.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "    </div>\n",
       "<div style=\"float: left; padding: 10px;\">\n",
       "    <p style='font-family:\"Courier New\", Courier, monospace'>df_t.describe()</p><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>3321.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>1660.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>958.834449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>830.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>1660.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>2490.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>3320.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "    </div>"
      ],
      "text/plain": [
       "df_v.describe()\n",
       "                ID        Class\n",
       "count  3321.000000  3321.000000\n",
       "mean   1660.000000     4.365854\n",
       "std     958.834449     2.309781\n",
       "min       0.000000     1.000000\n",
       "25%     830.000000     2.000000\n",
       "50%    1660.000000     4.000000\n",
       "75%    2490.000000     7.000000\n",
       "max    3320.000000     9.000000\n",
       "\n",
       "df_t.describe()\n",
       "                ID\n",
       "count  3321.000000\n",
       "mean   1660.000000\n",
       "std     958.834449\n",
       "min       0.000000\n",
       "25%     830.000000\n",
       "50%    1660.000000\n",
       "75%    2490.000000\n",
       "max    3320.000000"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "display_df_('df_v.describe()','df_t.describe()')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id='34'>C. Dataset shape</a> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape of variants dataset: (3321, 4)\n",
      "Shape of text dataset: (3321, 2)\n"
     ]
    }
   ],
   "source": [
    "print('Shape of variants dataset:', df_v.shape)\n",
    "print('Shape of text dataset:', df_t.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id=4 > 4. Data Cleansing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Convert the data into a format that our ML method can understand it better. Such as\n",
    " - Unnessaury numbers\n",
    " - Stop words (e.g; “the”, “is”, “in”, “for”, “where”, “when”, “to”, “at”, etc.) \n",
    " - Extra spaces."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id='41'>A.Text Processing</a> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To remove all English stop words like a, is, an, the, etc, we can use the following nltk library.\n",
    "stop_words = set(stopwords.words('english'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def data_text_preprocess(total_text, ind, col):\n",
    "    # Removing insignificant int values from text data \n",
    "    if type(total_text) is not int:\n",
    "        string = \"\"\n",
    "        # Replacing all special char with space\n",
    "        total_text = re.sub('[^a-zA-Z0-9\\n]', ' ', str(total_text))\n",
    "        # Replacing multiple spaces with single space\n",
    "        total_text = re.sub('\\s+',' ', str(total_text))\n",
    "        # Converting all uppercase characters to lowercase. \n",
    "        total_text = total_text.lower()\n",
    "        \n",
    "        for word in total_text.split():\n",
    "        # if the word is a not a stop word then retain that word from text.\n",
    "            if not word in stop_words:\n",
    "                string += word + \" \"\n",
    "        \n",
    "        df_t[col][ind] = string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Below code will take some time because its huge text (took 4 minute on my 16 GB RAM system), so run it and have a cup of coffee :)\n",
    "for index, row in df_t.iterrows():\n",
    "    if type(row['TEXT']) is str:\n",
    "        data_text_preprocess(row['TEXT'], index, 'TEXT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ensure that all spaces in Gene and Variation column to be replaced by _.\n",
    "df_v.Gene      = df_v.Gene.str.replace('\\s+', '_')\n",
    "df_v.Variation = df_v.Variation.str.replace('\\s+', '_')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id='42'>B. Merge Two datasets</a> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Class</th>\n",
       "      <th>TEXT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>FAM58A</td>\n",
       "      <td>Truncating_Mutations</td>\n",
       "      <td>1</td>\n",
       "      <td>cyclin dependent kinases cdks regulate variety...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>CBL</td>\n",
       "      <td>W802*</td>\n",
       "      <td>2</td>\n",
       "      <td>abstract background non small cell lung cancer...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>CBL</td>\n",
       "      <td>Q249E</td>\n",
       "      <td>2</td>\n",
       "      <td>abstract background non small cell lung cancer...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>CBL</td>\n",
       "      <td>N454D</td>\n",
       "      <td>3</td>\n",
       "      <td>recent evidence demonstrated acquired uniparen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>CBL</td>\n",
       "      <td>L399V</td>\n",
       "      <td>4</td>\n",
       "      <td>oncogenic mutations monomeric casitas b lineag...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID    Gene             Variation  Class  \\\n",
       "0   0  FAM58A  Truncating_Mutations      1   \n",
       "1   1     CBL                 W802*      2   \n",
       "2   2     CBL                 Q249E      2   \n",
       "3   3     CBL                 N454D      3   \n",
       "4   4     CBL                 L399V      4   \n",
       "\n",
       "                                                TEXT  \n",
       "0  cyclin dependent kinases cdks regulate variety...  \n",
       "1  abstract background non small cell lung cancer...  \n",
       "2  abstract background non small cell lung cancer...  \n",
       "3  recent evidence demonstrated acquired uniparen...  \n",
       "4  oncogenic mutations monomeric casitas b lineag...  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#merging both gene_variations and text data based on ID\n",
    "df = pd.merge(df_v, df_t,on='ID', how='left')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape of the dataset after concatination: (3321, 5)\n"
     ]
    }
   ],
   "source": [
    "print('Shape of the dataset after concatination:', df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id='43'>C. Handeling Missing Values</a> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Class</th>\n",
       "      <th>TEXT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1109</th>\n",
       "      <td>1109</td>\n",
       "      <td>FANCA</td>\n",
       "      <td>S1088F</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1277</th>\n",
       "      <td>1277</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>Truncating_Mutations</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1407</th>\n",
       "      <td>1407</td>\n",
       "      <td>FGFR3</td>\n",
       "      <td>K508M</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1639</th>\n",
       "      <td>1639</td>\n",
       "      <td>FLT1</td>\n",
       "      <td>Amplification</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2755</th>\n",
       "      <td>2755</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>G596C</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ID    Gene             Variation  Class TEXT\n",
       "1109  1109   FANCA                S1088F      1  NaN\n",
       "1277  1277  ARID5B  Truncating_Mutations      1  NaN\n",
       "1407  1407   FGFR3                 K508M      6  NaN\n",
       "1639  1639    FLT1         Amplification      6  NaN\n",
       "2755  2755    BRAF                 G596C      7  NaN"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.isnull().any(axis=1)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Based on the above table, there are many rows with missing values. To handel missing value we can either drop rows having missing values or do some imputation in it. Here, we consider imputation since we have a small dataset. \n",
    "\n",
    "For imputation, we replace NaN with the concatination value of Gene and Variation column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.loc[df['TEXT'].isnull(),'TEXT'] = df['Gene'] +' '+ df['Variation']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Class</th>\n",
       "      <th>TEXT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [ID, Gene, Variation, Class, TEXT]\n",
       "Index: []"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check for mising value after imputation\n",
    "df[df.isnull().any(axis=1)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There is no missing value in the dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id=5 > 5. Split Dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id='51'>A. Feature - Target Split </a> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "X = df.drop(target_class_name, axis=1)\n",
    "y = df[target_class_name].values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id='52'>B. Train - Test - Validation Split</a> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of data points in train data: 2124\n",
      "Number of data points in test data: 665\n",
      "Number of data points in validation data: 532\n"
     ]
    }
   ],
   "source": [
    "# Split the data into 80% train and 20% test set. \n",
    "X1_train, X_test, y1_train, y_test = train_test_split(df, y, stratify=y, test_size=0.2,random_state=42)\n",
    "# Split the train data now into 80% train and 20% validation set.\n",
    "X_train, X_cv, y_train, y_cv = train_test_split(X1_train, y1_train, stratify=y1_train, test_size=0.2,random_state=42)\n",
    "\n",
    "print('Number of data points in train data:', X_train.shape[0])\n",
    "print('Number of data points in test data:', X_test.shape[0])\n",
    "print('Number of data points in validation data:', X_cv.shape[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id='53'>C. Distribution of data in Train, Test and Validation set.</a> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>train data</th>\n",
       "      <th>test data</th>\n",
       "      <th>validation data</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>363</td>\n",
       "      <td>114</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>289</td>\n",
       "      <td>91</td>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>57</td>\n",
       "      <td>18</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>439</td>\n",
       "      <td>137</td>\n",
       "      <td>110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>155</td>\n",
       "      <td>48</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>176</td>\n",
       "      <td>55</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>609</td>\n",
       "      <td>191</td>\n",
       "      <td>153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>24</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   train data  test data  validation data\n",
       "1         363        114               91\n",
       "2         289         91               72\n",
       "3          57         18               14\n",
       "4         439        137              110\n",
       "5         155         48               39\n",
       "6         176         55               44\n",
       "7         609        191              153\n",
       "8          12          4                3\n",
       "9          24          7                6"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# To make sure that we have proper class distribution in each set.\n",
    "Distribution_df = pd.DataFrame(columns = ['train data','test data','validation data'],index = df['Class'].unique())\n",
    "\n",
    "# X_train['Class'].value_counts().sort_index()\n",
    "Distribution_df['train data'] = X_train['Class'].value_counts().sort_index()\n",
    "Distribution_df['test data'] = X_test['Class'].value_counts().sort_index()\n",
    "Distribution_df['validation data'] = X_cv['Class'].value_counts().sort_index()\n",
    "Distribution_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To better observe the distribution of classes in different sets, we can show the class distribution in %."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>train data %</th>\n",
       "      <th>test data %</th>\n",
       "      <th>validation data %</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>17.090</td>\n",
       "      <td>17.143</td>\n",
       "      <td>17.105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13.606</td>\n",
       "      <td>13.684</td>\n",
       "      <td>13.534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.684</td>\n",
       "      <td>2.707</td>\n",
       "      <td>2.632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>20.669</td>\n",
       "      <td>20.602</td>\n",
       "      <td>20.677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>7.298</td>\n",
       "      <td>7.218</td>\n",
       "      <td>7.331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>8.286</td>\n",
       "      <td>8.271</td>\n",
       "      <td>8.271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>28.672</td>\n",
       "      <td>28.722</td>\n",
       "      <td>28.759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.565</td>\n",
       "      <td>0.602</td>\n",
       "      <td>0.564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1.130</td>\n",
       "      <td>1.053</td>\n",
       "      <td>1.128</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   train data %  test data %  validation data %\n",
       "1        17.090       17.143             17.105\n",
       "2        13.606       13.684             13.534\n",
       "3         2.684        2.707              2.632\n",
       "4        20.669       20.602             20.677\n",
       "5         7.298        7.218              7.331\n",
       "6         8.286        8.271              8.271\n",
       "7        28.672       28.722             28.759\n",
       "8         0.565        0.602              0.564\n",
       "9         1.130        1.053              1.128"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# To make sure that we have proper class distribution in each set.\n",
    "Distribution_per_df = pd.DataFrame(columns = ['train data %','test data %','validation data %'],index = df['Class'].unique())\n",
    "\n",
    "# X_train['Class'].value_counts().sort_index()\n",
    "Distribution_per_df['train data %'] = np.round((Distribution_df['train data'].values[:]/X_train.shape[0]*100), 3)\n",
    "Distribution_per_df['test data %'] = np.round((Distribution_df['test data'].values[:]/X_test.shape[0]*100), 3)\n",
    "Distribution_per_df['validation data %'] = np.round((Distribution_df['validation data'].values[:]/X_cv.shape[0]*100), 3)\n",
    "Distribution_per_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Text(0, 0.5, 'Data points (%)'), Text(0.5, 0, 'Class')]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA3UAAAFhCAYAAADX1oOEAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAgAElEQVR4nO3debgkdXn3//eHAQQRBGFEdqIhRCQRERHFKC5EJS5oghEV96BXJGpCNGiiQZ9oiFEToyb+UFFEwQ0XFH5RVBR3BYICgg+CbDLCsDkDKsJwP39UHdMcztJz5lR315n367r6mu7qOt/7293nfKbvqurqVBWSJEmSpH7aYNwTkCRJkiQtnE2dJEmSJPWYTZ0kSZIk9ZhNnSRJkiT1mE2dJEmSJPWYTZ0kSZIk9ZhNnSQNSHJ0kg93NPbzk3xj4PbNSe67SGO/Nsn72uu7JqkkGy7S2Du3c122GOOtRd1tk5yZZHWSt63Fzy3q4+9KkguSHLBIY301yYvb689O8sWB+/ZPcnH7Gh680OdVkjS5Jvo/PElabEluHrh5d+BWYE17+yWjnEtV3WO+ddo3/R+uqh3nGevNizWvJJcBL66qL7VjXwHMO9cOHA5cB2xRS/BLVavqAR2N+xHgIwOL3gi8q6reAZDkdYzpeU3yfJrfrUd0MHYBu1XVTxZ7bEmadO6pk7Reqap7TF2AK4AnDyz7yHw/P4kmfY/UOtgF+NFSbOhGbBfggmm3F/S8LuHfNUnqNZs6SbqrjZN8qD087YIk+0zdkWT7JCcnWZnkp0lePtsgSbZOckqSVUm+B9xv2v2V5Hfb6wcl+VFb82dJ/jbJZsD/D2zfHjp3c1v/6CSfTPLhJKuA589y2OgLk1ydZEWSIwfqfjDJPw3cPiDJVe31E4Cdgc+19V49/XDGdg6nJLkhyU+S/MXAWEcn+fhsz98Mz9HDk3w/yS/afx8+NUfgecCr23k8boaf3TTJ25Jc3v78N5JsOsN6L0hyYTufS5O8ZOC+bZJ8PslN7eP5epIN2vv+rn0tVif5cZLHtss3SHJUkkuSXN8+3nu1923Svi7Xt2N+P8m2szz2y6Ye1wKetwOTXNQ+7ncBGbjvt4f5JrkEuO/A63nS9Od1nscz9dq/KMkVwFfa5S9sn9Mbk3whyS4D9SvJS9Mc8nljknencX/gPcDD2to3zfLYnt++TqvT/I09e+C+GesmObNd5Qft2H8+12srSUuN4SZJd/UU4KPAlsApwLugeTMPfA74AbAD8FjglUkeP8s47wZ+DWwHvLC9zOb9wEuqanNgT+ArVXUL8ETg6oG9iVe36z8V+GQ7x9n2MD4a2A34Y+ComRqj6arqMO68B/MtM6x2EnAVsD3wZ8Cbpxqe1ozP33Rt43Aq8B/A1sDbgVOTbF1Vz28f11vaeXxphiHeCjwYeDhwL+DVwB0zrHct8CRgC+AFwL8l2bu978j2sSwHtgVeC1SS3YEjgIe0r8njgcvan3k5cDDwqPY5uJHmtYamYbonsFP7mF4K/Gqmxz+DYZ+3bYCTgX8AtgEuAfafad2quh93fj0P5a7P61yPZ8qjgPsDj09yMM3z9HSa5+3rNL8Tg54EPAR4IPAM4PFVdSHN8/HttvaWMzy2zWh+H57YPu8PB85t75u1blU9sh3ige3YH2OW13am50mS+s6mTpLu6htVdVpVrQFOoHljCs2b1OVV9caq+k1VXQq8F3jm9AHSnFTkT4HXV9UtVXU+cPwcNW8D9kiyRVXdWFXnzDPHb1fVZ6rqjqqarWl4Q1v7POADwKHzjDmvJDsBjwD+rqp+XVXnAu8DDhtYbbbnb7o/AS6uqhOq6vaqOgm4CHjyEPPYgKZJfkVV/ayq1lTVt6rq1unrVtWpVXVJNb4GfBH4o/bu22ia7l2q6raq+np7WOIa4G40r8lGVXVZVV3S/sxLgL+vqqvaekcDf5ZmT+ZtNM3c77ZzOruqVs33eFrDPm8H0Rw++cmqug34d+DnQ9aYyVyPZ8rR7e/Sr9r1/7mqLqyq24E3A3sN7q0Djqmqm9rPY54B7LUW87kD2DPJplW1oqqmDh0dpu6g2V5bSVpybOok6a4G3yD/EtikfYO7C82hkDdNXWi2/s90eN1ympNRXTmw7PI5av4pzZv1y5N8LcnD5pnjlfPcP32dy2n2wqyr7YEbqmr1tLF3GLg92/M301jTn5PpY81mG2ATmr1Uc0ryxCTfaQ/Bu4nmed6mvftfgZ8AX2wP+TsKoD3ZxitpGpxrk3w0ydTztwvw6YHfgQtpmsBtaZqxLwAfTXPo61uSbDTE44G1e95++9q2jcowvw+zmevxTLly2vrvGFj/BprDP+f6HRjqRDvt3uk/p9mjtyLJqUl+fy3qDprxtZWkpcimTpKGdyXw06racuCyeVUdNMO6K4HbaQ7Dm7LzbANX1fer6qnAvYHPAB+fumu2HxlivtNrTx26eQvNmT+n3Gctxr4auFeSzaeN/bMh5jPTWNP3sgw71nU0h7beb66VktyN5lDFtwLbtof8nUb7GbSqWl1VR1bVfWn2EP7N1KGkVXVie5bGXWiek39ph72S5vDAwd+DTdo9hrdV1Ruqag+aQwefBDx3iMezNlYw8NomCXd+rdfWrI9nYJ2atv5Lpq2/aVV9a4ha8/7eVtUXqupAmr1sF9HsDV/runO9tpK01NjUSdLwvgesSnMCjU2TLEuyZ5KHTF+xPYTuU8DRSe6eZA+az1vdRZKN03y32D3bw+lW8b9fs3ANsHWSey5gvq9raz+A5rNkH2uXnwsclOReSe5Ds0dq0DU0J9e4i6q6EvgW8M9pTgryh8CLmP1zfXM5Dfi9JM9KsmGSPwf2AD4/3w9W1R3AccDb05y4ZVmSh7VN3KCNaQ6jXAncnuSJNJ8xBCDJk5L8btsYTT3va5LsnuQx7Xi/pvlc3NRr8h7gTQMn6Vie5Knt9Ucn+YP28NtVNIcATv3cYjkVeECSp7d78l7OXRvztTHr45lj/de0v1ckuWeSQ4asdQ2wY5KNZ7ozzXfoPaX9bN2twM3c+Xmfq+6dfm9ne22HnKck9YpNnSQNqW3Unkzz+aCf0uwteh/NiTFmcgTNYWc/Bz5I87m22RwGXJbmbJYvBZ7T1ryI5mQQl7aHna3NIZRfozn87MvAW6tq6gupT6A52ctlNJ8v+9i0n/tn4B/aen87w7iHArvS7Gn7NPCPVXX6WswLgKq6nmZP1pHA9TQnOnlSVV035BB/C5wHfJ/mULx/Ydr/a+1hoi+n2fN5I/AsmpOQTNkN+BJN8/Bt4D+r6qs0jeAxNK/xz2n2oL62/Zl3tGN8Mclq4DvAQ9v77kNzAptVNIcxfg1Y1C+zb5+fQ9r5Xd8+hm+uw5BzPZ6Z6n+a5rn+aPv7ej7NCX2G8RWar1f4eZKZXucNaH4frqZ5TR8F/OWQdY8Gjm9/b5/B7K+tJC058TPDkiRJktRf7qmTJEmSpB6zqZMkSZKkHrOpkyRJkqQes6mTJEmSpB6zqZMkSZKkHrOpkyRJkqQes6mTJEmSpB6zqZMkSZKkHrOpkyRJkqQes6lTp5K8J8nrFmmso5N8eDHGkqR1ZSZJmpLkgCRXDdy+IMkBw6y7gFqL9t5qnjrrNE+Nlk2dZpXksiSPW5cxquqlVfV/FmtOw0rywST/NOq6kkZrMXKqHef5Sb6xGHOaZXwzSVqPVNUDquqr6zrOTNk0rvdWc+k6QzU/mzotWJINxz0HSZIkaX1nU6cZJTkB2Bn4XJKbk7w6ya5JKsmLklwBfKVd9xNJfp7kF0nOTPKAgXF+u3V6ajd+kiOTXJtkRZIXzDGH30nytSSrk5wObDPt/hnrJjkceDbw6nbun2uXH5Xkkna8HyV52uI+a5JGaaacapfvl+RbSW5K8oPBQ6DarcmXtjnw0yTPTnJ/4D3Aw9pxbpqlnpkkLWHt3+Qnpy17R5L/aK+/IMmF7d/spUleMsdYvz2KIMmm7fuhG5P8CHjIDHXvkgWzZdP0Pf9J/iLJT5LckOSUJNsP3FdJXprk4rb+u5Nkljkv9jz/JMn/JFmV5MokR8/5AmjdVJUXLzNegMuAxw3c3hUo4EPAZsCm7fIXApsDdwP+HTh34Gc+CPxTe/0A4HbgjcBGwEHAL4GtZqn/beDt7biPBFYDHx64f6i6A8sOAban2Zjx58AtwHbjfp69ePGy8MsMObUDcH2bLxsAB7a3l7e5tQrYvV13O+AB7fXnA9+Yp5aZ5MXLEr4Au7TvS7Zoby8DVgD7tbf/BLgfEOBR7bp7t/cdAFw1MNZvswk4Bvg6cC9gJ+D8aevOmgUzZdO091aPAa4D9m6z553AmQPrFvB5YEuajWArgSfM8vgXe54HAH/Qrv+HwDXAweN+nZfqxT11Woijq+qWqvoVQFUdV1Wrq+pW4GjggUnuOcvP3ga8sapuq6rTgJuB3aevlGRnmi1Er6uqW6vqTOBzg+usZV2q6hNVdXVV3VFVHwMuBvZdy8cuabI9Bzitqk5r/9ZPB86iafIA7gD2TLJpVa2oqguGGdRMkpa+qrocOAc4uF30GOCXVfWd9v5Tq+qSanwN+CLwR0MM/QzgTVV1Q1VdCfzHtLrrkgXPBo6rqnPa7HkNzR6zXQfWOaaqbqqqK4AzgL1GMc+q+mpVndeu/0PgJJpmWB2wqdNCXDl1JcmyJMe0u+NX0WyZgmmHJQ24vqpuH7j9S+AeM6y3PXBjVd0ysOzydahLkucmObc9JOsmYM+51pfUS7sAh0z9nbd/64+g2Zp8C83W5ZcCK5KcmuT3hxzXTJLWDycCh7bXn9XeBiDJE5N8pz3M8SaajUXD/M1uz8B7Jwayox13XbJg+8HxqupmmqMTdhhY5+cD12d737Xo80zy0CRnJFmZ5Bc02WvGdcSmTnOpIZY/C3gq8DjgnjSHaEJzaMK6WAFslWSzgWU7r0XdO809yS7Ae4EjgK2rakuawwrWdZ6Sxmt6Tl0JnFBVWw5cNquqYwCq6gtVdSDNoZcX0eTCTONMZyZJ64dPAAck2RF4Gm1Tl+RuwMnAW4Ft27/Z0xjub3YFzeGMU36bHUNkwXzZdDXNxqyp8TYDtgZ+NsS8up7nicApwE5VdU+az92ZcR2xqdNcrgHuO886mwO30mwVujvw5sUo3B4CcRbwhiQbJ3kE8OS1qDt97pvRBM5KaD7sTLOFSVK/Tf9b/zDw5CSPb/eebZLmJE07Jtk2yVPaNz230hz+vWZgnB2TbDxTETNJWj9U1Urgq8AHgJ9W1YXtXRvTfGZtJXB7kicCfzzksB8HXpNkq7ZZ/KuB++bLgjmziaZxekGSvdrG883Ad6vqsiHn1uU8NwduqKpfJ9mXZuOXOmJTp7n8M/AP7W72v51lnQ/R7J7/GfAj4DuLWP9ZwEOBG4B/bGsNW/f9wB7t3D9TVT8C3kZzooNraD64+81FnKuk8bhTTrWfA3kq8FqaNx9XAq+i+f9uA+BImi3bN9B8tuMv23G+AlwA/DzJdbPUMpOk9cOJNHvdf3voZVWtBl5O0/jcSJMHpww53hto8uGnNJ/DO2Fg3PmyYM5sqqovA6+j2Yu4guZELs8ccl5dz/MvgTcmWQ28nua5U0dSNd9eXUmSJEnSpHJPnSRJkiT1mE2dJEmSJPVYZ01d++H07yX5QZILkryhXX6vJKe332x/epKtupqDJEmSJC11Xe6puxV4TFU9kOZLDp+QZD/gKODLVbUb8OX2tiSNhBucJE0q80nSQnXW1FXj5vbmRu2laM5Kdny7/Hjg4K7mIEkzcIOTpEllPklakA27HDzJMuBs4HeBd1fVd5NsW1UrAKpqRZJ7zzfONttsU7vuumuXU5U0YmefffZ1VbV81HWrOeXvbBucDmiXH0/zPUV/N9dYZpO0NJlPkibRXNnUaVNXVWuAvZJsCXw6ydBfrJrkcOBwgJ133pmzzjqro1lKGockl4+x9oI3OJlN0tLX13watOuuu5pP0hIzVzaN5OyXVXUTzValJwDXJNmundh2wLWz/MyxVbVPVe2zfPnIN5ZJWsKqak1V7QXsCOy7NhuczCZJXVqXfEpyeJKzkpy1cuXK7iYpaeJ0efbL5e0eOpJsCjwOuAg4BXheu9rzgM92NQdJmstCNjhJ0ii4QVzS2uhyT912wBlJfgh8Hzi9qj4PHAMcmORi4MD2tiSNhBucJE0q80nSQnX2mbqq+iHwoBmWXw88tqu6kjSP7YDj28+tbAB8vKo+n+TbwMeTvAi4AjhknJOUtF4ynyQtSKcnSpGkSeMGJ0mTynyStFAjOVGKJEmSJKkbNnWSJEmS1GM2dZIkSZLUYzZ1kiRJktRjNnWSJEmS1GM2dZIkSZLUY36lgbQOLnzTVzoZ9/5//5hOxpW0fjCbJE0q86kb7qmTJEmSpB6zqZMkSZKkHrOpkyRJkqQes6mTJEmSpB6zqZMkSZKkHrOpkyRJkqQes6mTJEmSpB6zqZMkSZKkHrOpkyRJkqQes6mTJEmSpB6zqZMkSZKkHrOpkyRJkqQes6mTJEmSpB6zqZMkSZKkHrOpkyRJkqQes6mTJEmSpB6zqZMkSZKkHrOpkyRJkqQes6mTJEmSpB6zqZMkSZKkHrOpkyRJkqQes6mTJEmSpB6zqZMkSZKkHrOpkyRJkqQes6mTJEmSpB6zqZMkSZKkHrOpkyRJkqQes6mTJEmSpB7rrKlLslOSM5JcmOSCJK9olx+d5GdJzm0vB3U1B0mSJEla6rrcU3c7cGRV3R/YD3hZkj3a+/6tqvZqL6d1OAdJuhM3OEmaVOaTpIXasKuBq2oFsKK9vjrJhcAOXdWTpCFNbXA6J8nmwNlJTm/v+7eqeusY5yZp/WY+SVqQkXymLsmuwIOA77aLjkjywyTHJdlqFHOQJGg2OFXVOe311YAbnCRNBPNJ0kJ13tQluQdwMvDKqloF/BdwP2Avmj15b5vl5w5PclaSs1auXNn1NCWthxaywclskjQKbhCXtDY6beqSbETT0H2kqj4FUFXXVNWaqroDeC+w70w/W1XHVtU+VbXP8uXLu5ympPXQQjc4mU2SuuYGcUlrq8uzXwZ4P3BhVb19YPl2A6s9DTi/qzlI0kzWZYOTJHXJDeKSFqKzE6UA+wOHAeclObdd9lrg0CR7AQVcBrykwzlI0p3MtcGpPcETuMFJ0hiYT5IWqsuzX34DyAx3+RUGksbJDU6SJpX5JGlButxTJ0kTxw1OkiaV+SRpoUbylQaSJEmSpG7Y1EmSJElSj9nUSZIkSVKP2dRJkiRJUo/Z1EmSJElSj9nUSZIkSVKP2dRJkiRJUo/Z1EmSJElSj9nUSZIkSVKP2dRJkiRJUo/Z1EmSJElSj9nUSZIkSVKP2dRJkiRJUo/Z1EmSJElSj9nUSZIkSVKP2dRJkiRJUo/Z1EmSJElSj9nUSZIkSVKP2dRJkiRJUo/Z1EmSJElSj9nUSZIkSVKP2dRJkiRJUo/Z1EmSJElSj9nUSZIkSVKP2dRJkiRJUo/Z1EmSJElSj9nUSZIkSVKP2dRJkiRJUo/Z1EmSJElSj9nUSZIkSVKP2dRJkiRJUo/Z1EmSJElSj9nUSZIkSVKPbTjuCWg8vvbIR3Uy7qPO/Fon40qSJEmamU2dJGliuMFJ0qTqIp/MJi2Wzg6/TLJTkjOSXJjkgiSvaJffK8npSS5u/92qqzlIkiRJ0lLX5WfqbgeOrKr7A/sBL0uyB3AU8OWq2g34cntbkkbCDU6SJpX5JGmhOmvqqmpFVZ3TXl8NXAjsADwVOL5d7Xjg4K7mIEkzcIOTpEllPklakJGc/TLJrsCDgO8C21bVCmgaP+Deo5iDJIEbnCRNLvNJ0kJ13tQluQdwMvDKqlq1Fj93eJKzkpy1cuXK7iYoab3lBidJk8p8krQ2Om3qkmxE09B9pKo+1S6+Jsl27f3bAdfO9LNVdWxV7VNV+yxfvrzLaUpaD7nBSdKkMp8kra0uz34Z4P3AhVX19oG7TgGe115/HvDZruYgSTNxg5OkSWU+SVqILvfU7Q8cBjwmybnt5SDgGODAJBcDB7a3JWkk3OAkaVKZT5IWqrMvH6+qbwCZ5e7HdlVXkuYxtcHpvCTntsteS7OB6eNJXgRcARwypvlJWn+ZT5IWpLOmTpImkRucJE0q80nSQo3kKw0kSZIkSd0YuqlLslmSZV1ORpIWwnySNInMJkmjMmtTl2SDJM9KcmqSa4GLgBVJLkjyr0l2G900Jel/mU+SJpHZJGlc5tpTdwZwP+A1wH2qaqequjfwR8B3gGOSPGcEc5Sk6cwnSZPIbJI0FnOdKOVxVXXb9IVVdQPN96ec3H6XiiSNmvkkaRKZTZLGYtambnooJdkEeA6wKXBiVV0/U3BJUtfMJ0mTyGySNC5rc/bLdwDLgF8Dn+lmOpK0IOaTpElkNkkaiblOlHJikvsNLLoX8BHgJGCrricmSbMxnyRNIrNJ0rjM9Zm6fwD+KcnVwP8B3gqcAmwCHN391CRpVuaTpElkNkkai7k+U3cp8KwkjwA+BpwKHFhVa0Y1OUmaifkkaRKZTZLGZa7DL7dK8jJgD+AZwC+ALyR50qgmJ0kzMZ8kTSKzSdK4zHWilM8At9IcMnBCVX0IeDLw4CSnjGJykjQL80nSJDKbJI3FXJ+p2xo4keY0vM8FqKpfAW9Ist0I5iZJszGfJE0is0nSWMzV1P0jcDqwBjhq8I6qWtHlpCRpHuaTpElkNkkai7lOlHIycPII5yJJQzGfJE0is0nSuMx1opRjk+w5y32bJXlhkmd3NzVJmpn5JGkSmU2SxmWuwy//E3h9kj8AzgdW0nzwdzdgC+A4mi/UlKRRM58kTSKzSdJYzHX45bnAM5LcA9gH2A74FXBhVf14RPOTpLswnyRNIrNJ0rjMtacOgKq6Gfhq91ORpLVjPkmaRGaTpFGb63vqJEmSJEkTbt49dZKk+T34VR/qZNyz//W5nYwrSZKWjrVq6pJsANyjqlZ1NJ95+cZJ0kwmIZ8kabpJyKYu3jv5vkmaLPMefpnkxCRbJNkM+BHw4ySv6n5qkjQ380nSJDKbJI3aMJ+p26PdunQwcBqwM3BYp7OSpOGYT5ImkdkkaaSGaeo2SrIRTTB9tqpu63hOkjQs80nSJDKbJI3UME3d/wdcBmwGnJlkF+AXXU5KkoZkPkmaRGaTpJEapqn7XFXtUFUHVVUBVwAv7HhekjQM80nSJDKbJI3UME3dyYM32nD6aDfTkaS1Yj5JmkRmk6SRmvUrDZL8PvAA4J5Jnj5w1xbAJl1PTJJmYz5JmkRmk6Rxmet76nYHngRsCTx5YPlq4C+6nJQkzcN8kjSJzCZJYzFrU1dVnwU+m+RhVfXtEc5JkuZkPkmaRGaTpHGZa0/dlJ8keS2w6+D6VeUHfiWNm/kkaRKZTZJGapim7rPA14EvAWu6nY4krRXzSdIkMpskjdQwTd3dq+rvOp+JJK0980nSJDKbJI3UMF9p8PkkB3U+E0lae+aTpElkNkkaqWGaulfQhNOvkqxKsjrJqq4nJklDMJ8kTSKzSdJIzdvUVdXmVbVBVW1aVVu0t7eY7+eSHJfk2iTnDyw7OsnPkpzbXtyKJWnBzCdJk8hskjRqc375eFVdlGTvme6vqnPmGfuDwLuAD01b/m9V9da1mqUkDTCfJE0is0nSuMx1opS/AQ4H3jbDfQU8Zq6Bq+rMJLsueGaSNDvzSdIkMpskjcVcXz5+ePvvoxe55hFJngucBRxZVTfOtFKSw2mCkZ133nmRpyCpz8adT5I0E7NJ0rjM+5m6JBsleXmST7aXI5JstMB6/wXcD9gLWMHMW7IAqKpjq2qfqtpn+fLlCywnaSkbRz4lOTzJWUnOWrly5QJLSVrKxvXeyXyS1l/DnP3yv4AHA//ZXh7cLltrVXVNVa2pqjuA9wL7LmQcSWqNPJ/c4CRpCGN572Q+SeuvYb58/CFV9cCB219J8oOFFEuyXVWtaG8+DTh/rvUlaR7mk6RJZDZJGqlhmro1Se5XVZcAJLkvsGa+H0pyEnAAsE2Sq4B/BA5IshfNh4UvA16ywHlLEphPkiaT2SRppIZp6l4FnJHkUiDALsAL5vuhqjp0hsXvX7vpSdKczCdJk8hskjRS8zZ1VfXlJLsBu9ME00VVdWvnM5OkeZhPkiaR2SRp1OZt6pJsAvwl8AiaXf9fT/Keqvp115OTpLmYT5ImkdkkadSGOfzyQ8Bq4J3t7UOBE4BDupqUJA3JfJI0icwmSSM1TFO3+7QzOJ2x0DM4SdIiM58kTSKzSdJIDfM9df+TZL+pG0keCnyzuylJ0tDMJ0mTyGySNFLD7Kl7KPDcJFe0t3cGLkxyHlBV9YedzU6S5mY+SZpEZpOkkRqmqXtC57OQpIVZL/Ppijf+QSfj7vz68zoZV1oPrZfZJGl8hvlKg8tHMZFJ5BsnabKtz/kkaXKZTZJGbZg9dZIkSdK83CAujccwJ0qRJEmSJE0omzpJkiRJ6rF5m7ok+yX5fpKbk/wmyZokq0YxOUmai/kkaRKZTZJGbZg9de8CDgUuBjYFXgy8s8tJSdKQzCdJk8hskjRSQ50opap+kmRZVa0BPpDkWx3PS5KGYj5JmkRmk6RRGqap+2WSjYFzk7wFWAFs1u20JGko5pOkSWQ2SRqpYQ6/PKxd7wjgFmAn4OldTkqShmQ+SZpEZpOkkRqmqTu4qn5dVauq6g1V9TfAk7qemCQNwXySNInMJkkjNUxT97wZlj1/kechSQthPkmaRGaTpJGa9TN1SQ4FngX8TpJTBu7aHLi+64lJ0mzMJ0mTyGySNC5znSjlWzQf7N0GeNvA8tXAD7uclCTNw3ySNInMJkljMWtTV1WXA5cDDxvddCRpfuaTpElkNkkal3k/U5dkvyTfT3Jzkt8kWZNk1SgmJ0lzMZ8kTSKzSdKoDXOilHcBhwIXA5sCLwbe2eWkJGlI5pOkSWQ2SRqpYb58nKr6SZJlVbUG+ECSb3U8L0kaivkkaRKZTZJGaZim7pdJNgbOTfIWmg8Ab9bttCRpKOaTpElkNkkaqWEOvzysXe8I4BZgJ+BPu5yUJA3JfJI0icwmSSM17566qro8yfL2+hu6n2m+ac4AAA4MSURBVJIkDcd8kjSJzCZJozbrnro0jk5yHXAR8H+TrEzy+tFNT5LuynySNInMJknjMtfhl68E9gceUlVbV9VWwEOB/ZP89UhmJ0kzM58kTSKzSdJYzNXUPRc4tKp+OrWgqi4FntPeJ0njYj5JmkRmk6SxmKup26iqrpu+sKpWAht1NyVJmpf5JGkSmU2SxmKupu43C7xPkrpmPkmaRGaTpLGY6+yXD0yyaoblATbpaD6SNAzzSdIkMpskjcWsTV1VLRvlRCRpWOaTpElkNkkal2G+fFySJEmSNKFs6iRJkiSpxzpr6pIcl+TaJOcPLLtXktOTXNz+u1VX9SVpNuaTpElkNklaqC731H0QeMK0ZUcBX66q3YAvt7cladQ+iPkkafJ8ELNJ0gJ01tRV1ZnADdMWPxU4vr1+PHBwV/UlaTbmk6RJZDZJWqhRf6Zu26paAdD+e+8R15ek2ZhPkiaR2SRpXhN7opQkhyc5K8lZK1euHPd0JAkwmyRNLvNJWn+Nuqm7Jsl2AO2/1862YlUdW1X7VNU+y5cvH9kEJa23hsons0nSiPneSdK8Rt3UnQI8r73+POCzI64vSbMxnyRNIrNJ0ry6/EqDk4BvA7snuSrJi4BjgAOTXAwc2N6WpJEynyRNIrNJ0kJt2NXAVXXoLHc9tquakjQM80nSJDKbJC3UxJ4oRZIkSZI0P5s6SZIkSeoxmzpJkiRJ6jGbOkmSJEnqMZs6SZIkSeoxmzpJkiRJ6jGbOkmSJEnqMZs6SZIkSeoxmzpJkiRJ6jGbOkmSJEnqsQ3HPQFpMb3pOX/Wybh//+FPdjKupPWH+SRpEplNS4N76iRJkiSpx2zqJEmSJKnHbOokSZIkqcds6iRJkiSpx2zqJEmSJKnHbOokSZIkqcds6iRJkiSpx2zqJEmSJKnH/PJxde5dR36uk3GPeNuTOxlXkiRpnHzvpLVlUydJWm918cbJN02SpFHz8EtJkiRJ6jGbOkmSJEnqMZs6SZIkSeoxmzpJkiRJ6jGbOkmSJEnqMZs6SZIkSeoxmzpJkiRJ6jGbOkmSJEnqMZs6SZIkSeoxmzpJkiRJ6jGbOkmSJEnqMZs6SZIkSeoxmzpJkiRJ6jGbOkmSJEnqMZs6SZIkSeqxDcdRNMllwGpgDXB7Ve0zjnlI0nTmk6RJZDZJmstYmrrWo6vqujHWl6TZmE+SJpHZJGlGHn4pSZIkST02rqaugC8mOTvJ4WOagyTNxHySNInMJkmzGtfhl/tX1dVJ7g2cnuSiqjpzcIU2sA4H2HnnnccxR0nrpznzyWySNCa+d5I0q7Hsqauqq9t/rwU+Dew7wzrHVtU+VbXP8uXLRz1FSeup+fLJbJI0Dr53kjSXkTd1STZLsvnUdeCPgfNHPQ9Jms58kjSJzCZJ8xnH4ZfbAp9OMlX/xKr67zHMQ5KmM58kTSKzSdKcRt7UVdWlwANHXVeS5mM+SZpEZpOk+fiVBpIkSZLUYzZ1kiRJktRjNnWSJEmS1GM2dZIkSZLUYzZ1kiRJktRjNnWSJEmS1GM2dZIkSZLUYzZ1kiRJktRjNnWSJEmS1GM2dZIkSZLUYzZ1kiRJktRjNnWSJEmS1GM2dZIkSZLUYzZ1kiRJktRjNnWSJEmS1GMbjnsCkqTJtv879+9k3G/+1Tc7GVeSpPWNTd2E8E2TJEnS8HzvJP0vD7+UJEmSpB6zqZMkSZKkHrOpkyRJkqQes6mTJEmSpB6zqZMkSZKkHrOpkyRJkqQes6mTJEmSpB6zqZMkSZKkHrOpkyRJkqQes6mTJEmSpB6zqZMkSZKkHrOpkyRJkqQes6mTJEmSpB7bcNwTkDSZPv6JfRd9zGcc8r1FH1PS+qWLbALzSdK6GXc2uadOkiRJknrMpk6SJEmSesymTpIkSZJ6zM/USZKkdXL00Uf3YkxJWqps6qSe6OoNjm+cJEnSUrQ+vXcay+GXSZ6Q5MdJfpLkqHHMQZJmYj5JmkRmk6S5jLypS7IMeDfwRGAP4NAke4x6HpI0nfkkaRKZTZLmM449dfsCP6mqS6vqN8BHgaeOYR6SNJ35JGkSmU2S5jSOpm4H4MqB21e1yyRp3MwnSZPIbJI0p1TVaAsmhwCPr6oXt7cPA/atqr+att7hwOHtzd2BHy+g3DbAdeswXWuNvt5SrTXqen2otUtVLV/syayLYfLJbJqoeku11qjr+djuaqLyaYTvnfxd6Ge9pVpr1PX6UGvWbBrH2S+vAnYauL0jcPX0larqWODYdSmU5Kyq2mddxljfa4263lKtNep6S7XWCMybT2bT5NRbqrVGXc/H1gsjee/k70I/6y3VWqOu1/da4zj88vvAbkl+J8nGwDOBU8YwD0maznySNInMJklzGvmeuqq6PckRwBeAZcBxVXXBqOchSdOZT5ImkdkkaT5j+fLxqjoNOG0EpdbpEClrjaXeUq016npLtVbnRpRPS/n1WaqPzeexn/WWTD4twWwadT0fW/9qjbper2uN/EQpkiRJkqTFM47P1EmSJEmSFsmSbOqSHJfk2iTnj6DWTknOSHJhkguSvKLDWpsk+V6SH7S13tBVrYGay5L8T5LPj6DWZUnOS3JukrM6rrVlkk8muah97R7WUZ3d28czdVmV5JVd1Bqo+dft78f5SU5KskmHtV7R1rmg68e1FJhNi1rTbFr3WiPNJ7Npci3VbGrrLdl8GmU2tfWW5HunUWZTW6+bfKqqJXcBHgnsDZw/glrbAXu31zcH/i+wR0e1Atyjvb4R8F1gv44f398AJwKfH8FzeRmwzYh+R44HXtxe3xjYcgQ1lwE/p/mOka5q7AD8FNi0vf1x4Pkd1doTOB+4O83nc78E7DaK16+vF7NpUWuaTYtbt9N8Mpsm+7JUs6mtsWTzaZTZ1NZbcu+dRplN7fid5dOS3FNXVWcCN4yo1oqqOqe9vhq4kOYXpItaVVU3tzc3ai+dfSgyyY7AnwDv66rGOCTZguY/sPcDVNVvquqmEZR+LHBJVV3ecZ0NgU2TbEgTGnf5LqNFcn/gO1X1y6q6Hfga8LSOai0JZtPiMJs6MYp8Mpsm1FLNpraG+bQIlvh7p1FlE3SYT0uyqRuXJLsCD6LZCtRVjWVJzgWuBU6vqs5qAf8OvBq4o8Magwr4YpKzkxzeYZ37AiuBD7SHR7wvyWYd1pvyTOCkLgtU1c+AtwJXACuAX1TVFzsqdz7wyCRbJ7k7cBB3/nJcTQizaZ0t9WyCjvPJbNJMRpFNbZ2lmk+jyiZYou+dRpxN0GE+2dQtkiT3AE4GXllVq7qqU1VrqmovYEdg3yR7dlEnyZOAa6vq7C7Gn8X+VbU38ETgZUke2VGdDWkOM/mvqnoQcAtwVEe1AEjzZbFPAT7RcZ2tgKcCvwNsD2yW5Dld1KqqC4F/AU4H/hv4AXB7F7W0cGbToliy2QSjySezSdONKptgSefTqLIJluh7p1FmE3SbTzZ1iyDJRjTB9JGq+tQoara7vL8KPKGjEvsDT0lyGfBR4DFJPtxRLQCq6ur232uBTwP7dlTqKuCqgS11n6QJqi49ETinqq7puM7jgJ9W1cqqug34FPDwropV1furau+qeiTNoTsXd1VLa89sWhxLPJtgNPlkNum3xpFNsPTyaYTZBEv3vdNIswm6yyebunWUJDTHF19YVW/vuNbyJFu21zel+UW8qItaVfWaqtqxqnal2fX9larqbMtFks2SbD51Hfhjml3Ui66qfg5cmWT3dtFjgR91UWvAoXR86GXrCmC/JHdvfzcfS/N5hU4kuXf7787A0xnNY9QQzKbFsR5kE4wmn8wmAaPNprbeksynUWYTLOn3TiPNJugunzZcjEEmTZKTgAOAbZJcBfxjVb2/o3L7A4cB57XHawO8tqpO66DWdsDxSZbRNOQfr6rOT+c9ItsCn27+ntgQOLGq/rvDen8FfKTdtX8p8IKuCrXHTB8IvKSrGlOq6rtJPgmcQ7M7/3+AYzsseXKSrYHbgJdV1Y0d1uo9s6mXlmw2wejyyWyabEs4m2Dp5tOoswmW4HunMWQTdJRPqersBECSJEmSpI55+KUkSZIk9ZhNnSRJkiT1mE2dJEmSJPWYTZ0kSZIk9ZhNnSRJkiT1mE2dOpXkPkk+muSSJD9KclqS30vS2XepSNIwzCdJk8hs0kIsye+p02Rov8Tx08DxVfXMdtleNN+tIkljYz5JmkRmkxbKPXXq0qOB26rqPVMLqupc4Mqp20l2TfL1JOe0l4e3y7dLcmaSc5Ocn+SPkixL8sH29nlJ/nr0D0nSEmE+SZpEZpMWxD116tKewNnzrHMtcGBV/TrJbsBJwD7As4AvVNWbkiwD7g7sBexQVXsCJNmyu6lLWuLMJ0mTyGzSgtjUadw2At7VHlqwBvi9dvn3geOSbAR8pqrOTXIpcN8k7wROBb44lhlLWl+YT5Imkdmku/DwS3XpAuDB86zz18A1wANptjJtDFBVZwKPBH4GnJDkuVV1Y7veV4GXAe/rZtqS1gPmk6RJZDZpQWzq1KWvAHdL8hdTC5I8BNhlYJ17Aiuq6g7gMGBZu94uwLVV9V7g/cDeSbYBNqiqk4HXAXuP5mFIWoLMJ0mTyGzSgnj4pTpTVZXkacC/JzkK+DVwGfDKgdX+Ezg5ySHAGcAt7fIDgFcluQ24GXgusAPwgSRTGyNe0/mDkLQkmU+SJpHZpIVKVY17DpIkSZKkBfLwS0mSJEnqMZs6SZIkSeoxmzpJkiRJ6jGbOkmSJEnqMZs6SZIkSeoxmzpJkiRJ6jGbOkmSJEnqMZs6SZIkSeqx/wc0XH8BTGxuRgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1080x360 with 3 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Class distributions visualization\n",
    "fig, axes = plt.subplots(1, 3, sharex = True, figsize = (15,5))\n",
    "fig.suptitle('The distribution of classes in different sets')\n",
    "\n",
    "# Train\n",
    "sns.barplot(ax = axes[0], x = Distribution_per_df.index, y = Distribution_per_df['train data %'].values)\n",
    "axes[0].set_title('train data')\n",
    "axes[0].set(xlabel ='Class', ylabel='Data points (%)')\n",
    "\n",
    "# Test\n",
    "sns.barplot(ax = axes[1], x = Distribution_per_df.index, y = Distribution_per_df['test data %'].values)\n",
    "axes[1].set_title('test data')\n",
    "axes[1].set(xlabel ='Class', ylabel='Data points (%)')\n",
    "\n",
    "# Validation\n",
    "sns.barplot(ax = axes[2], x = Distribution_per_df.index, y = Distribution_per_df['validation data %'].values)\n",
    "axes[2].set_title('validation data')\n",
    "axes[2].set(xlabel = 'Class', ylabel = 'Data points (%)')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Based on above tables and figures, our train, test, and cv sets have almost the same class distributions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id=6 > 6. Feature Engineering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id='61'>A. Handeling Categorical Features </a> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In our dataset, we have three categorical variable. In this section, We are going to \n",
    "\n",
    "- Evaluate the impact of each feature.\n",
    "- Transform the categorical values into numeric to enable our machine learning algorithm to understand features, accept them as input, and work with them.\n",
    "- Evaluate each transformed feature.\n",
    "- Concatenate important features\n",
    "\n",
    "We can encode categorical features using different techniques. Here, we consider the following encoding methods:\n",
    "\n",
    "- One-hot encoding\n",
    "- Response Encoding\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# code for response coding with Laplace smoothing.\n",
    "# alpha : used for laplace smoothing\n",
    "# feature: ['gene', 'variation']\n",
    "# df: ['train_df', 'test_df', 'cv_df']\n",
    "# algorithm\n",
    "# ----------\n",
    "# Consider all unique values and the number of occurrence of given feature in train data dataframe\n",
    "# build a vector (1*9) , the first element = (number of times it occured in class1 + 10*alpha / number of time it occurred in total data+90*alpha)\n",
    "# gv_dict is like a look up table, for every gene it store a (1*9) representation of it\n",
    "# for a value of feature in df:\n",
    "# if it is in train data:\n",
    "# we add the vector that was stored in 'gv_dict' look up table to 'gv_fea'\n",
    "# if it is not there is train:\n",
    "# we add [1/9, 1/9, 1/9, 1/9,1/9, 1/9, 1/9, 1/9, 1/9] to 'gv_fea'\n",
    "# return 'gv_fea'\n",
    "# ----------------------\n",
    "\n",
    "# get_gv_fea_dict: Get Gene varaition Feature Dict\n",
    "def get_gv_fea_dict(alpha, feature, train_df):\n",
    "    value_count = train_df[feature].value_counts()\n",
    "    \n",
    "    # gv_dict : Gene Variation Dict, which contains the probability array for each gene/variation\n",
    "    gv_dict = dict()\n",
    "    \n",
    "    # denominator will contain the number of time that particular feature occured in whole data\n",
    "    for i, denominator in value_count.items():\n",
    "        # vec will contain (p(yi==1/Gi) probability of gene/variation belongs to perticular class\n",
    "        # vec is 9 diamensional vector\n",
    "        vec = []\n",
    "        for k in range(1,10):\n",
    "            # print(train_df.loc[(train_df['Class']==1) & (train_df['Gene']=='BRCA1')])\n",
    "            #         ID   Gene             Variation  Class  \n",
    "            # 2470  2470  BRCA1                S1715C      1   \n",
    "            # 2486  2486  BRCA1                S1841R      1   \n",
    "            # 2614  2614  BRCA1                   M1R      1   \n",
    "            # 2432  2432  BRCA1                L1657P      1   \n",
    "            # 2567  2567  BRCA1                T1685A      1   \n",
    "            # 2583  2583  BRCA1                E1660G      1   \n",
    "            # 2634  2634  BRCA1                W1718L      1   \n",
    "            # cls_cnt.shape[0] will return the number of rows\n",
    "\n",
    "            cls_cnt = train_df.loc[(train_df['Class']==k) & (train_df[feature]==i)]\n",
    "            \n",
    "            # cls_cnt.shape[0](numerator) will contain the number of time that particular feature occured in whole data\n",
    "            vec.append((cls_cnt.shape[0] + alpha*10)/ (denominator + 90*alpha))\n",
    "\n",
    "        # we are adding the gene/variation to the dict as key and vec as value\n",
    "        gv_dict[i]=vec\n",
    "    return gv_dict\n",
    "\n",
    "# Get Gene variation feature\n",
    "def get_gv_feature(alpha, feature, df, train_df):\n",
    "    # print(gv_dict)\n",
    "    #     {'BRCA1': [0.20075757575757575, 0.03787878787878788, 0.068181818181818177, 0.13636363636363635, 0.25, 0.19318181818181818, 0.03787878787878788, 0.03787878787878788, 0.03787878787878788], \n",
    "    #      'TP53': [0.32142857142857145, 0.061224489795918366, 0.061224489795918366, 0.27040816326530615, 0.061224489795918366, 0.066326530612244902, 0.051020408163265307, 0.051020408163265307, 0.056122448979591837], \n",
    "    #      'EGFR': [0.056818181818181816, 0.21590909090909091, 0.0625, 0.068181818181818177, 0.068181818181818177, 0.0625, 0.34659090909090912, 0.0625, 0.056818181818181816], \n",
    "    #      'BRCA2': [0.13333333333333333, 0.060606060606060608, 0.060606060606060608, 0.078787878787878782, 0.1393939393939394, 0.34545454545454546, 0.060606060606060608, 0.060606060606060608, 0.060606060606060608], \n",
    "    #      'PTEN': [0.069182389937106917, 0.062893081761006289, 0.069182389937106917, 0.46540880503144655, 0.075471698113207544, 0.062893081761006289, 0.069182389937106917, 0.062893081761006289, 0.062893081761006289], \n",
    "    #      'KIT': [0.066225165562913912, 0.25165562913907286, 0.072847682119205295, 0.072847682119205295, 0.066225165562913912, 0.066225165562913912, 0.27152317880794702, 0.066225165562913912, 0.066225165562913912], \n",
    "    #      'BRAF': [0.066666666666666666, 0.17999999999999999, 0.073333333333333334, 0.073333333333333334, 0.093333333333333338, 0.080000000000000002, 0.29999999999999999, 0.066666666666666666, 0.066666666666666666],\n",
    "    #      ...\n",
    "    #     }\n",
    "    gv_dict = get_gv_fea_dict(alpha, feature, train_df)\n",
    "    # value_count is similar in get_gv_fea_dict\n",
    "    value_count = train_df[feature].value_counts()\n",
    "    \n",
    "    # gv_fea: Gene_variation feature, it will contain the feature for each feature value in the data\n",
    "    gv_fea = []\n",
    "    # for every feature values in the given data frame we will check if it is there in the train data \n",
    "    # then we will add the feature to gv_fea if not we will add [1/9,1/9,1/9,1/9,1/9,1/9,1/9,1/9,1/9] \n",
    "    # to gv_fea\n",
    "    for index, row in df.iterrows():\n",
    "        if row[feature] in dict(value_count).keys():\n",
    "            gv_fea.append(gv_dict[row[feature]])\n",
    "        else:\n",
    "            gv_fea.append([1/9,1/9,1/9,1/9,1/9,1/9,1/9,1/9,1/9])\n",
    "#             gv_fea.append([-1,-1,-1,-1,-1,-1,-1,-1,-1])\n",
    "    return gv_fea"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Unique Genes : 236\n",
      "BRCA1     154\n",
      "TP53      104\n",
      "EGFR       89\n",
      "PTEN       88\n",
      "BRCA2      83\n",
      "KIT        63\n",
      "BRAF       60\n",
      "ALK        43\n",
      "PDGFRA     41\n",
      "ERBB2      41\n",
      "Name: Gene, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Check Number of Unique Genes\n",
    "unique_genes = X_train['Gene'].value_counts()\n",
    "print('Number of Unique Genes :', unique_genes.shape[0])\n",
    "# the top 10 genes that occured most.predict_proba\n",
    "print(unique_genes.head(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAEJCAYAAACE39xMAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAgAElEQVR4nO2dfVxc5ZmwrzsEEgIkEAhJCEkmIWKIYIhgrJ9Vq6u2brVW39V23VprrV213W27u7rdfetud9e22/btp2vtx1q3rlZra+2n9Su11moCikKCwZBMDIGEQAIZCMnw8bx/nAEnJwMMMMPhzHNfvx+/zMw5c859X+fJPc8885xzxBiDoiiK4n9meR2AoiiKkhi0oCuKoqQIWtAVRVFSBC3oiqIoKYIWdEVRlBRBC7qiKEqKoAXdMkTkfBExIlIwxe0EItupTlRsE9h3g4jcFfU8KCKfScJ+jnMlIjeISE+i9xO1v6TkMRlE5C4R2R/J/wav41HiQwt6ghGRxSLydRFpFpFjIrJXRH4jIu/2OrbJIiKbRORbrpf3AEuBOg9CcnM6cE88K4rI/SLyyzi3+yJOjp2TDWyUGO4SkYYYi+LOI5mISDnwOeAWnPx/PMa67xSRJ0TkQKS97xSRn4jIu6YrXuVtZnsdQCohIgHgj0AIuBN4DedD813AvcAKr2JLNMaYQWCf13EAGGMOJHqbIpJujAkzjTkmI49Jsiby7+NmjDMPReRjOB9ADwJ/AewClgBVwDeAU5Icp+LGGKN/CfoDfg20AtkxluVFPTbA1a7lQeAzrnU+DvwcOAI0ARcAxcCTQC9O7/i0qPfcAPS4tnt+ZFsFozzPBx4CWoA+YCvw4aj33x9ZP/ovEPkzQDXOh1YLcLtr36WRdTZEni8A7gPacT70fg9Uj+O0MOKgD9gN3Ag0AHeN4e5jEV9HgQMRX7OBu2Lkcn5ULtcBz0b2dVsMVzcAPcCfR23/OWB11L7vAhpcOYwcl8hjdww3jJLHCuBnEVch4KdAsXtfwLVAc2Sdx4fjHcNpBfB0JM+DkWO8IGqbx8U3yjaWA8eAr46yXFzPz4oc7yPAXuC/gPlRyzfhfDj8B9ARaSNfBmZFrZMBfBGnrfUCW4BLopan43yQtEZi2wN8weu6MJ1/OuSSIERkIXAp8C1jzAnjrMaYQ5PY7D8BDwPrgRqcwvt9nIa/Aafh3j/JkIeZC7wCXI7To/o68J2or8yfBP4E/DfO1++lOP9RRjDGDEVi+6Br2x8EthljXhURAX4FLIvsawPwPPCsiCwdI777cXqMFwFXAn+FU4BjEhnT/zbwL8DJkff9NrL4y8AjOMVsOJcXo95+N47bdTiFMRZzcIYjPgycCaQBP4vkFw8/Br4CbI+K4YQhjcj2HgcWAxfifJgXAY+79hXA6R2/D/gzHK//PtrORWQejo8eYGPkfWcBP4is8mXgo5HHw/HF4mqcAvulWAtNpMJG9lkB/A54AqctXwVURu1zmA8CA5F4bgP+JpLbMP8NvBP4AM6H0g+BX4jI+sjyT0TyuRY4KfLe7aPEn5p4/YmSKn84/zkM8L441o23h3531PPyyGufinrtfGL0IF3bda9z3PNR4nsY+F7U8004H1TR6wQi26mOPD818nxN1DpvAndGHl+IU0QyXdupA/5+lDiGe/hnR722EhhklB46TrHoBnJG2eb9wC9HyeXT47i7YYx4Loo8v4sxeuijrRMjj4sj2w1ELV8NDLn2dZRI7zry2meBHWMc24+6/UTluSby/GpG6ZlHvee/gG7Xa5dHjvHw37mR1x8Avu9atzKyz8KoNvYn1zpPEWmHQEkk9xWudR4H7ok8/gbwDK5vBzb9aQ89ccTbQ5sIr0c93h/5tz7Ga4WT3YGIpInIZ0XkdRHpjMziuIoJjvcbY16PxPaByHbPwPlP+L+RVaqAecABEekZ/sP5oCoZZbNlOP+JN0ftZzfON5PReApnaGaXiDwoIh8SkZw406iJY53R4lkX5z7ipQxoNcYEo/a1M8a+dhtjuqOetzJ2eygDXjfGhKJeexEnr4nm4B5ffw6nUF8AZOF8ewHn2P+l67j/MbIs+thHt3c4PpfTcP6PbXNt5z1R27g/sv8mEfm2iLxHRKyqcfqjaOJ4E6eBl+GMe46F4cQPgPQY6/W73jPaa8ONdijO7UbzGeDTOEMr9Tg9q/9gch8SD+KMcf8rztfnP0QK3nCM+4FzY7zv8Cjbm/CHpDEmJCKnAefh9HLvBP5DRE43xoz1QQDOuOxUmcwxiIVwYsEcJvr1/hjLxipi8W53PJqABSKy1BjTBmCM6QV2xJgSOwv4HvD/Ymxnb9TjsXKZFXl+eoz1+iL7fyUyMeFSnG+EPwReE5GLjTMsmPJY9emVTIwxB3F+fLtNRLLdy0UkN+rpAaLGJkVkMaOPVU6EA8A8EZkf9VrlOO85B/iFMeZ/jDF1OD+ulbrWCfN2b2ssHgTWiMg7cMYvfxS17BWc8eAhY8wO11/7KNtrxGmjpw+/ICIrcMaSR8UYM2CMedYYcyfOUFAWznDARHIZjdHiaYy8dABY7Brndh+DeGLYBiyLFKjhfa2O7GvbZAKP2u5617eWs3Dyaoz9lpj8BKew3hnHuq8Ap8Q47juMMX1x7u9VnA+jJTG2MfKhYIwJGWMeNcZ8HKf3fiFvz9pJebSgJ5a/xml0NSJyjYicLCJrReTjHP918lngVhGpFpENOF8VjyZg/y/j9DLvFpE1IvL+SExj0QS8S0TOEZG1wLeAVa51gsDGyMlEBaN9jTXGtOD80HkvzoyWR6MWP43zNfvnInKZiKwSkTNF5F9EJFavHWPMdpwf8L4TWbcSx9WoRUBELheRT4rIBhFZiTMElMPbxSoIlEeOTYGITLT3PAB8LSqeH+LMDHo6snwTsBD4RxEpEZGP4IxJRxMEVorIaZEY5sTYz9M4014fFJGqyI+9D+IUx2cnGHM0D+K0kQdEpEJEzgO+A/zUGLMj3o0YY/bg/Gh5m4j8SEQujLSPDcCnIqsNRv79Ik77uTdyXNZEjtN3JrC/pkjs94vI1SKyOvL/5zMichWAiHxKRK4TkTIRWYNz7A/jzIqxAi3oCcQYswtnrO8pnEb8Os5/vvfiTKUb5tPATpz//D/B+To6Wi91Ivs/iDPUcTHO8MnNwD+P87Z/wxkT/g1OMe7F+Y8TzZdxepXbcHqgY42v/w/OTIZfGWO6omIzwLtxfHwXZ/bBIzgzUcYaCrkBZ37zs8AvcMbkg2Os34UzG+Zp4A2cIaWbjDF/iCz/Lk5xr4nkcvYY24rFMZxZJA/gfIDOAq6K5IcxphFnuunNOMf/YpwhrGgew5ni+kwkhuvcO4ls78rI8k0449P7gCuH9zUZjDFHgEuA+TjH/ec4s5hunMS27sE5xyIX51i+ifMBfArw3mHnkd9XzsP58fn3OB9Ud/P2b0Dx8mGcmS5fwjm2v4xsd3hYLwT8XSSvV3C+GV0WydkKZAptQ1EURZlBaA9dURQlRdCCriiKkiJoQVcURUkRtKAriqKkCJ6dWFRQUGACgcCk3hsOh8nIyEhsQD5DHagDUAc25l9bW9thjFkUa5lnBT0QCFBTE8+Z1icSCoXIyYn3bO7URB2oA1AHNuYvIrtHW+bLIZfa2lqvQ/AcdaAOQB3Ynr8bXxb0zMxMr0PwHHWgDkAd2J6/G18W9MmOvacS6kAdgDqwPX83vizojY0TuYZQaqIO1AGoA9vzd+PLgq6fyuoA1AGoA9vzdzNuQReRH4hI+yh3KUccviEiOyI3STgt8WEeTygUGn+lFEcdqANQB7bn7yaeHvr9OBeMH43LcO7fdxLOFeb+a+phjU1nZ2eydzHjUQfqANSB7fm7GXceujHm+eiL7MfgCuCByCU9XxKR3Oi7mCSDqqqqZG3aN6gDdQDqYCbnf2xgkK4j/RzsDXOoN8zBI2EOHennUG+YDStyOfekmOcGTYlEnFi0jOPvAt8See2Egi4iN+P04ikqKmLTpk2UlZURDAbp6+ujqqqK2tpaCgsLycjIoKWlhfLycpqamhgcHKSiooK6ujo6OjqoqKigra2NyspK6uvrSUtLo7S0lIaGBoqLiwmHw7S3t49sMzMzk0AgQGNjI4FAgFAoRGdn58jy7OxsioqKaGpqoqSkhM7OTrq6ukaW5+bmkp+fT3NzM6WlpbS2ttLT0zOyPD8/n5ycHILB4KRyWrrUuWFRvDk99thjlJWVpVROEz1O3d3drFy5MqVymuhxeuyxx6iqqkqpnCZynF544QUuv/zypOf08pYaTMY85i4ooKEpSNbCQlo7D9Pe3cu8vEKaW/ZxzMzmyOAsOkJHOTIgHOkf/a53f3X6EjIO7pzUcRqLuK6HHumh/9IYUx5j2a9w7k7/QuT5Mzh3cR9zxn91dbWZ7JmiNTU1VFdXT+q9qYI6UAegDiaT/8DgEF19Tk/5YG+YQ0fCHOzt59CRqJ50b5iDkd70od4woWMDo24ve85s8rLSWTgvg9x5GSzMyiBvXgYLs9LJy8o4/vWsdHIzM8iYPfn5KCJSa4yJmXQieugtwPKo58WMfQeaKVNUNOYtJa1AHagDUAeLlyzl4HGF+fiiPDzEMVKke8McPjp6cZ6XkRYpxhnkZWWwKn9eVDF2inNeVvrIOrnz0pkzeyq3qE0siSjoT+DcV/Bh4AygO5nj5wBNTU3WN2R1oA4gtRwMDRm6+yI95eFe83HF+cSedPeRfkYbY5ibPitSgJ3iuzxv3kgRfrsXnXFccZ6bPnOK82QYt6CLyEPA+UCBiLQAnwPSAYwx9+LcG/HdwA7gCM59/5JKSUlJsncx41EH6gBmroOhIUPo6EDkh8DwicMbkccjveoj/XQdCTM0SnXOmB1dnNNZVzSfhVkZpA30sXJJwUjRji7SmRn+Ls6TIZ5ZLifcwNa13AC3JiyiOOjs7GT58uXjr5jCqAN1ANPjwBhD6NjASFEemblxxDXUcSS6UPczOEp1zkibNTJskTcvg7VL5o+MQedFinH08MbCrAwy09MQkRO2VVdXR2XlqqTm7yc8u3zuVOjq6hp/pRRHHagDmJwDYwwHe8Ps7eo7odccPbxxqLd/5PnAKMV59iw5rviuKcyOeh75YXBexnHj0lkZsYvzZNA2cDxxzXJJBlOZ5WLjNZDdqAN1AGM7ODYwyO7OI+w80EPzgV52HuhlZ0cPOw/00t3Xf8L6abOEvHnprh5yBnnuMefIa3lZGeTMmZ2w4jwZbGwDyZ7lMu3U1tZy/vnnex2Gp6gDdQDOtL11VWfSfMAp1NFFu+XQkePGpBfPn8PqgmwuP3Upqxdls2LhPPKz3y7cOXNmM2uWd8V5MmgbOB5fFvTc3FyvQ/AcdWCXg77w4Eihji7aO/b30ffkMyPrZaansaogi1OLF3DlhmWULMpidUE2qxZlkT3Hl//dx8SmNhAPvjzC+fn5XofgOeog9RwMDRlau/siRbuHnR29I49bu4+OrCcCRQsyWb0oi3eXLWT96iWsLshm9aIslsyf67te9lRItTYwVXxZ0Jubm62f3aAO/OsgdLT/uF72zgO9NB/oIdjZy9Go08Vz5sxm9aIszlidz+qCLFYvcor2qoKskfnSmzZt4vwzAx5l4j1+bQPJwpcFvbS01OsQPEcdzHwHHT3HqG/ppnnkR0mn130gdGxknVkCyxfOY3VBFuesKRgp2qsXZbEoe864PzjOdAfJxvb83fiyoLe2tqbM2XGTRR3MLAdDQ4bmAz3U7D5ETfAQtbsPEuw8MrI8d146qwuyeGfpIqdgF2RTsiiLFfnzpnTq+Exy4AW25+/GlwW9p6fH6xA8Rx1466AvPMhrLV3U7j5ETfAgr7zVNTIVcGFWBlUr87h24wo2LM/lpMU5LMzKSEoctrcD2/N348uCPpOvgTxdqIPpdbD/8NFI8XZ631tbD4+cbLOmMJtLT1lCVSCP6pV5rCrImra52ba3A9vzd+PLgq5zT9UBJM/B4JBh+74QtW8dojZ4kJrdh2g51AfAnNmzWL88l4+et5rqlXmctiKPvCT1vuPB9nZge/5ufFnQdaqSOoDEOeg5NsBre7qoCR6iZvdB6t7qGrn+9aKcOVSvzOOGswJUrczjlKIFU7qWdaKxvR3Ynr8bXxZ02071jYU6mLyDvV191O5+u/fd2HaYIePM7z55cQ7vrSyiamUe1SsXsnxhpqento+H7e3A9vzd+LKgB4NBAoGA12F4ijqIz8HA4BCNbSFqdzvFu3b3IdoiJ+nMy0ijcnkut16whqqVeWxYkceCzPRpiDxx2N4ObM/fjS8LellZmdcheI46iO2gLzzI5uDBkd533Z4ujoQHAVi6YC5VK/NGet9lS3OYnTZzhk8mg+3twPb83fiyoAeDQRYvXux1GJ6iDt520B46yrON7Tzd2M4LOw5wtH+IWQJlS+dzdVWxU8ADC1mWm+l1yAnH9nZge/5ufFnQ+/r6vA7Bc2x2YIxh+/4QjzR0c3fNH6nb41wTe1luJn9RvZwL1hZSHViYkhejcmNzOwDN340vW7zOPbXPQXhgiC3Bgzy1bT9PN+4fmUa4vtjw6YtLuWjdYtYuyZnRP2AmA9vagRvb83fjy4Kuc0/tcNB9pJ9NTe08tW0/v99+gNCxAebMnsU5awq49YI1ZB7cwZWXnuN1mJ5iQzsYC9vzd+PLgl5YWOh1CJ6Tqg4O9oZ5om4vv926jy3BQwwOGQqy5/DuiqVctG4x56wpGLn577Ztetp3qraDeLE9fze+LOgZGd6dmTdTSCUHA4ND/L7pAI/WtPDMG/vpHzScvDiHW965movKFrO+ODfmNb5TycFksd2B7fm78WVBb2lpYc2aNV6H4Smp4ODN/SEerW3hp6/spaPnGAXZGXzozADXVC/n5CXjnzCSCg6miu0ObM/fjS8Lenl5udcheI5fHXT39fOL11p5tLaF1/Z0MXuWcMHaQq6pKuaCtYWkT2BeuF8dJBLbHdievxtfFvSmpiYKCgq8DsNT/ORgaMjwx+YOHq1p4cmt+zg2MMTJi3P4p/eUceWGZRRkz5nUdv3kIFnY7sD2/N34sqAPDg56HYLn+MHB0f5BHnulhe/9YRe7OnqZP3c2/6d6OddUF1OxbMGUpxj6wUGysd2B7fm78WVBr6io8DoEz5nJDg72hvmfP+3mgT8F6ewNU7FsAV+/tpJLTlkyci/MRDCTHUwXtjuwPX83vryQRV1dndcheM5MdLC7s5d/fryBs77wDP/v6SbWL8/loY++gyduO5srKpcltJjDzHQw3djuwPb83fiyh7506VKvQ/CcmeTg1bcOcd/zO/nt1n2kz5rFlRuKuOnc1ZQuTu6lTWeSA6+w3YHt+bvxZUFXvMcYw3Pb27l30042Bw8yf+5sPv7OEm44K0Dh/Lleh6coVuLLIZe2tjavQ/AcLx1s3nWQq+/9EzfeX8Perj7+7+XrePHOd/H3l66d1mKu7UAd2J6/G1/20CsrK70OwXO8cLCt9TD/+eQbPLf9AIvnz+Huqyq4uqp4QnPHE4m2A3Vge/5ufNlDr6+v9zoEz5lOB7s7e/nkw6/y7m/8gVfe6uLOy9by+7+7gOs2rvCsmIO2A1AHtufvxpc99LS0xM6W8CPT4aD98FG++ewOHtr8FrPThFsvKOHm80pmzG3atB2oA9vzd+PLgl5aWup1CJ6TTAdHwgPc81wz339hF/2DQ1y3cQW3X7hmxv3Yqe1AHdiev5u4vi+LyKUisl1EdojIHTGWLxCRX4jIayKyVUQ+nPhQ36ahoSGZm/cFyXBgjOHX9W1c9JXf863ndnDxusU88+l38vkry2dcMQdtB6AObM/fzbg9dBFJA74NXAy0AFtE5AljzLao1W4Fthlj/lxEFgHbReRBY0w4GUEXFxcnY7O+ItEOmg/0cNcTW/nDmx2ULZ3PN67bQHVgYUL3kWi0HagD2/N3E8+Qy0ZghzFmJ4CIPAxcAUQXdAPkiHNxjmzgIDCQ4FhHCIeT8jnhKxLl4Gj/IN945k2++4edzE1P41/eewofPGMFsz38sTNetB2oA9vzdxNPQV8G7Il63gKc4VrnW8ATQCuQA/yFMWbIvSERuRm4GaCoqIhNmzZRVlZGMBikr6+PqqoqamtrKSwsJCMjg5aWFsrLy2lqamJwcJCKigrq6uro6OggLS2NtrY2Kisrqa+vJy0tjdLSUhoaGiguLiYcDtPe3j6yzczMTAKBAI2NjQQCAUKhEJ2dnSPLs7OzKSoqoqmpiZKSEjo7O+nq6hpZnpubS35+Ps3NzZSWltLa2kpPT8/I8vz8fHJycggGg5PKafiMt3hz2rx5M6FQaEo5dQzN45tbDrP70DEuW7uQ96+ZxexwkL4j+Z7kNNHj1N3dTTgcntHHKdltb/PmzQwODqZUThM5Tps3b6awsDClchrvOI2FGGPGXkHkGuASY8xNkefXAxuNMbdHrXM1cDbwKaAEeApYb4w5PNp2q6urTU1NzbgBxiIUCpGTk9zTymc6U3HQPzjEN5/dwbef28Gi7Dl86epTOa90UYIjTD7aDtSBjfmLSK0xpjrWsni+V7cAy6OeF+P0xKP5MPBT47AD2AWsnUyw8VBbW5usTfuGyTrYvi/Eld/+I9945k2uqCziyb89z5fFHLQdgDqwPX838Qy5bAFOEpFVwF7gWuADrnXeAt4F/EFEFgMnAzsTGWg0mZmZydq0b5iog8Ehw3f/sJOv/q6J+Zmz+c71VVxyypIkRTc9aDtQB7bn72bcgm6MGRCR24AngTTgB8aYrSJyS2T5vcDngftFpB4Q4B+MMR3JCjoQCCRr075hIg7aQ0f5xEOv8tLOg1x6yhL+/X3l5E/yLkEzCW0H6sD2/N3EdWKRMebXwK9dr90b9bgV+LPEhjY6jY2NLF68eLp2NyOJ18FLOzu5/aFXCR3t58vXrOf9py2b8p2CZgraDtSB7fm78eWZovqpPL6DoSHDd57fyX8++QaB/Cx+9JEzOHlJav14pO1AHdievxtfFvRQKOR1CJ4zloPuI/18+tE6nm5s5z2nLuWL7z+V7Dm+PNRjou1AHdievxtf/i/v7Oz0OgTPGc1BfUs3H3+wlv2Hj3LXn6/jQ2cFUmaIxY22A3Vge/5ufFnQq6qqvA7Bc9wOjDH87+a3+JcntlGQncGPP3Ymp63I8yi66UHbgTqwPX83M//87hjo3NPjHRwJD/DpR17jsz9r4B0l+fzyE+emfDEHbQegDmzP340ve+jZ2dleh+A5ww52tPfw1w/W8mZ7D5+6uJTbLljDrFmpOcTiRtuBOrA9fze+LOhFRUVeh+A5RUVF/OK1Vu547HXmpKfxwI0bOfckf57xOVm0HagD2/N348shl6amJq9D8JTBIcPnf9HA7Q+9yslLcvjVJ86xrpiDtgNQB7bn78aXPfSSkhKvQ/CM0NF+PvlwHc/u6ucv37GCz/35KZ7e19NLbG4Hw9juwPb83fiyEtg6VWl3Zy9X3fMiv286wMeqFvBvV1ZYW8zB3nYQje0ObM/fjS976F1dXV6HMO3U7j7ER364BWPgf27cSLhFb71lYztwY7sD2/N348uCbtvc02ff2M9fP/gKS+bP5f4PbyRQkEVosV0OYmFbO4iF7Q5sz9+NL7+v2zT39LHaFj76QC1rCrP5ycfPIlCQBdjlYDTUgTqwPX83vuyh5+bmeh3CtHDf8838x6/f4Ow1+dz7l1XkzE0fWWaLg7FQB+rA9vzd+LKg5+fnex1CUhkaMnzht29w3/M7ec+pS/nq/1nPnNlpx62T6g7iQR2oA9vzd+PLIZfm5mavQ0gaxhj+9ZfbuO/5nfzVmSv5xrUbTijmkNoO4kUdqAPb83fjyx56aWmp1yEkja89/Sb3vxjkxrNX8c+Xl416pcRUdhAv6kAd2J6/G1/20Ftb3feoTg2+/8Iuvv7Mm1xTVTxmMYfUdTAR1IE6sD1/N74s6D09PV6HkHAeqdnD53+5jcvKl3D3VRXjXsM8FR1MFHWgDmzP340vC3qqzT39bUMbdzz2OueeVMDXrq1kdhxnf6aag8mgDtSB7fm78WVBT6W5p3948wCfeKiODSvy+M71VTF/AI1FKjmYLOpAHdievxtfFvRUmapUu/sQNz9QS0lhNj+44XTmZcT/G3WqOJgK6kAd2J6/G18W9Jwc/9+9vvlADzfev4UlC+bywI0bWZCZPv6bokgFB1NFHagD2/N348uCHgwGvQ5hShwIHeOG/95MeprwwI0bWZQzZ8Lb8LuDRKAO1IHt+bvx5Tz0srIyr0OYNEfCA3zkh1voCIX58cfewfKF8ya1HT87SBTqQB3Ynr8b7aFPIwODQ9z+v6/SsLebb163gVOLJ38dCr86SCTqQB3Ynr8bX/bQ+/r6vA5hwhhjuOsXW3nmjXY+f8UpXLRu8ZS250cHiUYdqAPb83fjyx66H+ee3vf8Tn700lt87J2ruf7MwJS350cHiUYdqAPb83fjy4Lut7mnv9u6j7t/8waXn7qUf7hkbUK26TcHyUAdqAPb83fjy4JeWFjodQhxs31fiL/9cR2nFi/gy9esZ9assU/pjxc/OUgW6kAd2J6/G18W9IyMDK9DiItDvWE++kAN8+bM5r7rq5mbHt9ZoPHgFwfJRB2oA9vzd+PLgt7S0uJ1COMyMDjEbQ+9wr7uo3zn+iqWLJib0O37wUGyUQfqwPb83fhylkt5ebnXIYzLv/2qkT/u6OTL16zntBV5Cd++HxwkG3WgDmzP301cPXQRuVREtovIDhG5Y5R1zheROhHZKiK/T2yYx9PU1JTMzU+ZR7bs4f4Xg3zknFVcXVWclH3MdAfTgTpQB7bn72bcHrqIpAHfBi4GWoAtIvKEMWZb1Dq5wD3ApcaYt0Qkqb9UDA4OJnPzU6ImeJDPPl7PuScVcOdliZnREouZ7GC6UAfqwPb83cTTQ98I7DDG7DTGhIGHgStc63wA+Kkx5i0AY0x7YsM8noqKimRuftK0dfdxy49eYVluJt+67rS4rms+WWaqg+lEHagD2/N3E88Y+jJgT9TzFuAM1zqlQLqIbAJygK8bYx5wb0hEbgZuBigqKmLTpk2UlZURDAbp60IVieYAABdHSURBVOujqqqK2tpaCgsLycjIoKWlhfLycpqamhgcHKSiooK6ujo6OjqoqKigra2NyspK6uvrSUtLo7S0lIaGBoqLiwmHw7S3t49sMzMzk0AgQGNjI4FAgFAoRGdn58jy7OxsioqKaGpqoqSkhM7OTrq6ukaW5+bmkp+fT3NzM6WlpbS2ttLT0/P28ryF/NNznfT0HeObV53CG/WvTCinpUuXAsSd0+OPP05ZWVlSc8rPzycnJ4dgMDip4zTRnCZ6nLq7u1m5cmVK5TTR4/T4449TVVWVUjlN5Di98MILXH755SmV03jHaSzEGDP2CiLXAJcYY26KPL8e2GiMuT1qnW8B1cC7gEzgT8B7jDGjDnBVV1ebmpqacQOMxfbt2zn55JMn9d5k8dWnmvjGM2/y9WsruaJyWdL3NxMdTDfqQB3YmL+I1BpjqmMti6eH3gIsj3peDLjvzNoCdBhjeoFeEXkeWA9Y8YvFSzs7+dazb/L+04qnpZgriqLEIp5B3i3ASSKySkQygGuBJ1zr/Bw4V0Rmi8g8nCGZxsSG+jZtbW3J2vSEOdQb5m8ermNlfhb/esUp07bfmeTAK9SBOrA9fzfj9tCNMQMichvwJJAG/MAYs1VEboksv9cY0ygivwVeB4aA7xljGpIVdGVlZbI2PSGMMfz9Y6/T2XuMn33obLLmTN+0/pniwEvUgTqwPX83cU3DMMb82hhTaowpMcb8e+S1e40x90at85/GmHXGmHJjzNeSFTBAfX19MjcfNz96aTdPbdvPP1y6lvJlC6Z13zPFgZeoA3Vge/5ufHnqf1pa4q6JMlne2HeYz/+qkfNPXsSNZ6+a9v3PBAdeow7Uge35u/FlQS8tLfV0/+GBIf7m4Trmz01P6BUUJ4LXDmYC6kAd2J6/G18W9IaGpA3Px8U3n32TN/aF+OL7KyjInvgNnhOB1w5mAupAHdievxtfFvTi4uRcHyUeGvZ2c8+mZq46bRnvKpvabeSmgpcOZgrqQB3Ynr8bXxb0cDjszX4HhvjMo69RkJ3B5y6fvimKMWPxyMFMQh2oA9vzd+PLgt7entRLxYzK8FDL3VdVsGBeuicxDOOVg5mEOlAHtufvxpcF3Ysbw0YPtVy41ruhlmH05rjqANSB7fm78WVBn+4bww4OGe78aT35Wd4PtQyjN8dVB6AObM/fjS8LemZm5rTu76HNb1G/t5vPvqfM86GWYabbwUxEHagD2/N348uCHggEpm1fB3vD/OeT23nH6oW8d33RtO13PKbTwUxFHagD2/N348uC3tiYtOt+ncCXfvsGvccG+NcryhGZ/hOIRmM6HcxU1IE6sD1/N74s6NP1qVzf0s3DW/Zww1kBShfnTMs+40V7JuoA1IHt+bvxZUEPhUJJ34cxhrt/00jevHQ+cdFJSd/fRJkOBzMddaAObM/fjS8LemdnZ9L38fumA7zY3Mkn3nUS8+fOjB9Co5kOBzMddaAObM/fjS8LerLnng4OGb7wmzdYsXAeHzxjZVL3NVl0/q06AHVge/5ufFnQkz339Gev7uWNfSH+7pKTyZg9MxXp/Ft1AOrA9vzdzMxqNQ7Z2dlJ2/bR/kG+8rvtrC9ewHsqxr/Ltlck04FfUAfqwPb83fiyoBcVJW8++H//MUhb91HuuKzMk+ucx0syHfgFdaAObM/fjS8LelNTU1K2293Xzz2bdnDh2kLOLMlPyj4SRbIc+Al1oA5sz9+NLwt6SUlJUrb7wItBQkcH+NTFM/8uKMly4CfUgTqwPX83vizoyZiq1HtsgB/8cRcXri2c9hs+TwadrqUOQB3Ynr8bXxb0rq6uhG/zoc1vcehIP7desCbh204GyXDgN9SBOrA9fze+LOiJnnt6tH+Q+57fyZmr86lamZfQbScLnX+rDkAd2J6/G18W9ETPPX20toX20DFuv9AfvXPQ+begDkAd2J6/G18W9Nzc3IRtq39wiHs3NbNhRe6Mn9kSTSId+BV1oA5sz9+NLwt6fn7iCu/P61rZ29XHbResmVGXxx2PRDrwK+pAHdievxtfFvTm5uaEbGdwyHDPczsoWzqfC9cWJmSb00WiHPgZdaAObM/fjS8LemlpYuaJ/6ahjZ0dvb7rnUPiHPgZdaAObM/fjS8Lemtr65S3YYzh2881s3pRFpeWL0lAVNNLIhz4HXWgDmzP340vC3pPT8+Ut/HsG+00th3mr89fQ9oMvmbLaCTCgd9RB+rA9vzd+LKgT3XuqTGGbz67g+K8TK6o9OfFfXT+rToAdWB7/m58WdCnOvf0xeZO6vZ0ccs7S0hP86UCnX+LOgB1YHv+bnxZzaY6Ven7L+yiIHsOV1cVJyii6Uena6kDUAe25+/GlwU9Jydn0u/dc/AIz21v57qNy5mbnpbAqKaXqThIFdSBOrA9fzdxFXQRuVREtovIDhG5Y4z1TheRQRG5OnEhnkgwGJz0e3/08m5mifCBM1YkLiAPmIqDVEEdqAPb83czbkEXkTTg28BlwDrgOhFZN8p6XwSeTHSQbsrKyib1vv7BIX5S08K71haydEFmgqOaXibrIJVQB+rA9vzdxNND3wjsMMbsNMaEgYeBK2KsdzvwGNCewPhiMtlP5WffaKezN8xfnL48sQF5gPZM1AGoA9vzdzM7jnWWAXuinrcAZ0SvICLLgPcBFwKnj7YhEbkZuBmcewFu2rSJsrIygsEgfX19VFVVUVtbS2FhIRkZGbS0tFBeXk5TUxODg4NUVFRQV1dHR0cHubm5tLW1UVlZSX19PWlpaZSWltLQ0EBxcTHhcJj29vaRbWZmZvKj1wZYMEdYOecI9fX1dHZ2jizPzs6mqKiIpqYmSkpK6OzspKura2R5bm4u+fn5NDc3U1paSmtrKz09PSPL8/PzycnJIRgMTiqnpUudG1LHm1NjYyMAgUCAxsZGAoEAoVDI1zkNH6d4c+ru7qauri6lcprocWpsbGTevHkpldNEjlNjYyMlJSUpldN4x2ksxBgz9goi1wCXGGNuijy/HthojLk9ap1Hga8YY14SkfuBXxpjfjLWdqurq01NTc24AcYiFApN+MeQA6FjvOPuZ7jp3FXceZn/v6ZNxkGqoQ7UgY35i0itMaY61rJ4hlxagOgximLAfb5tNfCwiASBq4F7ROTKScQaF5OZe/r4q3sZHDJcU+X/4RbQ+begDkAd2J6/m3iGXLYAJ4nIKmAvcC3wgegVjDGrhh9H9dAfT2Ccx1FYOLErIxpjeKRmDxtW5LKmMDtJUU0vE3WQiqgDdWB7/m7G7aEbYwaA23BmrzQCjxhjtorILSJyS7IDjEVGRsaE1n+tpZs323tSpncOE3eQiqgDdWB7/m7imodujPm1MabUGFNijPn3yGv3GmPujbHuDeONn0+VlpaWCa3/aM0e5qbP4vL14/+o4Bcm6iAVUQfqwPb83fjyTNHy8vK41z3aP8gTr7Vy6SlLmD83PYlRTS8TcZCqqAN1YHv+bnxZ0JuamuJe98mt+wgdHeCa6tQZboGJOUhV1IE6sD1/N74s6IODg3Gv+/ire1mWm8mZq1PrIj4TcZCqqAN1YHv+bnxZ0CsqKuJaL3S0nz/u6OSy8iXM8uFNLMYiXgepjDpQB7bn78aXBb2uri6u9Z7bfoDw4BCX+PAWc+MRr4NURh2oA9vzd+PLgh7PKbAAv9u6j4LsDE5bkZfkiKafeB2kMupAHdievxtfFvR4ODYwyKbtB7iobLEv7xmqKIoyUXxZ0Nva2sZd56WdB+k5NsCfnbJ4GiKafuJxkOqoA3Vge/5ufFnQKysrx13nqW37yExP46ySgmmIaPqJx0Gqow7Uge35u/FlQa+vrx9zuTGGp7e1c15pga9vMzcW4zmwAXWgDmzP340vC3pa2thFumHvYfYdPsrF61Jvdssw4zmwAXWgDmzP340vC3ppaemYy5/ato9ZAheuTd0rsY3nwAbUgTqwPX83vizoDQ0NYy7/3bb9VK9cyMKs1L0S23gObEAdqAPb83fjy4JeXFw86rI9B4/wxr4QF69Lzdktw4zlwBbUgTqwPX83vizo4XB41GXPNO4H4KIUL+hjObAFdaAObM/fjS8Lent7+6jLXt51kOULM1lVkDWNEU0/YzmwBXWgDmzP340vC3pVVdWoy17b00Xl8tQ71d/NWA5sQR2oA9vzd+PLgj7ajWHbDx+ltfso64sXTHNE04/eHFcdgDqwPX83vizomZmZMV9/raUbgMrludMZjieM5sAm1IE6sD1/N74s6IFAIObrr7d0kTZLOKUo9XvoozmwCXWgDmzP340vC3pjY2PM1+v2dHHy4hwyM1L/7LHRHNiEOlAHtufvxpcFPdansjGG1/Z0sX556vfOQXsmoA5AHdievxtfFvRQKHTCa8HOIxw+OsD64tQfP4fYDmxDHagD2/N348uC3tnZecJrr7d0AbDegh9EIbYD21AH6sD2/N34sqDHmntat6eLzPQ0TirM9iCi6Ufn36oDUAe25+/GlwU91tzT1/Z0UbFsAbPTfJnShNH5t+oA1IHt+bvxZfXLzj6+F94/OERD62FOteCEomHcDmxEHagD2/N348uCXlRUdNzzHe09hAeGqLCooLsd2Ig6UAe25+/GlwW9qanpuOe7OnoBWGPJ+Dmc6MBG1IE6sD1/N74s6CUlJcc9Hy7ogfzUvsJiNG4HNqIO1IHt+bvxZUF3T1XaeaCXJfPnkjVntkcRTT86XUsdgDqwPX83vizoXV1dxz3f1dGT8tc/d+N2YCPqQB3Ynr8bXxZ099zTXR29rFpkV0HX+bfqANSB7fm78WVBj557eqg3zKEj/ay2rIeu82/VAagD2/N3E1dBF5FLRWS7iOwQkTtiLP+giLwe+XtRRNYnPtS3yc19+/T+nZEfRG0bcol2YCvqQB3Ynr+bcQu6iKQB3wYuA9YB14nIOtdqu4B3GmNOBT4P3JfoQKPJz89/e8eWFvRoB7aiDtSB7fm7iaeHvhHYYYzZaYwJAw8DV0SvYIx50RhzKPL0JaA4sWEeT3Nz88jjXR09zJ4lLF84L5m7nHFEO7AVdaAObM/fTTzz/JYBe6KetwBnjLH+R4DfxFogIjcDN4NzhtemTZsoKysjGAzS19dHVVUVtbW1FBYWkpGRQUtLC+Xl5TQ1NTE4OEhFRQV1dXXMmTOH7du309bWxvbWuRRkCltefonS0lIaGhooLi4mHA7T3t4+ss3MzEwCgQCNjY0EAgFCoRCdnZ0jy7OzsykqKqKpqYmSkhI6Ozvp6uoaWZ6bm0t+fj7Nzc2UlpbS2tpKT0/PyPL8/HxycnIIBoOTymnp0qUAtLW1UVlZSX19PWlpaaPmFAqFePnll1Mqp4kepwULFlBXV5dSOU30OIVCIerr61Mqp4kcp1AoREdHR0rlNN5xGgsxxoy9gsg1wCXGmJsiz68HNhpjbo+x7gXAPcA5xpgxJ4hWV1ebmpqacQOMRU1NDdXV1QBc+rXnWZabyfdvOH1S2/Ir0Q5sRR2oAxvzF5FaY0zMpOMZcmkBlkc9LwZaY+zkVOB7wBXjFfOp0tPTA8DQkCHY2Wvd+Dm87cBm1IE6sD1/N/EU9C3ASSKySkQygGuBJ6JXEJEVwE+B640xSb+4wvDc032Hj3K0f8i6Oeig829BHYA6sD1/N+MWdGPMAHAb8CTQCDxijNkqIreIyC2R1f4vkA/cIyJ1IjK5sZQ4GZ57uvOAnTNcQOffgjoAdWB7/m7iuviJMebXwK9dr90b9fgm4KbEhjY6w1OVdnU4X7dWF9hzlcVhdLqWOgB1YHv+bnx5pmhOTg7gnFSUmZ7G4vlzPI5o+hl2YDPqQB3Ynr8bXxb0YDAIRK7hUpCFiHgbkAcMO7AZdaAObM/fjS8LellZGWDnRbmGGXZgM+pAHdievxtfFvRgMEh4YIg9B49QYuEPoqA9E1AHoA5sz9+NLwt6X18fbx08wpDB2h56X1+f1yF4jjpQB7bn78aXBb2qqoqdB5wZLqssnOECOv8W1AGoA9vzd+PLgl5bW/v2VRYtuo9oNDr/Vh2AOrA9fze+LOiFhYXs6uglPyuDBfPSvQ7HEwoLC70OwXPUgTqwPX83vizoGRkZ7Oyw8xouw2RkZHgdgueoA3Vge/5ufFnQW1paRuag20pLS4vXIXiOOlAHtufvxpcFfVVpGQdCxwhYXNDLy8u9DsFz1IE6sD1/N74s6C+97lzQcYVldymKpqkp6Re1nPGoA3Vge/5ufFnQ9/f0A7AsL9PjSLxjcHDQ6xA8Rx2oA9vzd+PLgp6R59yKqdjigl5RUeF1CJ6jDtSB7fm78WVBr3tzD3Nmz2JRtn1XWRymrq7O6xA8Rx2oA9vzdxPX9dBnGr1ksCwPK6+yOEw8N4xNddSBOrA9fze+7KHv7xlgWa69wy2Koiix8GVBb+s+RnGevTNcANra2rwOwXPUgTqwPX83vivoR8IDhPrt/kEUoLKy0usQPEcdqAPb83fju4K+95BzuUzbC3p9fb3XIXiOOlAHtufvxncFvWWkoNs95JKWluZ1CJ6jDtSB7fm78V1Bz5k7mwtOyrP6LFGA0tJSr0PwHHWgDmzP343vCnp1YCEfKgmzKMfeOegADQ0NXofgOepAHdievxvfFXSA4uJir0PwHHWgDkAd2J6/G18W9HA47HUInqMO1AGoA9vzd+PLgt7e3u51CJ6jDtQBqAPb83fjy4KuN4ZVB6AOQB3Ynr8bXxZ0vTGsOgB1AOrA9vzd+LKgZ2bafVIRqANQB6AObM/fjS8LeiAQ8DoEz1EH6gDUge35u/FlQW9sbPQ6BM9RB+oA1IHt+bsRY4w3OxY5AOye5NsLgI4EhuNH1IE6AHVgY/4rjTGLYi3wrKBPBRGpMcZUex2Hl6gDdQDqwPb83fhyyEVRFEU5ES3oiqIoKYJfC/p9XgcwA1AH6gDUge35H4cvx9AVRVGUE/FrD11RFEVxoQVdURQlRfBdQReRS0Vku4jsEJE7vI5nuhCRoIjUi0idiNREXlsoIk+JyJuRf/O8jjNRiMgPRKRdRBqiXhs1XxG5M9ImtovIJd5EnVhGcXCXiOyNtIM6EXl31LJUdLBcRJ4TkUYR2Soin4y8blVbiBtjjG/+gDSgGVgNZACvAeu8jmuacg8CBa7XvgTcEXl8B/BFr+NMYL7nAacBDePlC6yLtIU5wKpIG0nzOockObgL+EyMdVPVwVLgtMjjHKApkqtVbSHeP7/10DcCO4wxO40xYeBh4AqPY/KSK4AfRh7/ELjSw1gSijHmeeCg6+XR8r0CeNgYc8wYswvYgdNWfM0oDkYjVR20GWNeiTwOAY3AMixrC/Hit4K+DNgT9bwl8poNGOB3IlIrIjdHXltsjGkDp+EDhZ5FNz2Mlq9t7eI2EXk9MiQzPNSQ8g5EJABsAF5G20JM/FbQJcZrtsy7PNsYcxpwGXCriJzndUAzCJvaxX8BJUAl0AZ8JfJ6SjsQkWzgMeBvjDGHx1o1xmsp42E8/FbQW4DlUc+LgVaPYplWjDGtkX/bgZ/hfI3cLyJLASL/pvr9uEbL15p2YYzZb4wZNMYMAd/l7eGElHUgIuk4xfxBY8xPIy9b3xZi4beCvgU4SURWiUgGcC3whMcxJR0RyRKRnOHHwJ8BDTi5fyiy2oeAn3sT4bQxWr5PANeKyBwRWQWcBGz2IL6kM1zEIrwPpx1AijoQEQG+DzQaY74atcj6thCL2V4HMBGMMQMichvwJM6Mlx8YY7Z6HNZ0sBj4mdO2mQ38rzHmtyKyBXhERD4CvAVc42GMCUVEHgLOBwpEpAX4HPAFYuRrjNkqIo8A24AB4FZjzKAngSeQURycLyKVOMMIQeBjkLoOgLOB64F6EamLvPaPWNYW4kVP/VcURUkR/DbkoiiKooyCFnRFUZQUQQu6oihKiqAFXVEUJUXQgq4oipIiaEFXFEVJEbSgK4qipAj/H/nXWmJd00hHAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "s = sum(unique_genes.values);\n",
    "h = unique_genes.values/s;\n",
    "c = np.cumsum(h)\n",
    "plt.plot(c)\n",
    "plt.title(\"Cumulative distribution of Genes\", fontsize = 14)\n",
    "plt.grid(color='gray', linestyle='--', linewidth=0.5)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The figure shows that the first 50 genes have almost 75% of occurrence. Therefore, gene is important for us and we need to transform it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2124, 236)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# one-hot encoding of Gene feature.\n",
    "gene_vectorizer = CountVectorizer()\n",
    "X_train_gene_onehotCoding = gene_vectorizer.fit_transform(X_train['Gene'])\n",
    "X_test_gene_onehotCoding = gene_vectorizer.transform(X_test['Gene'])\n",
    "X_cv_gene_onehotCoding = gene_vectorizer.transform(X_cv['Gene'])\n",
    "\n",
    "X_train_gene_onehotCoding.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2124, 9)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# response-coding of the Gene feature\n",
    "# alpha is used for laplace smoothing\n",
    "alpha = 1\n",
    "\n",
    "X_train_gene_responseCoding = np.array(get_gv_feature(alpha, \"Gene\", X_train,X_train))\n",
    "X_test_gene_responseCoding = np.array(get_gv_feature(alpha, \"Gene\", X_test, X_train))\n",
    "X_cv_gene_responseCoding = np.array(get_gv_feature(alpha, \"Gene\", X_cv, X_train))\n",
    "\n",
    "X_train_gene_responseCoding.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variation feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Unique Variatis : 1923\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAEJCAYAAACE39xMAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAgAElEQVR4nO2dfVxc5ZmwrzsEEgIkEAhJCEkmIWKIYIhgrJ9Vq6u2brVW39V23VprrV213W27u7rdfetud9e22/btp2vtx1q3rlZra+2n9Su11moCikKCwZBMDIGEQAIZCMnw8bx/nAEnJwMMMMPhzHNfvx+/zMw5c859X+fJPc8885xzxBiDoiiK4n9meR2AoiiKkhi0oCuKoqQIWtAVRVFSBC3oiqIoKYIWdEVRlBRBC7qiKEqKoAXdMkTkfBExIlIwxe0EItupTlRsE9h3g4jcFfU8KCKfScJ+jnMlIjeISE+i9xO1v6TkMRlE5C4R2R/J/wav41HiQwt6ghGRxSLydRFpFpFjIrJXRH4jIu/2OrbJIiKbRORbrpf3AEuBOg9CcnM6cE88K4rI/SLyyzi3+yJOjp2TDWyUGO4SkYYYi+LOI5mISDnwOeAWnPx/PMa67xSRJ0TkQKS97xSRn4jIu6YrXuVtZnsdQCohIgHgj0AIuBN4DedD813AvcAKr2JLNMaYQWCf13EAGGMOJHqbIpJujAkzjTkmI49Jsiby7+NmjDMPReRjOB9ADwJ/AewClgBVwDeAU5Icp+LGGKN/CfoDfg20AtkxluVFPTbA1a7lQeAzrnU+DvwcOAI0ARcAxcCTQC9O7/i0qPfcAPS4tnt+ZFsFozzPBx4CWoA+YCvw4aj33x9ZP/ovEPkzQDXOh1YLcLtr36WRdTZEni8A7gPacT70fg9Uj+O0MOKgD9gN3Ag0AHeN4e5jEV9HgQMRX7OBu2Lkcn5ULtcBz0b2dVsMVzcAPcCfR23/OWB11L7vAhpcOYwcl8hjdww3jJLHCuBnEVch4KdAsXtfwLVAc2Sdx4fjHcNpBfB0JM+DkWO8IGqbx8U3yjaWA8eAr46yXFzPz4oc7yPAXuC/gPlRyzfhfDj8B9ARaSNfBmZFrZMBfBGnrfUCW4BLopan43yQtEZi2wN8weu6MJ1/OuSSIERkIXAp8C1jzAnjrMaYQ5PY7D8BDwPrgRqcwvt9nIa/Aafh3j/JkIeZC7wCXI7To/o68J2or8yfBP4E/DfO1++lOP9RRjDGDEVi+6Br2x8EthljXhURAX4FLIvsawPwPPCsiCwdI777cXqMFwFXAn+FU4BjEhnT/zbwL8DJkff9NrL4y8AjOMVsOJcXo95+N47bdTiFMRZzcIYjPgycCaQBP4vkFw8/Br4CbI+K4YQhjcj2HgcWAxfifJgXAY+79hXA6R2/D/gzHK//PtrORWQejo8eYGPkfWcBP4is8mXgo5HHw/HF4mqcAvulWAtNpMJG9lkB/A54AqctXwVURu1zmA8CA5F4bgP+JpLbMP8NvBP4AM6H0g+BX4jI+sjyT0TyuRY4KfLe7aPEn5p4/YmSKn84/zkM8L441o23h3531PPyyGufinrtfGL0IF3bda9z3PNR4nsY+F7U8004H1TR6wQi26mOPD818nxN1DpvAndGHl+IU0QyXdupA/5+lDiGe/hnR722EhhklB46TrHoBnJG2eb9wC9HyeXT47i7YYx4Loo8v4sxeuijrRMjj4sj2w1ELV8NDLn2dZRI7zry2meBHWMc24+6/UTluSby/GpG6ZlHvee/gG7Xa5dHjvHw37mR1x8Avu9atzKyz8KoNvYn1zpPEWmHQEkk9xWudR4H7ok8/gbwDK5vBzb9aQ89ccTbQ5sIr0c93h/5tz7Ga4WT3YGIpInIZ0XkdRHpjMziuIoJjvcbY16PxPaByHbPwPlP+L+RVaqAecABEekZ/sP5oCoZZbNlOP+JN0ftZzfON5PReApnaGaXiDwoIh8SkZw406iJY53R4lkX5z7ipQxoNcYEo/a1M8a+dhtjuqOetzJ2eygDXjfGhKJeexEnr4nm4B5ffw6nUF8AZOF8ewHn2P+l67j/MbIs+thHt3c4PpfTcP6PbXNt5z1R27g/sv8mEfm2iLxHRKyqcfqjaOJ4E6eBl+GMe46F4cQPgPQY6/W73jPaa8ONdijO7UbzGeDTOEMr9Tg9q/9gch8SD+KMcf8rztfnP0QK3nCM+4FzY7zv8Cjbm/CHpDEmJCKnAefh9HLvBP5DRE43xoz1QQDOuOxUmcwxiIVwYsEcJvr1/hjLxipi8W53PJqABSKy1BjTBmCM6QV2xJgSOwv4HvD/Ymxnb9TjsXKZFXl+eoz1+iL7fyUyMeFSnG+EPwReE5GLjTMsmPJY9emVTIwxB3F+fLtNRLLdy0UkN+rpAaLGJkVkMaOPVU6EA8A8EZkf9VrlOO85B/iFMeZ/jDF1OD+ulbrWCfN2b2ssHgTWiMg7cMYvfxS17BWc8eAhY8wO11/7KNtrxGmjpw+/ICIrcMaSR8UYM2CMedYYcyfOUFAWznDARHIZjdHiaYy8dABY7Brndh+DeGLYBiyLFKjhfa2O7GvbZAKP2u5617eWs3Dyaoz9lpj8BKew3hnHuq8Ap8Q47juMMX1x7u9VnA+jJTG2MfKhYIwJGWMeNcZ8HKf3fiFvz9pJebSgJ5a/xml0NSJyjYicLCJrReTjHP918lngVhGpFpENOF8VjyZg/y/j9DLvFpE1IvL+SExj0QS8S0TOEZG1wLeAVa51gsDGyMlEBaN9jTXGtOD80HkvzoyWR6MWP43zNfvnInKZiKwSkTNF5F9EJFavHWPMdpwf8L4TWbcSx9WoRUBELheRT4rIBhFZiTMElMPbxSoIlEeOTYGITLT3PAB8LSqeH+LMDHo6snwTsBD4RxEpEZGP4IxJRxMEVorIaZEY5sTYz9M4014fFJGqyI+9D+IUx2cnGHM0D+K0kQdEpEJEzgO+A/zUGLMj3o0YY/bg/Gh5m4j8SEQujLSPDcCnIqsNRv79Ik77uTdyXNZEjtN3JrC/pkjs94vI1SKyOvL/5zMichWAiHxKRK4TkTIRWYNz7A/jzIqxAi3oCcQYswtnrO8pnEb8Os5/vvfiTKUb5tPATpz//D/B+To6Wi91Ivs/iDPUcTHO8MnNwD+P87Z/wxkT/g1OMe7F+Y8TzZdxepXbcHqgY42v/w/OTIZfGWO6omIzwLtxfHwXZ/bBIzgzUcYaCrkBZ37zs8AvcMbkg2Os34UzG+Zp4A2cIaWbjDF/iCz/Lk5xr4nkcvYY24rFMZxZJA/gfIDOAq6K5IcxphFnuunNOMf/YpwhrGgew5ni+kwkhuvcO4ls78rI8k0449P7gCuH9zUZjDFHgEuA+TjH/ec4s5hunMS27sE5xyIX51i+ifMBfArw3mHnkd9XzsP58fn3OB9Ud/P2b0Dx8mGcmS5fwjm2v4xsd3hYLwT8XSSvV3C+GV0WydkKZAptQ1EURZlBaA9dURQlRdCCriiKkiJoQVcURUkRtKAriqKkCJ6dWFRQUGACgcCk3hsOh8nIyEhsQD5DHagDUAc25l9bW9thjFkUa5lnBT0QCFBTE8+Z1icSCoXIyYn3bO7URB2oA1AHNuYvIrtHW+bLIZfa2lqvQ/AcdaAOQB3Ynr8bXxb0zMxMr0PwHHWgDkAd2J6/G18W9MmOvacS6kAdgDqwPX83vizojY0TuYZQaqIO1AGoA9vzd+PLgq6fyuoA1AGoA9vzdzNuQReRH4hI+yh3KUccviEiOyI3STgt8WEeTygUGn+lFEcdqANQB7bn7yaeHvr9OBeMH43LcO7fdxLOFeb+a+phjU1nZ2eydzHjUQfqANSB7fm7GXceujHm+eiL7MfgCuCByCU9XxKR3Oi7mCSDqqqqZG3aN6gDdQDqYCbnf2xgkK4j/RzsDXOoN8zBI2EOHennUG+YDStyOfekmOcGTYlEnFi0jOPvAt8See2Egi4iN+P04ikqKmLTpk2UlZURDAbp6+ujqqqK2tpaCgsLycjIoKWlhfLycpqamhgcHKSiooK6ujo6OjqoqKigra2NyspK6uvrSUtLo7S0lIaGBoqLiwmHw7S3t49sMzMzk0AgQGNjI4FAgFAoRGdn58jy7OxsioqKaGpqoqSkhM7OTrq6ukaW5+bmkp+fT3NzM6WlpbS2ttLT0zOyPD8/n5ycHILB4KRyWrrUuWFRvDk99thjlJWVpVROEz1O3d3drFy5MqVymuhxeuyxx6iqqkqpnCZynF544QUuv/zypOf08pYaTMY85i4ooKEpSNbCQlo7D9Pe3cu8vEKaW/ZxzMzmyOAsOkJHOTIgHOkf/a53f3X6EjIO7pzUcRqLuK6HHumh/9IYUx5j2a9w7k7/QuT5Mzh3cR9zxn91dbWZ7JmiNTU1VFdXT+q9qYI6UAegDiaT/8DgEF19Tk/5YG+YQ0fCHOzt59CRqJ50b5iDkd70od4woWMDo24ve85s8rLSWTgvg9x5GSzMyiBvXgYLs9LJy8o4/vWsdHIzM8iYPfn5KCJSa4yJmXQieugtwPKo58WMfQeaKVNUNOYtJa1AHagDUAeLlyzl4HGF+fiiPDzEMVKke8McPjp6cZ6XkRYpxhnkZWWwKn9eVDF2inNeVvrIOrnz0pkzeyq3qE0siSjoT+DcV/Bh4AygO5nj5wBNTU3WN2R1oA4gtRwMDRm6+yI95eFe83HF+cSedPeRfkYbY5ibPitSgJ3iuzxv3kgRfrsXnXFccZ6bPnOK82QYt6CLyEPA+UCBiLQAnwPSAYwx9+LcG/HdwA7gCM59/5JKSUlJsncx41EH6gBmroOhIUPo6EDkh8DwicMbkccjveoj/XQdCTM0SnXOmB1dnNNZVzSfhVkZpA30sXJJwUjRji7SmRn+Ls6TIZ5ZLifcwNa13AC3JiyiOOjs7GT58uXjr5jCqAN1ANPjwBhD6NjASFEemblxxDXUcSS6UPczOEp1zkibNTJskTcvg7VL5o+MQedFinH08MbCrAwy09MQkRO2VVdXR2XlqqTm7yc8u3zuVOjq6hp/pRRHHagDmJwDYwwHe8Ps7eo7odccPbxxqLd/5PnAKMV59iw5rviuKcyOeh75YXBexnHj0lkZsYvzZNA2cDxxzXJJBlOZ5WLjNZDdqAN1AGM7ODYwyO7OI+w80EPzgV52HuhlZ0cPOw/00t3Xf8L6abOEvHnprh5yBnnuMefIa3lZGeTMmZ2w4jwZbGwDyZ7lMu3U1tZy/vnnex2Gp6gDdQDOtL11VWfSfMAp1NFFu+XQkePGpBfPn8PqgmwuP3Upqxdls2LhPPKz3y7cOXNmM2uWd8V5MmgbOB5fFvTc3FyvQ/AcdWCXg77w4Eihji7aO/b30ffkMyPrZaansaogi1OLF3DlhmWULMpidUE2qxZlkT3Hl//dx8SmNhAPvjzC+fn5XofgOeog9RwMDRlau/siRbuHnR29I49bu4+OrCcCRQsyWb0oi3eXLWT96iWsLshm9aIslsyf67te9lRItTYwVXxZ0Jubm62f3aAO/OsgdLT/uF72zgO9NB/oIdjZy9Go08Vz5sxm9aIszlidz+qCLFYvcor2qoKskfnSmzZt4vwzAx5l4j1+bQPJwpcFvbS01OsQPEcdzHwHHT3HqG/ppnnkR0mn130gdGxknVkCyxfOY3VBFuesKRgp2qsXZbEoe864PzjOdAfJxvb83fiyoLe2tqbM2XGTRR3MLAdDQ4bmAz3U7D5ETfAQtbsPEuw8MrI8d146qwuyeGfpIqdgF2RTsiiLFfnzpnTq+Exy4AW25+/GlwW9p6fH6xA8Rx1466AvPMhrLV3U7j5ETfAgr7zVNTIVcGFWBlUr87h24wo2LM/lpMU5LMzKSEoctrcD2/N348uCPpOvgTxdqIPpdbD/8NFI8XZ631tbD4+cbLOmMJtLT1lCVSCP6pV5rCrImra52ba3A9vzd+PLgq5zT9UBJM/B4JBh+74QtW8dojZ4kJrdh2g51AfAnNmzWL88l4+et5rqlXmctiKPvCT1vuPB9nZge/5ufFnQdaqSOoDEOeg5NsBre7qoCR6iZvdB6t7qGrn+9aKcOVSvzOOGswJUrczjlKIFU7qWdaKxvR3Ynr8bXxZ02071jYU6mLyDvV191O5+u/fd2HaYIePM7z55cQ7vrSyiamUe1SsXsnxhpqento+H7e3A9vzd+LKgB4NBAoGA12F4ijqIz8HA4BCNbSFqdzvFu3b3IdoiJ+nMy0ijcnkut16whqqVeWxYkceCzPRpiDxx2N4ObM/fjS8LellZmdcheI46iO2gLzzI5uDBkd533Z4ujoQHAVi6YC5VK/NGet9lS3OYnTZzhk8mg+3twPb83fiyoAeDQRYvXux1GJ6iDt520B46yrON7Tzd2M4LOw5wtH+IWQJlS+dzdVWxU8ADC1mWm+l1yAnH9nZge/5ufFnQ+/r6vA7Bc2x2YIxh+/4QjzR0c3fNH6nb41wTe1luJn9RvZwL1hZSHViYkhejcmNzOwDN340vW7zOPbXPQXhgiC3Bgzy1bT9PN+4fmUa4vtjw6YtLuWjdYtYuyZnRP2AmA9vagRvb83fjy4Kuc0/tcNB9pJ9NTe08tW0/v99+gNCxAebMnsU5awq49YI1ZB7cwZWXnuN1mJ5iQzsYC9vzd+PLgl5YWOh1CJ6Tqg4O9oZ5om4vv926jy3BQwwOGQqy5/DuiqVctG4x56wpGLn577Ztetp3qraDeLE9fze+LOgZGd6dmTdTSCUHA4ND/L7pAI/WtPDMG/vpHzScvDiHW965movKFrO+ODfmNb5TycFksd2B7fm78WVBb2lpYc2aNV6H4Smp4ODN/SEerW3hp6/spaPnGAXZGXzozADXVC/n5CXjnzCSCg6miu0ObM/fjS8Lenl5udcheI5fHXT39fOL11p5tLaF1/Z0MXuWcMHaQq6pKuaCtYWkT2BeuF8dJBLbHdievxtfFvSmpiYKCgq8DsNT/ORgaMjwx+YOHq1p4cmt+zg2MMTJi3P4p/eUceWGZRRkz5nUdv3kIFnY7sD2/N34sqAPDg56HYLn+MHB0f5BHnulhe/9YRe7OnqZP3c2/6d6OddUF1OxbMGUpxj6wUGysd2B7fm78WVBr6io8DoEz5nJDg72hvmfP+3mgT8F6ewNU7FsAV+/tpJLTlkyci/MRDCTHUwXtjuwPX83vryQRV1dndcheM5MdLC7s5d/fryBs77wDP/v6SbWL8/loY++gyduO5srKpcltJjDzHQw3djuwPb83fiyh7506VKvQ/CcmeTg1bcOcd/zO/nt1n2kz5rFlRuKuOnc1ZQuTu6lTWeSA6+w3YHt+bvxZUFXvMcYw3Pb27l30042Bw8yf+5sPv7OEm44K0Dh/Lleh6coVuLLIZe2tjavQ/AcLx1s3nWQq+/9EzfeX8Perj7+7+XrePHOd/H3l66d1mKu7UAd2J6/G1/20CsrK70OwXO8cLCt9TD/+eQbPLf9AIvnz+Huqyq4uqp4QnPHE4m2A3Vge/5ufNlDr6+v9zoEz5lOB7s7e/nkw6/y7m/8gVfe6uLOy9by+7+7gOs2rvCsmIO2A1AHtufvxpc99LS0xM6W8CPT4aD98FG++ewOHtr8FrPThFsvKOHm80pmzG3atB2oA9vzd+PLgl5aWup1CJ6TTAdHwgPc81wz339hF/2DQ1y3cQW3X7hmxv3Yqe1AHdiev5u4vi+LyKUisl1EdojIHTGWLxCRX4jIayKyVUQ+nPhQ36ahoSGZm/cFyXBgjOHX9W1c9JXf863ndnDxusU88+l38vkry2dcMQdtB6AObM/fzbg9dBFJA74NXAy0AFtE5AljzLao1W4Fthlj/lxEFgHbReRBY0w4GUEXFxcnY7O+ItEOmg/0cNcTW/nDmx2ULZ3PN67bQHVgYUL3kWi0HagD2/N3E8+Qy0ZghzFmJ4CIPAxcAUQXdAPkiHNxjmzgIDCQ4FhHCIeT8jnhKxLl4Gj/IN945k2++4edzE1P41/eewofPGMFsz38sTNetB2oA9vzdxNPQV8G7Il63gKc4VrnW8ATQCuQA/yFMWbIvSERuRm4GaCoqIhNmzZRVlZGMBikr6+PqqoqamtrKSwsJCMjg5aWFsrLy2lqamJwcJCKigrq6uro6OggLS2NtrY2Kisrqa+vJy0tjdLSUhoaGiguLiYcDtPe3j6yzczMTAKBAI2NjQQCAUKhEJ2dnSPLs7OzKSoqoqmpiZKSEjo7O+nq6hpZnpubS35+Ps3NzZSWltLa2kpPT8/I8vz8fHJycggGg5PKafiMt3hz2rx5M6FQaEo5dQzN45tbDrP70DEuW7uQ96+ZxexwkL4j+Z7kNNHj1N3dTTgcntHHKdltb/PmzQwODqZUThM5Tps3b6awsDClchrvOI2FGGPGXkHkGuASY8xNkefXAxuNMbdHrXM1cDbwKaAEeApYb4w5PNp2q6urTU1NzbgBxiIUCpGTk9zTymc6U3HQPzjEN5/dwbef28Gi7Dl86epTOa90UYIjTD7aDtSBjfmLSK0xpjrWsni+V7cAy6OeF+P0xKP5MPBT47AD2AWsnUyw8VBbW5usTfuGyTrYvi/Eld/+I9945k2uqCziyb89z5fFHLQdgDqwPX838Qy5bAFOEpFVwF7gWuADrnXeAt4F/EFEFgMnAzsTGWg0mZmZydq0b5iog8Ehw3f/sJOv/q6J+Zmz+c71VVxyypIkRTc9aDtQB7bn72bcgm6MGRCR24AngTTgB8aYrSJyS2T5vcDngftFpB4Q4B+MMR3JCjoQCCRr075hIg7aQ0f5xEOv8tLOg1x6yhL+/X3l5E/yLkEzCW0H6sD2/N3EdWKRMebXwK9dr90b9bgV+LPEhjY6jY2NLF68eLp2NyOJ18FLOzu5/aFXCR3t58vXrOf9py2b8p2CZgraDtSB7fm78eWZovqpPL6DoSHDd57fyX8++QaB/Cx+9JEzOHlJav14pO1AHdievxtfFvRQKOR1CJ4zloPuI/18+tE6nm5s5z2nLuWL7z+V7Dm+PNRjou1AHdievxtf/i/v7Oz0OgTPGc1BfUs3H3+wlv2Hj3LXn6/jQ2cFUmaIxY22A3Vge/5ufFnQq6qqvA7Bc9wOjDH87+a3+JcntlGQncGPP3Ymp63I8yi66UHbgTqwPX83M//87hjo3NPjHRwJD/DpR17jsz9r4B0l+fzyE+emfDEHbQegDmzP340ve+jZ2dleh+A5ww52tPfw1w/W8mZ7D5+6uJTbLljDrFmpOcTiRtuBOrA9fze+LOhFRUVeh+A5RUVF/OK1Vu547HXmpKfxwI0bOfckf57xOVm0HagD2/N348shl6amJq9D8JTBIcPnf9HA7Q+9yslLcvjVJ86xrpiDtgNQB7bn78aXPfSSkhKvQ/CM0NF+PvlwHc/u6ucv37GCz/35KZ7e19NLbG4Hw9juwPb83fiyEtg6VWl3Zy9X3fMiv286wMeqFvBvV1ZYW8zB3nYQje0ObM/fjS976F1dXV6HMO3U7j7ER364BWPgf27cSLhFb71lYztwY7sD2/N348uCbtvc02ff2M9fP/gKS+bP5f4PbyRQkEVosV0OYmFbO4iF7Q5sz9+NL7+v2zT39LHaFj76QC1rCrP5ycfPIlCQBdjlYDTUgTqwPX83vuyh5+bmeh3CtHDf8838x6/f4Ow1+dz7l1XkzE0fWWaLg7FQB+rA9vzd+LKg5+fnex1CUhkaMnzht29w3/M7ec+pS/nq/1nPnNlpx62T6g7iQR2oA9vzd+PLIZfm5mavQ0gaxhj+9ZfbuO/5nfzVmSv5xrUbTijmkNoO4kUdqAPb83fjyx56aWmp1yEkja89/Sb3vxjkxrNX8c+Xl416pcRUdhAv6kAd2J6/G1/20Ftb3feoTg2+/8Iuvv7Mm1xTVTxmMYfUdTAR1IE6sD1/N74s6D09PV6HkHAeqdnD53+5jcvKl3D3VRXjXsM8FR1MFHWgDmzP340vC3qqzT39bUMbdzz2OueeVMDXrq1kdhxnf6aag8mgDtSB7fm78WVBT6W5p3948wCfeKiODSvy+M71VTF/AI1FKjmYLOpAHdievxtfFvRUmapUu/sQNz9QS0lhNj+44XTmZcT/G3WqOJgK6kAd2J6/G18W9Jwc/9+9vvlADzfev4UlC+bywI0bWZCZPv6bokgFB1NFHagD2/N348uCHgwGvQ5hShwIHeOG/95MeprwwI0bWZQzZ8Lb8LuDRKAO1IHt+bvx5Tz0srIyr0OYNEfCA3zkh1voCIX58cfewfKF8ya1HT87SBTqQB3Ynr8b7aFPIwODQ9z+v6/SsLebb163gVOLJ38dCr86SCTqQB3Ynr8bX/bQ+/r6vA5hwhhjuOsXW3nmjXY+f8UpXLRu8ZS250cHiUYdqAPb83fjyx66H+ee3vf8Tn700lt87J2ruf7MwJS350cHiUYdqAPb83fjy4Lut7mnv9u6j7t/8waXn7qUf7hkbUK26TcHyUAdqAPb83fjy4JeWFjodQhxs31fiL/9cR2nFi/gy9esZ9assU/pjxc/OUgW6kAd2J6/G18W9IyMDK9DiItDvWE++kAN8+bM5r7rq5mbHt9ZoPHgFwfJRB2oA9vzd+PLgt7S0uJ1COMyMDjEbQ+9wr7uo3zn+iqWLJib0O37wUGyUQfqwPb83fhylkt5ebnXIYzLv/2qkT/u6OTL16zntBV5Cd++HxwkG3WgDmzP301cPXQRuVREtovIDhG5Y5R1zheROhHZKiK/T2yYx9PU1JTMzU+ZR7bs4f4Xg3zknFVcXVWclH3MdAfTgTpQB7bn72bcHrqIpAHfBi4GWoAtIvKEMWZb1Dq5wD3ApcaYt0Qkqb9UDA4OJnPzU6ImeJDPPl7PuScVcOdliZnREouZ7GC6UAfqwPb83cTTQ98I7DDG7DTGhIGHgStc63wA+Kkx5i0AY0x7YsM8noqKimRuftK0dfdxy49eYVluJt+67rS4rms+WWaqg+lEHagD2/N3E88Y+jJgT9TzFuAM1zqlQLqIbAJygK8bYx5wb0hEbgZuBigqKmLTpk2UlZURDAbp60IVieYAABdHSURBVOujqqqK2tpaCgsLycjIoKWlhfLycpqamhgcHKSiooK6ujo6OjqoqKigra2NyspK6uvrSUtLo7S0lIaGBoqLiwmHw7S3t49sMzMzk0AgQGNjI4FAgFAoRGdn58jy7OxsioqKaGpqoqSkhM7OTrq6ukaW5+bmkp+fT3NzM6WlpbS2ttLT0/P28ryF/NNznfT0HeObV53CG/WvTCinpUuXAsSd0+OPP05ZWVlSc8rPzycnJ4dgMDip4zTRnCZ6nLq7u1m5cmVK5TTR4/T4449TVVWVUjlN5Di98MILXH755SmV03jHaSzEGDP2CiLXAJcYY26KPL8e2GiMuT1qnW8B1cC7gEzgT8B7jDGjDnBVV1ebmpqacQOMxfbt2zn55JMn9d5k8dWnmvjGM2/y9WsruaJyWdL3NxMdTDfqQB3YmL+I1BpjqmMti6eH3gIsj3peDLjvzNoCdBhjeoFeEXkeWA9Y8YvFSzs7+dazb/L+04qnpZgriqLEIp5B3i3ASSKySkQygGuBJ1zr/Bw4V0Rmi8g8nCGZxsSG+jZtbW3J2vSEOdQb5m8ermNlfhb/esUp07bfmeTAK9SBOrA9fzfj9tCNMQMichvwJJAG/MAYs1VEboksv9cY0ygivwVeB4aA7xljGpIVdGVlZbI2PSGMMfz9Y6/T2XuMn33obLLmTN+0/pniwEvUgTqwPX83cU3DMMb82hhTaowpMcb8e+S1e40x90at85/GmHXGmHJjzNeSFTBAfX19MjcfNz96aTdPbdvPP1y6lvJlC6Z13zPFgZeoA3Vge/5ufHnqf1pa4q6JMlne2HeYz/+qkfNPXsSNZ6+a9v3PBAdeow7Uge35u/FlQS8tLfV0/+GBIf7m4Trmz01P6BUUJ4LXDmYC6kAd2J6/G18W9IaGpA3Px8U3n32TN/aF+OL7KyjInvgNnhOB1w5mAupAHdievxtfFvTi4uRcHyUeGvZ2c8+mZq46bRnvKpvabeSmgpcOZgrqQB3Ynr8bXxb0cDjszX4HhvjMo69RkJ3B5y6fvimKMWPxyMFMQh2oA9vzd+PLgt7entRLxYzK8FDL3VdVsGBeuicxDOOVg5mEOlAHtufvxpcF3Ysbw0YPtVy41ruhlmH05rjqANSB7fm78WVBn+4bww4OGe78aT35Wd4PtQyjN8dVB6AObM/fjS8LemZm5rTu76HNb1G/t5vPvqfM86GWYabbwUxEHagD2/N348uCHggEpm1fB3vD/OeT23nH6oW8d33RtO13PKbTwUxFHagD2/N348uC3tiYtOt+ncCXfvsGvccG+NcryhGZ/hOIRmM6HcxU1IE6sD1/N74s6NP1qVzf0s3DW/Zww1kBShfnTMs+40V7JuoA1IHt+bvxZUEPhUJJ34cxhrt/00jevHQ+cdFJSd/fRJkOBzMddaAObM/fjS8LemdnZ9L38fumA7zY3Mkn3nUS8+fOjB9Co5kOBzMddaAObM/fjS8LerLnng4OGb7wmzdYsXAeHzxjZVL3NVl0/q06AHVge/5ufFnQkz339Gev7uWNfSH+7pKTyZg9MxXp/Ft1AOrA9vzdzMxqNQ7Z2dlJ2/bR/kG+8rvtrC9ewHsqxr/Ltlck04FfUAfqwPb83fiyoBcVJW8++H//MUhb91HuuKzMk+ucx0syHfgFdaAObM/fjS8LelNTU1K2293Xzz2bdnDh2kLOLMlPyj4SRbIc+Al1oA5sz9+NLwt6SUlJUrb7wItBQkcH+NTFM/8uKMly4CfUgTqwPX83vizoyZiq1HtsgB/8cRcXri2c9hs+TwadrqUOQB3Ynr8bXxb0rq6uhG/zoc1vcehIP7desCbh204GyXDgN9SBOrA9fze+LOiJnnt6tH+Q+57fyZmr86lamZfQbScLnX+rDkAd2J6/G18W9ETPPX20toX20DFuv9AfvXPQ+begDkAd2J6/G18W9Nzc3IRtq39wiHs3NbNhRe6Mn9kSTSId+BV1oA5sz9+NLwt6fn7iCu/P61rZ29XHbResmVGXxx2PRDrwK+pAHdievxtfFvTm5uaEbGdwyHDPczsoWzqfC9cWJmSb00WiHPgZdaAObM/fjS8LemlpYuaJ/6ahjZ0dvb7rnUPiHPgZdaAObM/fjS8Lemtr65S3YYzh2881s3pRFpeWL0lAVNNLIhz4HXWgDmzP340vC3pPT8+Ut/HsG+00th3mr89fQ9oMvmbLaCTCgd9RB+rA9vzd+LKgT3XuqTGGbz67g+K8TK6o9OfFfXT+rToAdWB7/m58WdCnOvf0xeZO6vZ0ccs7S0hP86UCnX+LOgB1YHv+bnxZzaY6Ven7L+yiIHsOV1cVJyii6Uena6kDUAe25+/GlwU9Jydn0u/dc/AIz21v57qNy5mbnpbAqKaXqThIFdSBOrA9fzdxFXQRuVREtovIDhG5Y4z1TheRQRG5OnEhnkgwGJz0e3/08m5mifCBM1YkLiAPmIqDVEEdqAPb83czbkEXkTTg28BlwDrgOhFZN8p6XwSeTHSQbsrKyib1vv7BIX5S08K71haydEFmgqOaXibrIJVQB+rA9vzdxNND3wjsMMbsNMaEgYeBK2KsdzvwGNCewPhiMtlP5WffaKezN8xfnL48sQF5gPZM1AGoA9vzdzM7jnWWAXuinrcAZ0SvICLLgPcBFwKnj7YhEbkZuBmcewFu2rSJsrIygsEgfX19VFVVUVtbS2FhIRkZGbS0tFBeXk5TUxODg4NUVFRQV1dHR0cHubm5tLW1UVlZSX19PWlpaZSWltLQ0EBxcTHhcJj29vaRbWZmZvKj1wZYMEdYOecI9fX1dHZ2jizPzs6mqKiIpqYmSkpK6OzspKura2R5bm4u+fn5NDc3U1paSmtrKz09PSPL8/PzycnJIRgMTiqnpUudG1LHm1NjYyMAgUCAxsZGAoEAoVDI1zkNH6d4c+ru7qauri6lcprocWpsbGTevHkpldNEjlNjYyMlJSUpldN4x2ksxBgz9goi1wCXGGNuijy/HthojLk9ap1Hga8YY14SkfuBXxpjfjLWdqurq01NTc24AcYiFApN+MeQA6FjvOPuZ7jp3FXceZn/v6ZNxkGqoQ7UgY35i0itMaY61rJ4hlxagOgximLAfb5tNfCwiASBq4F7ROTKScQaF5OZe/r4q3sZHDJcU+X/4RbQ+begDkAd2J6/m3iGXLYAJ4nIKmAvcC3wgegVjDGrhh9H9dAfT2Ccx1FYOLErIxpjeKRmDxtW5LKmMDtJUU0vE3WQiqgDdWB7/m7G7aEbYwaA23BmrzQCjxhjtorILSJyS7IDjEVGRsaE1n+tpZs323tSpncOE3eQiqgDdWB7/m7imodujPm1MabUGFNijPn3yGv3GmPujbHuDeONn0+VlpaWCa3/aM0e5qbP4vL14/+o4Bcm6iAVUQfqwPb83fjyTNHy8vK41z3aP8gTr7Vy6SlLmD83PYlRTS8TcZCqqAN1YHv+bnxZ0JuamuJe98mt+wgdHeCa6tQZboGJOUhV1IE6sD1/N74s6IODg3Gv+/ire1mWm8mZq1PrIj4TcZCqqAN1YHv+bnxZ0CsqKuJaL3S0nz/u6OSy8iXM8uFNLMYiXgepjDpQB7bn78aXBb2uri6u9Z7bfoDw4BCX+PAWc+MRr4NURh2oA9vzd+PLgh7PKbAAv9u6j4LsDE5bkZfkiKafeB2kMupAHdievxtfFvR4ODYwyKbtB7iobLEv7xmqKIoyUXxZ0Nva2sZd56WdB+k5NsCfnbJ4GiKafuJxkOqoA3Vge/5ufFnQKysrx13nqW37yExP46ySgmmIaPqJx0Gqow7Uge35u/FlQa+vrx9zuTGGp7e1c15pga9vMzcW4zmwAXWgDmzP340vC3pa2thFumHvYfYdPsrF61Jvdssw4zmwAXWgDmzP340vC3ppaemYy5/ato9ZAheuTd0rsY3nwAbUgTqwPX83vizoDQ0NYy7/3bb9VK9cyMKs1L0S23gObEAdqAPb83fjy4JeXFw86rI9B4/wxr4QF69Lzdktw4zlwBbUgTqwPX83vizo4XB41GXPNO4H4KIUL+hjObAFdaAObM/fjS8Lent7+6jLXt51kOULM1lVkDWNEU0/YzmwBXWgDmzP340vC3pVVdWoy17b00Xl8tQ71d/NWA5sQR2oA9vzd+PLgj7ajWHbDx+ltfso64sXTHNE04/eHFcdgDqwPX83vizomZmZMV9/raUbgMrludMZjieM5sAm1IE6sD1/N74s6IFAIObrr7d0kTZLOKUo9XvoozmwCXWgDmzP340vC3pjY2PM1+v2dHHy4hwyM1L/7LHRHNiEOlAHtufvxpcFPdansjGG1/Z0sX556vfOQXsmoA5AHdievxtfFvRQKHTCa8HOIxw+OsD64tQfP4fYDmxDHagD2/N348uC3tnZecJrr7d0AbDegh9EIbYD21AH6sD2/N34sqDHmntat6eLzPQ0TirM9iCi6Ufn36oDUAe25+/GlwU91tzT1/Z0UbFsAbPTfJnShNH5t+oA1IHt+bvxZfXLzj6+F94/OERD62FOteCEomHcDmxEHagD2/N348uCXlRUdNzzHe09hAeGqLCooLsd2Ig6UAe25+/GlwW9qanpuOe7OnoBWGPJ+Dmc6MBG1IE6sD1/N74s6CUlJcc9Hy7ogfzUvsJiNG4HNqIO1IHt+bvxZUF3T1XaeaCXJfPnkjVntkcRTT86XUsdgDqwPX83vizoXV1dxz3f1dGT8tc/d+N2YCPqQB3Ynr8bXxZ099zTXR29rFpkV0HX+bfqANSB7fm78WVBj557eqg3zKEj/ay2rIeu82/VAagD2/N3E1dBF5FLRWS7iOwQkTtiLP+giLwe+XtRRNYnPtS3yc19+/T+nZEfRG0bcol2YCvqQB3Ynr+bcQu6iKQB3wYuA9YB14nIOtdqu4B3GmNOBT4P3JfoQKPJz89/e8eWFvRoB7aiDtSB7fm7iaeHvhHYYYzZaYwJAw8DV0SvYIx50RhzKPL0JaA4sWEeT3Nz88jjXR09zJ4lLF84L5m7nHFEO7AVdaAObM/fTTzz/JYBe6KetwBnjLH+R4DfxFogIjcDN4NzhtemTZsoKysjGAzS19dHVVUVtbW1FBYWkpGRQUtLC+Xl5TQ1NTE4OEhFRQV1dXXMmTOH7du309bWxvbWuRRkCltefonS0lIaGhooLi4mHA7T3t4+ss3MzEwCgQCNjY0EAgFCoRCdnZ0jy7OzsykqKqKpqYmSkhI6Ozvp6uoaWZ6bm0t+fj7Nzc2UlpbS2tpKT0/PyPL8/HxycnIIBoOTymnp0qUAtLW1UVlZSX19PWlpaaPmFAqFePnll1Mqp4kepwULFlBXV5dSOU30OIVCIerr61Mqp4kcp1AoREdHR0rlNN5xGgsxxoy9gsg1wCXGmJsiz68HNhpjbo+x7gXAPcA5xpgxJ4hWV1ebmpqacQOMRU1NDdXV1QBc+rXnWZabyfdvOH1S2/Ir0Q5sRR2oAxvzF5FaY0zMpOMZcmkBlkc9LwZaY+zkVOB7wBXjFfOp0tPTA8DQkCHY2Wvd+Dm87cBm1IE6sD1/N/EU9C3ASSKySkQygGuBJ6JXEJEVwE+B640xSb+4wvDc032Hj3K0f8i6Oeig829BHYA6sD1/N+MWdGPMAHAb8CTQCDxijNkqIreIyC2R1f4vkA/cIyJ1IjK5sZQ4GZ57uvOAnTNcQOffgjoAdWB7/m7iuviJMebXwK9dr90b9fgm4KbEhjY6w1OVdnU4X7dWF9hzlcVhdLqWOgB1YHv+bnx5pmhOTg7gnFSUmZ7G4vlzPI5o+hl2YDPqQB3Ynr8bXxb0YDAIRK7hUpCFiHgbkAcMO7AZdaAObM/fjS8LellZGWDnRbmGGXZgM+pAHdievxtfFvRgMEh4YIg9B49QYuEPoqA9E1AHoA5sz9+NLwt6X18fbx08wpDB2h56X1+f1yF4jjpQB7bn78aXBb2qqoqdB5wZLqssnOECOv8W1AGoA9vzd+PLgl5bW/v2VRYtuo9oNDr/Vh2AOrA9fze+LOiFhYXs6uglPyuDBfPSvQ7HEwoLC70OwXPUgTqwPX83vizoGRkZ7Oyw8xouw2RkZHgdgueoA3Vge/5ufFnQW1paRuag20pLS4vXIXiOOlAHtufvxpcFfVVpGQdCxwhYXNDLy8u9DsFz1IE6sD1/N74s6C+97lzQcYVldymKpqkp6Re1nPGoA3Vge/5ufFnQ9/f0A7AsL9PjSLxjcHDQ6xA8Rx2oA9vzd+PLgp6R59yKqdjigl5RUeF1CJ6jDtSB7fm78WVBr3tzD3Nmz2JRtn1XWRymrq7O6xA8Rx2oA9vzdxPX9dBnGr1ksCwPK6+yOEw8N4xNddSBOrA9fze+7KHv7xlgWa69wy2Koiix8GVBb+s+RnGevTNcANra2rwOwXPUgTqwPX83vivoR8IDhPrt/kEUoLKy0usQPEcdqAPb83fju4K+95BzuUzbC3p9fb3XIXiOOlAHtufvxncFvWWkoNs95JKWluZ1CJ6jDtSB7fm78V1Bz5k7mwtOyrP6LFGA0tJSr0PwHHWgDmzP343vCnp1YCEfKgmzKMfeOegADQ0NXofgOepAHdievxvfFXSA4uJir0PwHHWgDkAd2J6/G18W9HA47HUInqMO1AGoA9vzd+PLgt7e3u51CJ6jDtQBqAPb83fjy4KuN4ZVB6AOQB3Ynr8bXxZ0vTGsOgB1AOrA9vzd+LKgZ2bafVIRqANQB6AObM/fjS8LeiAQ8DoEz1EH6gDUge35u/FlQW9sbPQ6BM9RB+oA1IHt+bsRY4w3OxY5AOye5NsLgI4EhuNH1IE6AHVgY/4rjTGLYi3wrKBPBRGpMcZUex2Hl6gDdQDqwPb83fhyyEVRFEU5ES3oiqIoKYJfC/p9XgcwA1AH6gDUge35H4cvx9AVRVGUE/FrD11RFEVxoQVdURQlRfBdQReRS0Vku4jsEJE7vI5nuhCRoIjUi0idiNREXlsoIk+JyJuRf/O8jjNRiMgPRKRdRBqiXhs1XxG5M9ImtovIJd5EnVhGcXCXiOyNtIM6EXl31LJUdLBcRJ4TkUYR2Soin4y8blVbiBtjjG/+gDSgGVgNZACvAeu8jmuacg8CBa7XvgTcEXl8B/BFr+NMYL7nAacBDePlC6yLtIU5wKpIG0nzOockObgL+EyMdVPVwVLgtMjjHKApkqtVbSHeP7/10DcCO4wxO40xYeBh4AqPY/KSK4AfRh7/ELjSw1gSijHmeeCg6+XR8r0CeNgYc8wYswvYgdNWfM0oDkYjVR20GWNeiTwOAY3AMixrC/Hit4K+DNgT9bwl8poNGOB3IlIrIjdHXltsjGkDp+EDhZ5FNz2Mlq9t7eI2EXk9MiQzPNSQ8g5EJABsAF5G20JM/FbQJcZrtsy7PNsYcxpwGXCriJzndUAzCJvaxX8BJUAl0AZ8JfJ6SjsQkWzgMeBvjDGHx1o1xmsp42E8/FbQW4DlUc+LgVaPYplWjDGtkX/bgZ/hfI3cLyJLASL/pvr9uEbL15p2YYzZb4wZNMYMAd/l7eGElHUgIuk4xfxBY8xPIy9b3xZi4beCvgU4SURWiUgGcC3whMcxJR0RyRKRnOHHwJ8BDTi5fyiy2oeAn3sT4bQxWr5PANeKyBwRWQWcBGz2IL6kM1zEIrwPpx1AijoQEQG+DzQaY74atcj6thCL2V4HMBGMMQMichvwJM6Mlx8YY7Z6HNZ0sBj4mdO2mQ38rzHmtyKyBXhERD4CvAVc42GMCUVEHgLOBwpEpAX4HPAFYuRrjNkqIo8A24AB4FZjzKAngSeQURycLyKVOMMIQeBjkLoOgLOB64F6EamLvPaPWNYW4kVP/VcURUkR/DbkoiiKooyCFnRFUZQUQQu6oihKiqAFXVEUJUXQgq4oipIiaEFXFEVJEbSgK4qipAj/H/nXWmJd00hHAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Truncating_Mutations    66\n",
      "Amplification           47\n",
      "Deletion                42\n",
      "Fusions                 23\n",
      "Overexpression           4\n",
      "T58I                     3\n",
      "G12V                     2\n",
      "P34R                     2\n",
      "R170W                    2\n",
      "Q61H                     2\n",
      "Name: Variation, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Check Number of Unique Variations\n",
    "unique_Variation = X_train['Variation'].value_counts()\n",
    "print('Number of Unique Variatis :', unique_Variation.shape[0])\n",
    "h = unique_genes.values/s;\n",
    "c = np.cumsum(h)\n",
    "plt.plot(c)\n",
    "plt.title(\"Cumulative distribution of Genes\", fontsize = 14)\n",
    "plt.grid(color='gray', linestyle='--', linewidth=0.5)\n",
    "plt.show()\n",
    "\n",
    "# The figure shows that the first 50 genes have almost 75% of occurrence. Therefor gene is important for us and we need to transform it.ons:', unique_Variation.shape[0])\n",
    "# the top 10 genes that occured most\n",
    "print(unique_Variation.head(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEJCAYAAACT/UyFAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAgAElEQVR4nO2dfXwU1b24n0MgEAgQDAQJAaKRLYFEwACW6xuKWLG+oJZaq/Wl9qqt/qrV3l6tpWK1V22lt74V6/tra22vBVTwBTGKohSC0SSELASDhMQEAgkJhIQk5/fHLHSJ2c1mZnbnu3Cez2c/ycycPfOc78x+9+zZszNKa43BYDAYjhx6eS1gMBgMhthiEr/BYDAcYZjEbzAYDEcYJvEbDAbDEYZJ/AaDwXCEYRK/wWAwHGGYxH8YoZSaoZTSSqmhDuvJDNQzxS23Huy7WCk1P2i5Qin18yjs55BYKaWuUko1ub2foP1FpR12UErNV0rVBNp/VYz37dq5FajnO254HWmYxN8DlFLDlVIPKqXKlVItSqltSqllSqlzvHazi1IqXyn1SKfVW4ERQKEHSp2ZCvwpkoJKqWeVUq9HWO8qrDbW2RUL4TBfKVXcxaaI2xFNlFI5wJ3A9Vjt/1un7cOVUvuVUpeHeP7vlFJblVJ2c0ePz60wx3UE8JpNjyOa3l4LxAtKqUzgI6ARuB34DOuNcybwGDDaKze30Vq3A1957QGgtd7udp1KqT5a61Zi2MZotMMmxwX+LtJd/HpTa10TSLLXAC8Gb1NK9QZ+ADyhte7o6Y6VUoluxl1rLeIcjUu01uYRwQNYClQByV1sGxL0vwa+02l7BfDzTmV+DCwG9gJ+4HQgA3gL2IPVIzoh6DlXAU2d6p0RqGtoiOVU4K9AJdAMlABXBz3/2UD54Edm4KGBKVhvbpXA/+u0b1+gzOTA8mDgcaAW683xfWBKNzFNC8SgGdgC/BAoBuaHid11gXjtA7YH4tUbmN9FW2YEteVSYEVgXzd2EaurgCbgvKD63wOODdr3fKC4UxsOHpfA/50drgrRjtHAPwOxagReBTI67wv4HlAeKLPogG+YmOYCywPt3Bk4xoOD6jzEL0Qd3wY6gtseWH9BYH0m1ieYt4EdwG7gQ2B6p/IauCHQtj3AA0HHY0qgTALwFPBFwHkj8AugVyhnYEZXr7VwbQ86318HbgK2AbuAZ4D+XueXWD/MUE8EKKWOAs4GHtFaf20cWGu9y0a1vwJeBiYCa7ES9FNYwwGTsd5knrWpfIB+wDrgXGAC8CDwZ6XUzMD2m4CPsU7+EYHH1uAKtNWz+ytwWae6LwPWa60/VUop4A1gZGBfk4EPgBVKqRFh/J7F6oGeCcwBrsBKDF0SGBd+FLgL+EbgeW8GNj8AvIL1wj/QllVBT78XK7bjsRJoV/TFGga5GpiOlZT+GWhfJPwNWACUBTn8rXOhQH2LgOHAGVhv+unAok77ygQuAS4EzsKK629D7Vwp1R8rHk3AtMDz/gN4OlDkAeA/A/8f8OuKN7HOv6s7rb8GeFdrXQEMBF4ATgnsqxBY2sX3S3didZpysY5dZ3phJeHvAtnAHcAvg/bd3XGNtO0HOAXIwTp3DsT2pq6CcFjj9TtPPDywTiQNXBhB2Uh7/PcGLecE1t0StG4GXfRIO9XbucwhyyH8XgaeDFrOx3pDCy6TyaG9suMDy8cFldkI3B74/wysF1xSp3oKgV+E8DjwieGkoHVjgHZC9PiBi4AGYGCIOp8FXg/Rllu7id1VYXzODCzPJ0yPP1SZLtoxK1BvZtD2Y7F608H72sehPdY7gE1hju1/do5PUDuPCyx/hxA9/U513YPVCTjQ8z4a2A9cEqK8AqqByzud5w+HO7dC1HUfsDzcce38Wouw7c8G2tQ7qMwTwfs6Uh6mxx8Zkfb4esLnQf/XBP4WdbEuze4OlFIJSqk7lFKfK6XqArNWLqKH30dorT8PuH0/UO+JQBbwl0CRPKA/sF0p1XTggfWGlhWi2mysRPevoP1swepphuIdrCGhL5RSLymlrlRKDYywGWsjKBPKZ3yE+4iUbKBKWz3nA/va3MW+tmitG4KWqwh/PmQDn2utG4PWrcJqV0/b8DTWJ7izAstXYiXWRQBKqTSl1J+VUn6lVAPWUFQaXz+3uo27Uup6pdRapdT2wHnzsy7q6Y5I275ea90WtNxdTA9LTOKPjI1YPYfsCMpqvv5G0aeLcvs7PSfUugPHqCPCeoP5OXAr8HusL6EnYb1wE7t5Xle8xL+Hey4DVgYS4wHHmkD9wY9xwLwQ9fX4zTTwoj4Ba1jgS6wv2TcopdIjePqenu6vC+wcg65Q/Pv4diZ4/f4utoV7zUZab7cE3ojew/rehcDfF7XWLYHl57DG+X+GNaQyCeu7oM7nVti4K6UuAf6I1Rv/VqCeP3VRT3dEK6aHJUdcg+2gtd6J9SXijUqp5M7blVIpQYvbCRo7VUoNJ/RYak/YDvRXSg0KWjepm+ecDLymtX5Ba12I9SWhr1OZVqyx7O54CThOKfVNrLHR4Bkf67DGqzu01ps6PWpD1FeKdf5NPbBCKTUaa6w7JFrrNq31Cq317VhDUAOwvlfoSVtCEcqnNLBqOzC80zh852MQicN6YGRgptiBfR0b2Nd6O+JB9U7s9CnoP7DaVdr1U8LyJHCBUupCrPPmyaBtJ2MN47yhtS7B6vHbOc9PBlZrrR/RWq/TWm/i658SI42pm20/rDGJP3J+gtWrWKuUmquU+oZSapxS6sccOmyzArhBKTVFKTUZqyezz4X9r8bqPd2rlDpOKXVxwCkcfmCmUupkpdQ44BHgmE5lKoBpgR/WDA01P1trXYn1he1jWDN4/h60eTnWVNfFSqnZSqljlFLTlVJ3KaVOCVFfGdaXcX8OlJ2EFavmUI1RSp2rlLpJKTVZKTUGa+hpIP9+YVcAOYFjM1Qp1dPeeBvwxyCf57BmQi0PbM8HjgJ+qZTKUkpdgzVmHkwFMEYpdULAoW8X+1mONR34JaVUXuBL65ew3kBX9NA5mJewzpHnlVK5SqlTgT8DrwYSak95Feu7m6eAf2mtg3+f4AcuV0qNV0pNxfruqNXGPvzACYHzZqxSah5wWqcyFXR/XN1u+2GNSfwRorX+AmuY4R3gfqxkvwI4H2uK4QFuBTZjJYl/YPWSQvV6e7L/nVhDLLOwxtuvJfQwygHuwRqzXoaVtPdgvUCCeQDrBbseq0cbbmz1BaxZSG9oreuD3DRwDlY8nsCa1fIK1sybcGP2V2FN41uB9UOcv2C9yENRjzX7ZzmwAWso60da65WB7U9gvQmsDbTlpDB1dUUL1qyZ57HeaHsBF+kD3yRqXYo1DfdarOM/C/ifTnX8H9YslncDDpd23kmgvjmB7flYQypfAXMO7MsOWuu9WMMlg7CO+2KsWVs/DPe8MPW1YJ0vQzi0t0+gzmSgACvpP034YxeKP2OdK38B1mB9+bugU5luj6vbbT/cUQ7OM4PBYDDEIabHbzAYDEcYJvEbDAbDEYZJ/AaDwXCEYRK/wWAwHGF4dnXOoUOH6szMTFvPbW1tJTHRzm+QYoPxs49kN5DtJ9kNjJ8Tgt0KCgp2aK2HOanPs8SfmZnJ2rWR/Ir+6zQ2NjJwYKS/1I89xs8+kt1Atp9kNzB+Tgh2U0pt6aZ4t8TlUE9BQYHXCmExfvaR7Aay/SS7gfFzgttucZn4k5KSvFYIi/Gzj2Q3kO0n2Q2MnxPcdovLxG/3u4FYYfzsI9kNZPtJdgPj5wS33eIy8ZeWyr7mkvGzj2Q3kO0n2Q2MnxPcdovLxC/5nRmMnxMku4FsP8luYPycEPMev1LqaaVUrVKqOMR2pZR6SCm1KXDDjxNcNeyCxsbG7gt5iPGzj2Q3kO0n2Q2MnxPcdoukx/8s1v1mQzEbGBt4XAssdK4Vnrq6umjvwhHGzz6S3UC2n2Q3MH5OcNut28Svtf4A6471obgAeF5bfAKkdHODbcfk5eVFs3rHGD/7SHYD2X6S3cD4AbS2dbAwv5xPv9zVo+e57ebGD7hGYt3A+ACVgXXVnQsqpa7F+lRAeno6+fn5ZGdnU1FRQXNzM3l5eRQUFJCWlkZiYiKVlZXk5OTg9/tpb28nNzeXwsJCduzYQW5uLtXV1UyaNImioiISEhLw+XwUFxeTkZFBa2srtbW1B+tMSkoiMzOT0tJSMjMzaWxspK6u7uD25ORk0tPT8fv9ZGVlUVdXR319/cHtKSkppKamUl5ejs/no6qqiqampoPbU1NTGThwIBUVFTQ1NTFs2LAetWnECOu9MhZtamlpoW/fvj1qk53jZKdN27Zt48wzz4zJcTrczr38/Hxmz54dk+Nkp02FhYVcfPHFMTlOEs+9bfsH8Pi6BrbsauGKqUfTWt0WcZsWLVrE9OnTXUjXASK5IzvWzRGKQ2x7Azg5aPldIK+7OvPy8rRd1qxZY/u5scD42Ueym9ay/SS7aX3k+tU0NOuf/nWdHvPfr+uT739Xv1v6VY/rCHYD1uoI8na4hxs9/kpgVNByBuHvuuSY9PRI7q3tHcbPPpLdQLafZDc48vza2jt4/uMt/O87flraOvjpzLH8ZEYW/fr0/LbQbru5MZ1zCXBFYHbPN4EGrfXXhnncxO/3R7N6xxg/+0h2A9l+kt3gyPIr2LKL8x75iN+8vp7JY4bw1s9O5ZZZPltJ3203iGCMXyn1V2AGMFQpVQncCfQB0Fo/hnV/0XOATcBe4GpXDbsgKysr2rtwhPGzj2Q3kO0n2Q2ODL+de1q5b1kpr6ytZMTgfiy87ATOzjkapZTnbsF0m/i11l+7WXSn7Rq4wTWjCKirq2PUqFHdF/QI42cfyW4g20+yGxzefh0dmpfXbOV3b22gaV8b1516LD+dOZYBfd25ALLbsfPsssxOqK+v91ohLMbPPpLdQLafZDc4fP2KKhv41eJiPttaz4nHHMXdc3LwDXf38s5ux05ZHfbYM2XKFG2ux+8Nkv0ku4FsP8lucPj5NTTvZ8HbZbzwyRZSB/TlV9/O5oJJ6Y6HdbpzU0oVaK2nOKkvLq/VI/m62WD8nCDZDWT7SXaDw8dPa83/FVQyc0E+L36yhSunZ7Li56cxZ/LIqCT9nrhFSlwO9aSkpHitEBbjZx/JbiDbT7IbHB5+ZV81Mm9RMf+q2Mnk0Sk8e/U0ckYOFuHWE+Iy8aempnqtEBbjZx/JbiDbT7IbxLdfU0sbDy738/RHFQzq15v7L85lbt4oevWKTg+/J252iMuhnvLycq8VwmL87CPZDWT7SXaD+PTTWvP651XMXJDPEyu/4LtTMlhx6wwumTo6Zkk/lJsT4rLH7/P5vFYIi/Gzj2Q3kO0n2Q3iz2/z9ibuXFLCyo07mJA+iIWX53HC6CEi3JwSlz3+qqqoXhHCMcbPPpLdQLafZDeIH7/m1nYeeKuMs/+4ksKt9dx1/gSW3HiyZ0k/2M0t4rLH39TU5LVCWIyffSS7gWw/yW4QH37L19cw/7USKnc1c9Hkkdx+TjbDBvb1Ws312Jl5/FHA+NlHshvI9pPsBrL9tu7cy7x/fkb+xp2MTUvm7jk5fPNYOV9Gm3n8HD7zgb1Csp9kN5DtJ9kNZPq1tLXzyIqNnPmH9/l4805+ec44lt50iqikD2YePxDf08IkINlPshvI9pPsBvL8Vm7czp2LS9i8Yw/n5B7N3ON6cfqJMi8k53bs4jLxS/24eADjZx/JbiDbT7IbyPH7qmEfd7+xnjc+ryYztT/P/XAap/mGUVFR4bVaSNyOXVwO9Ug+QGD8nCDZDWT7SXYD7/32t3fwxAebmbkgn+Xra7hllo83bz6V03zDRPiFw223uOzxZ2dne60QFuNnH8luINtPsht46/evL3Yyb1ExZTWNnDEujfnnTWB0av9DykiOn9tupscfBYyffSS7gWw/yW7gjd/2xhZueaWQ7/75Y5pa2nj8B3k8deWUryV9r/wixfT4gebmZq8VwmL87CPZDWT7SXaD2Pq1d2j+snoLv3urjH3727nh9CxuPH0sSYmhb30oOX5uu5l5/FHA+NlHshvI9pPsBrHzK9xaz7xFxRRta+Dk44Zy1wUTyBqWLMbPDmYePzLnAwdj/Owj2Q1k+0l2g+j77drTyu2vFnHhnz6itnEfD186mReumRZR0o+FnxPMPH4gLS3Na4WwGD/7SHYD2X6S3SB6fh0dmr8XbOW+ZRvYva+Na046hptn+Uju4f1uJcfPbbe4TPyJiYleK4TF+NlHshvI9pPsBtHxK6lqYN6iYtZ9Wc/UzCHcPSeHcUcPslWX5Pi57RaXQz2VlZVeK4TF+NlHshvI9pPsBu767d63n/lLSjjv4Q/ZUreXB+ZO5JXrpttO+m77uY3bbnHZ48/JyfFaISzGzz6S3UC2n2Q3cMdPa82Sz6q4541SdjS1cNmJo/mvs8YxuH8fEX7Rwm23uOzx+/1+rxXCYvzsI9kNZPtJdgPnfptqG/n+E6u56eVC0gf3Y/ENJ3HPnFxXkr4bftHEbbe47PG3t7d7rRAW42cfyW4g20+yG9j329vaxkPvbuLJlZsZ0Lc398zJ4dJpo0lw+daHkuPntltcJv7c3FyvFcJi/Owj2Q1k+0l2g577aa15q+QrfvPaeqoa9jE3L4PbZo8jNTk6N0aRHD+33eJyqKewsNBrhbAYP/tIdgPZfpLdoGd+W+r2cPWza7j+xXUMSurDP66fzu/nToxa0u+pX6xx2y0ue/wjRozwWiEsxs8+kt1Atp9kN4jMb9/+dhbml7Pw/XISE3ox79zxXDl9DL0Tot9HlRw/t93iMvEbDIbDj/fKapm/pIQtdXs5b2I6v/p2NsMH9fNa67AkLod6qqurvVYIi/Gzj2Q3kO0n2Q1C+22rb+a6F9Zy9TNrSOileOlHJ/LwpZNjnvQlx89tt7i8SFt9fT0pKSkuG7mH8bOPZDeQ7SfZDb7u19rWwVMffsFD725Eo/l/Z4zlP085lsTe3vRHJccv2O2IvUhbUVGR1wphMX72kewGsv0ku8GhfqvKd3DOQyu5/80NnDJ2KMtvOY0bTj/Os6Tf2U8abrtFNMavlDobeBBIAJ7UWt/Xaftg4EVgdKDOB7TWz7hqGkRCQuhrakvA+NlHshvI9pPsBpZf7e59/HZpKYsLqxh9VH+evmoKZ4wb7rUaIDt+brt1m/iVUgnAo8AsoBJYo5RaorVeH1TsBmC91vo8pdQwoEwp9ZLWutVV2wA+ny8a1bqG8bOPZDeQ7SfZra29g4LGQVy34H1a2ju4aeZYfjwji3595CRbyfFz2y2Sz1XTgE1a682BRP4ycEGnMhoYqJRSQDKwE2hz1TSI4uLiaFXtCsbPPpLdQLafVLeCLTs575GPWPDeFk4YM4S3bz6Vn83yiUr6IDd+4L5bJEM9I4GtQcuVwImdyjwCLAGqgIHAJVrrjs4VKaWuBa4FSE9PJz8/n+zsbCoqKmhubiYvL4+CggLS0tJITEyksrKSnJwc/H4/7e3t5ObmHvwhQ1lZGdXV1UyaNImioiISEhLw+XwUFxeTkZFBa2srtbW1B+tMSkoiMzOT0tJSMjMzaWxspK6u7uD25ORk0tPT8fv9ZGVlUVdXR319/cHtKSkppKamUl5ejs/no6qqiqampoPbU1NTGThwIBUVFQwYMIDVq1f3qE0H5unGok1HHXUU+fn5PWqTneNkp0179uyhpqYmJsfpcDv3du7cSU1NTUyOUyRtSkoZxn3LNrByWxupSQlcddx+bvnOONatW0O9Ofd61KadO3dSVlYWQbqOEK112AcwF2tc/8DyD4CHO5X5DvC/gAKOA74ABoWrNy8vT9ulpKTE9nNjgfGzj2Q3rWX7SXFrb+/QL35SoY+f/5bOuv0N/T9L1+umffvF+IVCsl+wG7BWd5O3u3tEMtRTCYwKWs7A6tkHczXwasBrUyDxj7P7ZtQdtbW10araFYyffSS7gWw/CW5FlQ1cuHAVd/yzmHFHD2TZTadw++xsBvTtLcIvHJL93Hbrdh6/Uqo34AdmAtuANcD3tdYlQWUWAjVa6/lKqeHAOmCi1npHqHrNzda9Q7KfZDeQ7eelW8Pe/Tzwdhkvrt5C6oC+/Orb2VwwKR3raz/v/SJBsl/Mb7autW4DbgTeAkqBV7TWJUqp65VS1weK3Q38h1KqCHgX+O9wSd8pkm+KDMbPCZLdQLafF25aa/6voJIzFuTz0uotXDk9kxU/P405k0cekvS98usJkv08udm61nopsLTTuseC/q8CznLVLAxJSUmx2pUtjJ99JLuBbL9Yu5V91ci8RcX8q2Ink0en8NwPp5EzcnDI8pJjB7L93HaLy4u0ZWZmeq0QFuNnH8luINsvVm5NLW08uNzP0x9VMKhfb+6/OJe5eaPo1c2NUSTHDmT7ue0Wl5dsKC0t9VohLMbPPpLdQLZftN201rz+eRUzF+TzxMov+O6UDFbcOoNLpo7uNunHws8pkv3cdjM9/ihg/Owj2Q1k+0XTbfP2Ju5cUsLKjTuYkD6IhZfnccLoIT2qQ3LsQLaf225xmfgbGxu9VgiL8bOPZDeQ7RcNt+bWdh59bxOPf7CZvn16cdf5E7j8m2Ns3e9WcuxAtp/bbnGZ+Ovq6rxWCIvxs49kN5Dt57bb8vU1zH+thMpdzVw0eSS3n5PNsIH2b30oOXYg289tt7i8Hr/k+bZg/Jwg2Q1k+7nltnXnXu56rYTlpbWMTUvm7jk5fPPYVDF+0UKyX8zn8UtE8nxbMH5OkOwGsv2curW0tfPIio2c+Yf3WVVexy/PGcfSm05xJem74RdtJPt5Mo9fGsnJyV4rhMX42UeyG8j2c+K2cuN27lxcwuYdezgn92jmnTueEYPdnTsuOXYg289tt7hM/Onp6V4rhMX42UeyG8j2s+NW3dDMPa+X8kZRNccMHcBzP5zGab5hUbCTHTuQ7ee2W1wO9fj9fq8VwmL87CPZDWT79cRtf3sHj39QzswF77O8tIZbZ/l48+ZTopb0e+rnBZL93HaLyx5/VlaW1wphMX72kewGsv0idVu9uY55i4vx1zQxc1wa88+fwKij+kfZTnbsQLaf225x2eOXPO0KjJ8TJLuBbL/u3LY3tnDL3wq55PFP2NPSzhNXTOGpq6bGJOlH4uc1kv3cdovLHn99fb3XCmExfvaR7Aay/UK5tXdoXlq9hd+/Vca+/e3ccHoWN54+lqTE2N76UHLsQLaf225mHn8UMH72kewGsv26cvv0y13MW1xM8bbdnHzcUO66YAJZw7yZvSI5diDbz8zjR/Z8WzB+TpDsBrL9gt127Wnl9leLuGjhKrY3tvDwpZN54ZppniX9zn4Skexn5vEDKSkpXiuExfjZR7IbyPZLSUmho0Pz94Kt3LdsA7v3tXHNScdw8ywfyX29f6lLjh3I9nPbzfuzwQapqe78kjBaGD/7SHYD2X47O/rzncdWse7LeqZmDuHuOTmMO3qQ11oHkRw7kO3ntltcDvWUl5d7rRAW42cfyW4g02/3vv3MX1LCNa9sZEvdXh6YO5FXrpsuKumDzNgFI9nPbbe47PH7fD6vFcJi/Owj2Q1k+WmtWfJZFfe8UcqOphbm5A5l/oUnMLh/H6/VukRS7LpCsp/bbnHZ46+qqvJaISzGzz6S3UCO36baRr7/xGpuermQ9MH9WHzDSVz2jQSxSR/kxC4Ukv3cdovLHn9TU5PXCmExfvaR7Abe++1tbeOhdzfx5MrNDOjbm3vm5HDptNEk9FLkbzKxc4JkP7fdzDz+KGD87CPZDbzz01rzVslX/Oa19VQ17GNuXga3zR5HavK/b4xiYucMyX5mHj+y59uC8XOCZDfwxm9L3R6ufnYN17+4jkFJffjH9dP5/dyJhyR9r9x6gvGzj5nHj+xpV2D8nCDZDWLrt29/Owvzy1n4fjmJCb2Yd+54rpw+ht4JXffXTOycIdnPbbe4TPxSP44dwPjZR7IbxM7vvQ213LmkhC937uX8ienc8e1shg/qJ8LNLsbPPm67xeVQT0VFhdcKYTF+9pHsBtH321bfzHUvrOXqZ9fQJ0Hxlx+dyEOXTu426cfCzSnGzz5uu8Vljz87O9trhbAYP/tIdoPo+bW2dfDkh5t5+N1NAPzi7G/wo5OPJbF35H2zIzV2biHZz2030+OPAsbPPpLdIDp+q8p3MPvBD/jdm2WcMnYo79xyKj+ZcVyPkn603NzE+NnH9PiB5uZmrxXCYvzsI9kN3PWr3b2Pe94oZclnVYw+qj/PXDWV08eliXCLBsbPPm67mXn8UcD42UeyG7jj19bewfMfb+EP7/hpbe/gx6dl8eMZWfTr4+zGKEdC7KKJZD8zjx/Z823B+DlBshs49yvYspPzHvmI37y+nrwxQ3j75lP52Syf46Tvhlu0MX72MfP4gbQ0+x+HY4Hxs49kN7DvV9fUwv1vbuCVtZWMGNyPxy4/gW9NOBqllOduscL42cdtt4gSv1LqbOBBIAF4Umt9XxdlZgB/BPoAO7TWp7noeQiJiYnRqtoVjJ99JLtBz/06OjR/XfMlv3uzjD0tbVx32rH89IyxDIjCjVEOt9jFGsl+brt1O9SjlEoAHgVmA+OBS5VS4zuVSQH+BJyvtZ4AzHXVshOVlZXRrN4xxs8+kt2gZ35FlQ1cuHAVd/yzmHFHD2TZTadw++zsqCT9nrp5gfGzj9tukZyB04BNWuvNAEqpl4ELgPVBZb4PvKq1/hJAa13rqmUncnJyolm9Y4yffSS7QWR+DXv388DbZby4egupA/ryx0smccGkdFeHdey6eYnxs4/bbpEk/pHA1qDlSuDETmV8QB+lVD4wEHhQa/1854qUUtcC1wKkp6eTn59PdnY2FRUVNDc3k5eXR0FBAWlpaSQmJlJZWUlOTg5+v5/29nZyc3MpLCykoaGBcePGUV1dzaRJkygqKiIhIQGfz0dxcTEZGRm0trZSW1t7sM6kpCQyMzMpLS0lMzOTxsZG6urqDm5PTk4mPT0dv99PVlYWdXV11NfXH9yekpJCamoq5eXl+Hw+qqqqaGpqOrg9NTWVgQMHUlFRQUtLCykpKT1q04gRIwBi0qaOjg569erVozbZOcoTWqcAACAASURBVE522vTVV19x+umnx+Q4uX3uFRUVUbpvEH9eXcvufR18f2o6Jw2qJ7W1ktra3lFv08qVKznrrLNicpzsnHtFRUXMmTMnJsfpcDv3li1bxrRp0yLN692jtQ77wBq2eTJo+QfAw53KPAJ8AgwAhgIbAV+4evPy8rRdPvjgA9vPjQXGzz6S3bQO7behereeu3CVHvPfr+s5j36oiyrrY2wWv7GTgmS/YDdgre4mb3f3iKTHXwmMClrOADrfDqYS6wvdPcAepdQHwETAb/P9KCy5ubnRqNY1jJ99JLvB1/2aWtp4cLmfpz+qYFC/3tx/cS5z80bRq1d0h3UicZOG8bOP226RzONfA4xVSh2jlEoEvgcs6VRmMXCKUqq3Uqo/1lBQqaumQRQWFkaralcwfvaR7Ab/9tNa8/rnVcxckM8TK7/gu1MyWHHrDC6ZOtqTpB/sJhXjZx+33brt8Wut25RSNwJvYU3nfFprXaKUuj6w/TGtdalS6k3gc6ADa2io2FXTIA6M30nF+NlHshtYfuXbm5i/pISVG3cwIX0QCy/P44TRQ7xWi4vYSUayn9tuEc0r01ovBZZ2WvdYp+XfA793T81gkEVzazvPFtTxj5Jy+vVJ4K7zJ3D5N8eQ4FEP32CwS1xesqG6utprhbAYP/tIdXtnfQ1n/uF9/vr5Ls47Pp0Vt87gyv/IFJX0pcbuAMbPPm67xeUlGyZNmuS1QliMn32kuW3duZf5S0p4d0MtvuHJPPX9HGYeP8ZrrS6RFrvOGD/7uO0Wlz3+oqIirxXCYvzsI8Wtpa2dh9/dyJl/eJ+PN9fxy3PG8cZPTyGx4Uuv1UIiJXahMH72cdstLnv8CQnOr2QYTYyffSS4feDfzp1LSvhixx6+nTuCX52bzYjBSYAMv1BIdgPj5wS33eIy8ft8Pq8VwmL87OOlW3VDM/e8XsobRdUcM3QAz/9wGqf6hh1SxsTOPsbPPm67xeVQT3Fx1GaKuoLxs48XbvvbO3j8g3JmLnif5aU13DrLx5s3n/K1pO+VX6RIdgPj5wS33eKyx5+RkeG1QliMn31i7bZ6cx3zFhfjr2li5rg05p8/gVFH9Q9Z3sTOPsbPPm67xWXib21t9VohLMbPPrFy297Ywr1LS3n1022MTEniiSumMGv88G6fZ2JnH+NnH7fd4nKop7Y2qld9dozxs0+03do7NM9/XMEZC/J57fMqbjg9i+W3nBZR0o+FnxMku4Hxc4LbbuZm61HA+Nknmm6ffrmLeYuLKd62m5OPG8pdF0wga1iyGD+nSHYD4+cEc7N1ZN8UGYyfE6LhtmtPK7e/WsRFC1exvbGFhy+dzAvXTOtx0o+Wn1tIdgPj5wRzs3UgKSnJa4WwGD/7uOnW0aH5e8FW7lu2gd372rjmpGO4eZaPZAe3PjxSYhcNjJ993HaLy8SfmZnptUJYjJ993HIrqWpg3qJi1n1Zz9TMIdw9J4dxRw9yXO+RELtoYfzs47ZbXA71lJZG7VL/rmD87OPUbfe+/cxfUsJ5D3/Ilrq9PDB3Iq9cN92VpO+GXzSR7AbGzwluu5kefxQwfvax66a1ZnFhFb9dWsqOphYuO3E0/3XWOAb37yPCLxZIdgPj5wS33eIy8Tc2NnqtEBbjZx87bhtrGpm3uJhPNu9kYsZgnrpyCsdnpETB7vCLXSwxfvZx2y0uE39dXZ3XCmExfvbpidueljYeWrGRp1Z+wYC+vfnthTl8b+roqF4j/3CJnRcYP/u47Wbm8UcB42efSNy01rxV8hW/eW09VQ37mJuXwW2zx5Ga3FeEn1dIdgPj5wQzjx/Z823B+DmhO7eKHXu46pk1XP/iOgYl9eEf10/n93MnxiTpR+LnJZLdwPg5wczjB5KTe/7Dm1hi/OwTym3f/nYW5pez8P1yEhN68etzx3PF9DH0Toht3yUeYycF42cft93iMvGnp6d7rRAW42efrtze21DLnUtK+HLnXs6fmM4d385m+KB+HtjFX+wkYfzs47ZbXA71+P1+rxXCYvzsE+y2rb6Z615Yy9XPrqFPguIvPzqRhy6d7FnS7+wnDcluYPyc4LZbXPb4s7KyvFYIi/GzT1ZWFq1tHTz54WYefncTAL84+xv86ORjSeztfT9FeuwkY/zs47ab968kG0iedgXGzwn566uY/eAH/O7NMk4ZO5R3bjmVn8w4TkTSB9mxk+wGxs8JbrvFZY+/vr7ea4WwGL+eU7t7H/e8UcqSz3Yw+qj+PHPVVE4fl+a11teQGLsDSHYD4+cEt93MPP4oYPwip629g+c/3sIf3vHT2t7BNdMzuOms8fTrk+C1WpdIil1nJLuB8XOCmceP7Pm2YPwipWDLTs575CN+8/p68sYM4e2bT+XEAXVikz7IiV1XSHYD4+cEM48fSEmJznVY3ML4haeuqYX739zAK2srGTG4H49dfgLfmnA0SinqTexsI9kNjJ8T3HaLy8SfmprqtUJYjF/XdHRo/rrmS373Zhl7Wtq47rRj+ekZYxkQdGMUEzv7SHYD4+cEt93icqinvLzca4WwGL+vU1TZwIV/+og7/lnMuKMHsuymU7h9dvYhSd8rt54g2U+yGxg/J7jtFpc9fp/P57VCWIzfv2nYu58H3i7jxdVbSB3Qlz9eMokLJqWjVNdX0DSxs49kNzB+TnDbLaIev1LqbKVUmVJqk1LqtjDlpiql2pVS33FP8etUVVVFs3rHGD/rCpr/KKjkjAX5vLR6C1dOz2TFz09jzuSRIZN+rNycINlPshsYPye47dZtj18plQA8CswCKoE1SqklWuv1XZS7H3jLVcMuaGpqivYuHHGk+234ajfzFhWzpmIXk0en8NwPp5EzcrAIN6dI9pPsBsbPCW67dTuPXyk1HZivtf5WYPl2AK31vZ3K3QzsB6YCr2ut/xGuXjOP3zui5dfU0sYf3/HzzKoKBvXrzW2zxzE3bxS9enBjlCM1dm4g2Q2MnxO8mMc/EtgatFwZWHcQpdRI4ELgMScykSJ5vi0ceX5aa17/vIqZC/J56qMv+O6UUay4dQaXTB3do6QfDTe3kewn2Q2MnxO8mMff1Su388eEPwL/rbVuDzd+q5S6FrgWrMuM5ufnk52dTUVFBc3NzeTl5VFQUEBaWhqJiYlUVlaSk5OD3++nvb2d3NxcCgsLaW1tpaysjOrqaiZNmkRRUREJCQn4fD6Ki4vJyMigtbWV2trag3UmJSWRmZlJaWkpmZmZNDY2UldXd3B7cnIy6enp+P1+srKyqKuro76+/uD2lJQUUlNTKS8vx+fzUVVVRVNT08HtqampDBw4kIqKCnr37s3q1at71KYRI0YAxKRN/fv3Jz8/v0dtCnWcavbCA+9toaSug8zBCfx6en8uPHUUhWtW2WpTfX09NTU1MTlOh9u5V1NTQ01NTY/bFKtzr6amhsbGxpgcp8Pt3KupqaGsrCyCdB0Zrgz1KKW+4N9vEEOBvcC1WutFoep1MtRTUVHh+l3n3eRI8GtubefR9zbx5w/K6dcngf/61je47MQxju93eyTELlpIdgPj54Rgt1gN9awBxiqljlFKJQLfA5YEF9BaH6O1ztRaZwL/AH4SLuk7paKiIlpVu8Lh7vfO+hrO/MP7PPLeJs47Pp0Vt87giumZrtzk/HCPXTSR7AbGzwluu3U71KO1blNK3Yg1WycBeFprXaKUuj6wPSbj+sFkZ2fHepc94nD127pzL/OXlPDuhlp8w5P527Xf5MRj3f1F4eEau1gg2Q2MnxPcdovoB1xa66XA0k7rukz4WuurnGuFp6KiguHDh0d7N7Y53Pxa2tp5/P3NPPLeJhJ6KX55zjiuPukY+kThfreHW+xiiWQ3MH5OcNstLn+529zc7LVCWA4nvw/827lzSQlf7NjDt3NH8KtzsxkxOEmEmxdI9pPsBsbPCW67mevxR4HDwa+6oZl7Xi/ljaJqjhk6gLvOn8CpvmEi3LxEsp9kNzB+TjDX40f2fFuIb7/97R08/kE5Mxe8z/LSGm6d5ePNm0+JSdLvzk0Ckv0ku4Hxc4K5Hj+QlibvlnzBxKvf6s11zFtcjL+miZnj0ph//gRGHdVfhJsUJPtJdgPj5wS33eIy8ScmJnqtEJZ489ve2MK9S0t59dNtjExJ4okrpjBrvDdfcsVb7CQh2Q2MnxPcdovLoZ7KykqvFcISL37tHZrnP67gjAX5vPZ5FTecnsXyW07zLOkHu0lFsp9kNzB+TnDbLS57/Dk5OV4rhCUe/D79chfzFhdTvG03Jx83lLsumEDWsGSv1eIidlKR7AbGzwluu8Vlj9/v93utEBbJfrv2tPJff/+UixauYntjCw9fOpkXrpkmIumD7NiBbD/JbmD8nOC2W1z2+Nvb271WCItEv44Ozd8LtnLfsg00NO/nmpOO4eZZPpL7yjoFJMYuGMl+kt3A+DnBbTdZr/oIyc3N9VohLNL8SqoamLeomHVf1jM1cwi/OGMMU30ju3+iB0iLXWck+0l2A+PnBLfd4nKop7Cw0GuFsEjx271vP/OXlHDewx+ypW4vC+ZO5JXrprOnaqPXaiGRErtQSPaT7AbGzwluu8Vlj//Atbal4rWf1prFhVX8dmkpO5pauPzEMfz8rG8wuH8fEX7hkOwGsv0ku4Hxc4LbbnGZ+A2h2VjTyLzFxXyyeScTMwbz1JVTOD4jxWstg8EgiLgc6qmurvZaISxe+O1paePeZaXMfnAlpdWN/PbCHF79yUldJn3J8ZPsBrL9JLuB8XOC225xeZG2+vp6UlLk9mJj6ae15q2Sr/jNa+upatjH3LwMbps9jtTkviL8eopkN5DtJ9kNjJ8Tgt2O2Iu0FRUVea0Qllj5VezYw1XPrOH6F9cxKKkP/7h+Or+fOzFs0o+lnx0ku4FsP8luYPyc4LZbXI7xJyQkeK0Qlmj77dvfzsL8cha+X05iQi9+fe54rpg+ht4R3hhFcvwku4FsP8luYPyc4LZbXCZ+n8/ntUJYoun33oZa7lxSwpc793L+xHTu+HY2wwf161EdkuMn2Q1k+0l2A+PnBLfd4nKop7i42GuFsETDb1t9M9e9sJarn11DnwTFX350Ig9dOrnHST9afm4h2Q1k+0l2A+PnBLfd4rLHn5GR4bVCWNz0a23r4MkPN/Pwu5sA+MXZ3+BHJx9LYm/779mS4yfZDWT7SXYD4+cEt93iMvG3trZ6rRAWt/xWbdrBvMXFlG/fw1njh/Pr88aTMcT5jVEkx0+yG8j2k+wGxs8JbrvF5VBPbW2t1wphcepXu3sfP/3rp3z/ydXsb9c8c9VUHr9iiitJ3w2/aCLZDWT7SXYD4+cEt93ich6/5Jsig32/tvYOnvt4C//7jp/W9g5+fFoWP56RRb8+7n6jLzl+kt1Atp9kNzB+TjA3W0f2TZHBnt/aip2c+/CH3P36evLGDOHtm0/lZ7N8rid9u36xQrIbyPaT7AbGzwnmZutAUlKS1wph6YlfXVML9y3bwN8LKhkxuB+PXX4C35pwNEopEX6xRrIbyPaT7AbGzwluu8Vl4s/MzPRaISyR+LV3aF5e8yW/e7OMPS1tXHfasfz0jLEMiMGNUSTHT7IbyPaT7AbGzwluu8XlUE9paanXCmHpzq+osoGL/vQRd/yzmOwRA1l20yncPjs7Jkk/Ej8vkewGsv0ku4Hxc4LbbqbHHwVC+TXs3c8Db5fx4uotpA7oy4Pfm8T5E9OjOqzTEz8JSHYD2X6S3cD4OcFtt7hM/I2NjV4rhKWzn9aa/1u3jXuXlrJrbytXTs/klrN8DOrXR4SfJCS7gWw/yW5g/JzgtltcJv66ujqvFcIS7Lfhq93MW1TMmopdnDA6heevmcaE9MEe2smOn2Q3kO0n2Q2MnxPcdjPz+KNAY2MjKjGJP77j55lVFQzq15vbZo9jbt4oevWK7bBOKD+p8ZPsBrL9JLuB8XOCmceP7Pm2WmsWvr6amQvyeeqjL/julFGsuHUGl0wdLSLpg+z4SXYD2X6S3cD4OcFtt4gSv1LqbKVUmVJqk1Lqti62X6aU+jzwWKWUmuiqZSeSk5OjWb1tyrc38YOn/sWfPmthaHJfXv3xf3DvRbkMGZDotdohSI0fyHYD2X6S3cD4OcFtt27H+JVSCcCjwCygElijlFqitV4fVOwL4DSt9S6l1GzgceBEV02DSE9Pj1bVtmhubefR9zbx5w/K6dcngVtPy+An3zqeBCE9/M5Ii18wkt1Atp9kNzB+TnDbLZIe/zRgk9Z6s9a6FXgZuCC4gNZ6ldZ6V2DxEyCq1zf1+/3RrL5HvLO+hjP/8D6PvLeJ845PZ8WtM8hN2iU26YOs+HVGshvI9pPsBsbPCW67RTKrZySwNWi5kvC9+WuAZV1tUEpdC1wL1jtYfn4+2dnZVFRU0NzcTF5eHgUFBaSlpZGYmEhlZSU5OTn4/X7a29vJzc2lsLCQPn36UFZWRnV1NZMmTaKoqIiEhAR8Ph/FxcVkZGTQ2tpKbW3twTqTkpLIzMyktLSUzMxMGhsbqaurO7g9OTmZ9PR0/H4/WVlZ1NXVUV9ff3B7SkoKqamplJeX4/P5+HTjVh5fu4vPtrczMllx75lpTD8uhZKCjxkyZAirV6/uUZtGjBgBEJM2jRgxgvz8/K+1qaqqiqampoPPT01NZeDAgVRUVNg6Tnba1NLSQk1NjWvHye02STj3QrWpoaGBmpqamBwnO21qaGigsbExJsfpcDv3GhoaKCsriyBdR4jWOuwDmAs8GbT8A+DhEGVPB0qB1O7qzcvL03b59NNPbT/XKfv2t+mHlvu1746lOnveMv3n9zfp1rb2Q8p46RcJkv0ku2kt20+ym9bGzwnBbsBa3U1+7e4RSY+/EhgVtJwBVHUupJQ6HngSmK21juqE2Pr6+mhWH5IP/Nu5c0kJX+zYw7dzR/Crc7MZMfjrF0/yyi9SJPtJdgPZfpLdwPg5wW23bufxK6V6A35gJrANWAN8X2tdElRmNLACuEJrvSqSHcfTPP7qhmbufn09S4u+4pihA7jr/Amc6hsmxq+nSPaT7Aay/SS7gfFzQszn8Wut24AbgbewhnFe0VqXKKWuV0pdHyj2ayAV+JNSqlApZS+jR0is5tvub+/g8Q/Kmbngfd4treXWWT7evPmUsEk/ln52kewn2Q1k+0l2A+PnBE+ux6+1Xgos7bTusaD/fwT8yFWzMKSkpER9H6s31zFvcTH+miZmjktj/vkTGHVUZLc+jIWfEyT7SXYD2X6S3cD4OcFtt7i8Vk9qamrU6t7e2MK9S0t59dNtjExJ4okrpjBr/PAe1RFNPzeQ7CfZDWT7SXYD4+cEt93i8pIN5eXlrtfZ3qF5blUFZyzI57XPq7jx9ONYfstpPU760fJzE8l+kt1Atp9kNzB+TnDbLS57/D6fz9X6Pv1yF/MWF1O8bTcnHzeUuy6YQNYw+z+RdtvPbST7SXYD2X6S3cD4OcFtt7js8VdVfW02qS127Wnl9lc/56KFq9je2MIj35/MC9dMc5T03fSLFpL9JLuBbD/JbmD8nOC2W1z2+Juamhw9v6ND8/eCrdy3bAO797VxzUnHcPMsH8ku3frQqV+0kewn2Q1k+0l2A+PnBLfdjrjr8ZdUNTBvUTHrvqxnauYQ7p6Tw7ijB9mqKxp+sUCyn2Q3kO0n2Q2MnxPM9fixN6d19779zF9SwnkPf8iWur0smDuRV66b7nrSt+sXSyT7SXYD2X6S3cD4OcGTefzS6MnUJq01iwur+O3SUnY0tXD5iWP4+VnfYHD/6N3vVvK0MJDtJ9kNZPtJdgPj5wS33eIy8Uf6cWxjTSPzFhfzyeadTMwYzFNXTuH4jOj/SEPqx8UDSPaT7Aay/SS7gfFzgttucTnUU1FREXb7npY27l1WyuwHV1Ja3chvL8zh1Z+cFJOkH4mf10j2k+wGsv0ku4Hxc4LbbnHZ48/Ozu5yvdaaN4u/4u7X11PVsI+5eRncNnscqcl9RfhJQbKfZDeQ7SfZDYyfE9x2O2x6/BU79nDVM2v48UvrGJTUh39cP53fz50Y86Qfyk8Skv0ku4FsP8luYPycYHr8QHNz8yHLC/PL+d/lfhITevHrc8dzxfQx9E7w7j2ts580JPtJdgPZfpLdwPg5wW23uJ/H//zHFfx6cQlnTziauy6YwPBB/Vy0tIfk+cAg20+yG8j2k+wGxs8JZh4//57T+r5/O3e9tp6Z49J49LITRCR9kD0fGGT7SXYD2X6S3cD4OcFtt7hM/GlpaWysaeTGl9YxNi2ZBy+dTEIv5bXWQdLS0rxWCItkP8luINtPshsYPye47RaXib+5oxfXPLeWvn0SeOqqqa5dY8ctEhMTvVYIi2Q/yW4g20+yGxg/J7jtFpeJ/+H8LVQ3NPPEFXmMTPn6zc69prKy0muFsEj2k+wGsv0ku4Hxc4LbbrK6yhHi353AWeOHMXn0EK9VuiQnJ8drhbBI9pPsBrL9JLuB8XOC225x2eNv3NfKkAHRu9aOU/x+v9cKYZHsJ9kNZPtJdgPj5wS33eIy8e9r0/RPlPthpb293WuFsEj2k+wGsv0ku4Hxc4LbbnGX+Ns7NPs7oH9igtcqIcnNzfVaISyS/SS7gWw/yW5g/JzgtlvcJf7m/dY7n+TEX1hY6LVCWCT7SXYD2X6S3cD4OcFtt7hL/Htb2gBED/WMGDHCa4WwSPaT7Aay/SS7gfFzgttu8Zf4W+X3+A0Gg0EyJvFHgerqaq8VwiLZT7IbyPaT7AbGzwluu8Vd4m/eL3+oZ9KkSV4rhEWyn2Q3kO0n2Q2MnxPcdou7xL+nRX6Pv6ioyGuFsEj2k+wGsv0ku4Hxc4LbbnGX+P891CO3x5+QIPdNCWT7SXYD2X6S3cD4OcFtt7hL/MMGJnLG2CEcNUDuBZV8Pp/XCmGR7CfZDWT7SXYD4+cEt93iLvHnjTmKK7JaOXqwjGvvd0VxcbHXCmGR7CfZDWT7SXYD4+cEt93iLvEDZGRkeK0QFuNnH8luINtPshsYPye47RZR4ldKna2UKlNKbVJK3dbFdqWUeiiw/XOl1AmuWnaitbU1mtU7xvjZR7IbyPaT7AbGzwluu3Wb+JVSCcCjwGxgPHCpUmp8p2KzgbGBx7XAQlctO1FbWxvN6h1j/Owj2Q1k+0l2A+PnBLfdIunxTwM2aa03a61bgZeBCzqVuQB4Xlt8AqQopaL2++e8vLxoVe0Kxs8+kt1Atp9kNzB+TnDbLZI5kSOBrUHLlcCJEZQZCRzyczOl1LVYnwhIT08nPz+f7OxsKioqaG5uJi8vj4KCAtLS0khMTKSyspKcnBz8fj/t7e3k5uZSWFjIjh07yM3Npbq6mkmTJlFUVERCQgI+n4/i4mIyMjJobW2ltrb2YJ1JSUlkZmZSWlpKZmYmjY2N1NXVHdyenJxMeno6fr+frKws6urqqK+vP7g9JSWF1NRUysvL8fl8VFVV0dTUdHB7amoqAwcOpKKigqamJoYNG9ajNh24Fkcs2tTS0kLfvn171CY7x8lOm7Zt28aZZ54Zk+N0uJ17+fn5zJ49OybHyU6bCgsLufjii2NynA63c2/RokVMnz49kpweGVrrsA9gLvBk0PIPgIc7lXkDODlo+V0gL1y9eXl52i6ffPKJ7efGAuNnH8luWsv2k+ymtfFzQrAbsFZ3k7e7e0Qy1FMJjApazgCqbJRxjczMzGhV7QrGzz6S3UC2n2Q3MH5OcNstksS/BhirlDpGKZUIfA9Y0qnMEuCKwOyebwINWuuoXfGotLQ0WlW7gvGzj2Q3kO0n2Q2MnxPcdut2jF9r3aaUuhF4C0gAntZalyilrg9sfwxYCpwDbAL2Ald3V29BQcEOpdQWm95DgR02nxsLjJ99JLuBbD/JbmD8nBDsNsZpZcoaMoovlFJrtdZTvPYIhfGzj2Q3kO0n2Q2MnxPcdovLX+4aDAaDwT4m8RsMBsMRRrwm/se9FugG42cfyW4g20+yGxg/J7jqFpdj/AaDwWCwT7z2+A0Gg8FgE5P4DQaD4Qgj7hJ/d5eIjsH+Ryml3lNKlSqlSpRSNwXWz1dKbVNKFQYe5wQ95/aAb5lS6lsxcKxQShUFPNYG1h2llHpHKbUx8HdIrP2UUt8Iik+hUmq3UupmL2OnlHpaKVWrlCoOWtfjWCml8gIx3xS4RLmKot/vlVIbApdA/6dSKiWwPlMp1RwUx8ei6RfCrcfHMsax+1uQW4VSqjCwPtaxC5VHYnPuOb3mQywfWD8gKweOBRKBz4DxMXYYAZwQ+H8g4Me6XPV84OddlB8f8OwLHBPwT4iyYwUwtNO63wG3Bf6/DbjfK7+gY/kV1o9RPIsdcCpwAlDsJFbAv4DpgAKWAbOj6HcW0Dvw//1BfpnB5TrV47pfCLceH8tYxq7T9gXArz2KXag8EpNzL956/JFcIjqqaK2rtdbrAv83AqVYVyINxQXAy1rrFq31F1i/bp4WfdMuPZ4L/P8cMMdjv5lAudY63K+3o+6mtf4A2NnFfiOOlbIuQT5Ia/2xtl6Jzwc9x3U/rfXbWuu2wOInWNfGCkm0/ELELhQiYneAQK/4u8Bfw9URxdiFyiMxOffiLfGHuvyzJyilMoHJwOrAqhsDH7+fDvqI5oWzBt5WShUo61LYAMN14PpJgb9pHvqBdc2n4BedlNhBz2M1MvB/rD0BfojVyzvAMUqpT5VS7yulTgmsi7VfT46lV7E7BajRWm8MWudJ7DrlkZice/GW+Lsau/JkPqpSKhn4P+BmrfVurLuOZQGTsO5DsOBA0S6eHm3nk7TWJ2DdGe0GpdSpYcrG3E9ZF/s7H/h7YJWk2IUjlI8nnkqpO4A24KXAqmpgtNZ6MnAL8Bel1KAY+/X0WHp1jC/l0I6HJ7HrIo+ELBrCw5ZfvCX+mF7+ORRKqT5YB+slrfWrAFrrGq11u9a6A3iCfw9JSkd4MgAAAcdJREFUxNxZa10V+FsL/DPgUhP4WHjg4+uBe7l5EdPZwDqtdU3AU0zsAvQ0VpUcOtwSdU+l1JXAucBlgY/4BIYB6gL/F2CNA/ti6WfjWHoRu97ARcDfgrxjHruu8ggxOvfiLfFHconoqBIYG3wKKNVa/yFoffCtJi8EDswkWAJ8TynVVyl1DNZ9if8VRb8BSqmBB/7H+iKwOOBxZaDYlcBiL/wCHNLbkhK7IHoUq8BH8kal1DcD58cVQc9xHaXU2cB/A+drrfcGrR+mrHtko5Q6NuC3OZZ+PT2WsY5dgDOBDVrrg0MksY5dqDxCrM49p99Ox/qBdflnP9Y78h0e7P9krI9SnwOFgcc5wAtAUWD9EmBE0HPuCPiW4dKMhTB+x2J9+/8ZUHIgRkAq1p3RNgb+HuWRX3+gDhgctM6z2GG9AVUD+7F6T9fYiRUwBSvJlQOPEPhVfJT8NmGN9x44/x4LlL04cMw/A9YB50XTL4Rbj49lLGMXWP8scH2nsrGOXag8EpNzz1yywWAwGI4w4m2ox2AwGAwOMYnfYDAYjjBM4jcYDIYjDJP4DQaD4QjDJH6DwWA4wjCJ32AwGI4wTOI3GAyGI4z/Dwj9txGcD85jAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "s = sum(unique_Variation.values);\n",
    "h = unique_Variation.values/s;\n",
    "c = np.cumsum(h)\n",
    "plt.plot(c)\n",
    "plt.title(\"Cumulative distribution of Variation\", fontsize = 14)\n",
    "plt.grid(color='gray', linestyle='--', linewidth=0.5)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The figure shows that the 1500 variations have almost 80% of occurrence. Therefore, variations is important for us and we need to transform it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2124, 1952)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# one-hot encoding of Variation feature.\n",
    "variation_vectorizer = CountVectorizer()\n",
    "X_train_variation_onehotCoding = variation_vectorizer.fit_transform(X_train['Variation'])\n",
    "X_test_variation_onehotCoding = variation_vectorizer.transform(X_test['Variation'])\n",
    "X_cv_variation_onehotCoding = variation_vectorizer.transform(X_cv['Variation'])\n",
    "\n",
    "X_train_variation_onehotCoding.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 0, 0, ..., 0, 0, 0], dtype=int64)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_vectorizer.fit_transform(X_train['Variation']).toarray()[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2124, 9)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# response-coding of the Variation feature\n",
    "# alpha is used for laplace smoothing\n",
    "alpha = 1\n",
    "    \n",
    "X_train_variation_responseCoding = np.array(get_gv_feature(alpha, \"Variation\", X_train,X_train))\n",
    "X_test_variation_responseCoding = np.array(get_gv_feature(alpha, \"Variation\", X_test,X_train))\n",
    "X_cv_variation_responseCoding = np.array(get_gv_feature(alpha, \"Variation\", X_cv,X_train))\n",
    "\n",
    "X_train_variation_responseCoding.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Text feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Total number of repeatition of each word in TEXT:\n",
    "\n",
    "# for each row in df['TEXT']\n",
    "# split words by space.\n",
    "# make a dict with those words.\n",
    "# increment the word's count whenever it occures.\n",
    "\n",
    "def Count_number_of_each_word_in_TEXT(df):\n",
    "    dictionary = defaultdict(int)\n",
    "    for index, row in df.iterrows():\n",
    "        for word in row['TEXT'].split():\n",
    "            dictionary[word] +=1\n",
    "    return dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(int,\n",
       "            {'etv6': 1800,\n",
       "             'ntrk3': 569,\n",
       "             'en': 830,\n",
       "             'chimeric': 880,\n",
       "             'tyrosine': 13607,\n",
       "             'kinase': 42629,\n",
       "             'generated': 5647,\n",
       "             '12': 19265,\n",
       "             '15': 17705,\n",
       "             'translocations': 1387,\n",
       "             'dominantly': 116,\n",
       "             'acting': 609,\n",
       "             'oncoprotein': 352,\n",
       "             'diverse': 2029,\n",
       "             'tumor': 44997,\n",
       "             'types': 7110,\n",
       "             'previously': 18276,\n",
       "             'showed': 17464,\n",
       "             'insulin': 1613,\n",
       "             'like': 6826,\n",
       "             'growth': 24516,\n",
       "             'factor': 10399,\n",
       "             '1': 138036,\n",
       "             'receptor': 16441,\n",
       "             'igf1r': 590,\n",
       "             'essential': 3514,\n",
       "             'mediated': 9055,\n",
       "             'oncogenesis': 708,\n",
       "             'substrate': 5820,\n",
       "             'irs1': 187,\n",
       "             'constitutively': 3006,\n",
       "             'phosphorylated': 7369,\n",
       "             'bound': 5925,\n",
       "             'transformed': 2960,\n",
       "             'cells': 117058,\n",
       "             'given': 5744,\n",
       "             'also': 42734,\n",
       "             'adapter': 215,\n",
       "             'hypothesized': 963,\n",
       "             'might': 8079,\n",
       "             'localize': 310,\n",
       "             'membrane': 5662,\n",
       "             'activate': 4743,\n",
       "             'phosphatidylinositol': 1277,\n",
       "             '3': 87670,\n",
       "             'pi3k': 8209,\n",
       "             'akt': 13000,\n",
       "             'critical': 5072,\n",
       "             'study': 23887,\n",
       "             'examined': 7154,\n",
       "             'complexes': 4232,\n",
       "             'detail': 926,\n",
       "             'find': 1326,\n",
       "             'active': 10477,\n",
       "             'required': 8001,\n",
       "             'transformation': 5428,\n",
       "             'present': 9977,\n",
       "             'high': 17668,\n",
       "             'molecular': 11857,\n",
       "             'weight': 1341,\n",
       "             'colocalizes': 44,\n",
       "             'plasma': 2860,\n",
       "             'activity': 48313,\n",
       "             'intact': 1629,\n",
       "             'cytoplasmic': 2281,\n",
       "             'y950': 3,\n",
       "             'residue': 9135,\n",
       "             'docking': 702,\n",
       "             'site': 17185,\n",
       "             'confirming': 1092,\n",
       "             'importance': 2519,\n",
       "             'interaction': 12215,\n",
       "             'dual': 1525,\n",
       "             'specificity': 3814,\n",
       "             'insr': 57,\n",
       "             'inhibitor': 13197,\n",
       "             'bms': 151,\n",
       "             '536924': 39,\n",
       "             'blocks': 1161,\n",
       "             'cell': 82203,\n",
       "             'survival': 12558,\n",
       "             'irs': 347,\n",
       "             'proteins': 26222,\n",
       "             'induces': 1924,\n",
       "             'striking': 804,\n",
       "             'shift': 1645,\n",
       "             'smaller': 1156,\n",
       "             'sized': 200,\n",
       "             'conclude': 785,\n",
       "             'tripartite': 14,\n",
       "             'complex': 12566,\n",
       "             'localizes': 260,\n",
       "             'findings': 6572,\n",
       "             'provide': 6593,\n",
       "             'explanation': 1135,\n",
       "             'observed': 21224,\n",
       "             'dependency': 478,\n",
       "             'blocking': 877,\n",
       "             'may': 31978,\n",
       "             'therefore': 10686,\n",
       "             'tractable': 121,\n",
       "             'therapeutic': 5491,\n",
       "             'strategy': 1568,\n",
       "             'many': 7606,\n",
       "             'driven': 3126,\n",
       "             'keywords': 973,\n",
       "             'top': 3175,\n",
       "             'page': 4309,\n",
       "             'introduction': 3972,\n",
       "             'p13': 332,\n",
       "             'q25': 44,\n",
       "             'congenital': 456,\n",
       "             'fibrosarcoma': 112,\n",
       "             'knezevich': 22,\n",
       "             'et': 68636,\n",
       "             'al': 68565,\n",
       "             '1998': 3090,\n",
       "             'cellular': 7733,\n",
       "             'mesoblastic': 12,\n",
       "             'nephroma': 13,\n",
       "             'argani': 1,\n",
       "             '2000': 3572,\n",
       "             'adult': 1527,\n",
       "             'acute': 2135,\n",
       "             'myeloid': 3173,\n",
       "             'leukemia': 5566,\n",
       "             'eguchi': 8,\n",
       "             '1999': 2917,\n",
       "             'secretory': 140,\n",
       "             'breast': 14399,\n",
       "             'carcinoma': 6040,\n",
       "             'tognon': 22,\n",
       "             '2002': 3426,\n",
       "             'recently': 6680,\n",
       "             'fusion': 12382,\n",
       "             'transcripts': 2068,\n",
       "             'detected': 11771,\n",
       "             'mammary': 1434,\n",
       "             'analog': 135,\n",
       "             'salivary': 118,\n",
       "             'glands': 252,\n",
       "             'skalova': 1,\n",
       "             '2010': 4708,\n",
       "             'skin': 2404,\n",
       "             'kazakov': 1,\n",
       "             'chronic': 2083,\n",
       "             'eosinophilic': 196,\n",
       "             'forghieri': 1,\n",
       "             '2011': 5471,\n",
       "             'encode': 893,\n",
       "             'sterile': 148,\n",
       "             'motif': 5517,\n",
       "             'dimerization': 2703,\n",
       "             'sam': 186,\n",
       "             'domain': 41267,\n",
       "             'transcription': 10646,\n",
       "             'linked': 2931,\n",
       "             'neurotrophin': 38,\n",
       "             'expression': 49606,\n",
       "             'leads': 3320,\n",
       "             'elevated': 3022,\n",
       "             'cyclin': 8844,\n",
       "             'd1': 6096,\n",
       "             'mrna': 5948,\n",
       "             'protein': 54276,\n",
       "             'correlates': 1012,\n",
       "             'increased': 14950,\n",
       "             'cycle': 6682,\n",
       "             'progression': 8658,\n",
       "             '2001': 3201,\n",
       "             'nih3t3': 1635,\n",
       "             'fibroblasts': 1890,\n",
       "             'activates': 1589,\n",
       "             'two': 34188,\n",
       "             'major': 4377,\n",
       "             'effector': 2049,\n",
       "             'pathways': 8658,\n",
       "             'wt': 19422,\n",
       "             'namely': 584,\n",
       "             'ras': 27442,\n",
       "             'erk': 8310,\n",
       "             'cascades': 353,\n",
       "             'blocked': 1115,\n",
       "             'inhibition': 12698,\n",
       "             'either': 13024,\n",
       "             'pathway': 17764,\n",
       "             'wild': 36197,\n",
       "             'type': 46235,\n",
       "             'utilizes': 67,\n",
       "             'juxtamembrane': 1537,\n",
       "             'y516': 3,\n",
       "             'bind': 4672,\n",
       "             'adapters': 74,\n",
       "             'shc': 254,\n",
       "             'sh2b': 5,\n",
       "             'p85': 3697,\n",
       "             'subunit': 3897,\n",
       "             'downstream': 7135,\n",
       "             'signaling': 22783,\n",
       "             'huang': 368,\n",
       "             'reichardt': 7,\n",
       "             '2003': 3477,\n",
       "             'however': 20249,\n",
       "             'direct': 5013,\n",
       "             'binding': 40215,\n",
       "             'known': 12990,\n",
       "             'demonstrable': 41,\n",
       "             'wai': 16,\n",
       "             'lannon': 5,\n",
       "             'sorensen': 4,\n",
       "             '2005': 4255,\n",
       "             'likely': 11794,\n",
       "             'tyr': 1253,\n",
       "             '516': 48,\n",
       "             'absent': 1280,\n",
       "             'due': 7624,\n",
       "             'position': 6726,\n",
       "             'point': 7576,\n",
       "             'instead': 1051,\n",
       "             'directly': 4924,\n",
       "             'binds': 2926,\n",
       "             'npxy': 8,\n",
       "             'c': 65012,\n",
       "             'terminus': 3529,\n",
       "             'ptb': 66,\n",
       "             '2004': 3802,\n",
       "             'morrison': 72,\n",
       "             'recruit': 432,\n",
       "             'grb2': 323,\n",
       "             'respectively': 10499,\n",
       "             'fails': 335,\n",
       "             'transform': 807,\n",
       "             'mouse': 8372,\n",
       "             'embryo': 420,\n",
       "             'derived': 8475,\n",
       "             'mice': 14734,\n",
       "             'targeted': 5644,\n",
       "             'disruption': 1653,\n",
       "             'gene': 39954,\n",
       "             'r': 6725,\n",
       "             'reintroduction': 174,\n",
       "             'restores': 265,\n",
       "             'dependent': 11686,\n",
       "             'fractions': 852,\n",
       "             'martin': 104,\n",
       "             '2006': 5016,\n",
       "             'myristoylated': 92,\n",
       "             'form': 7110,\n",
       "             'enmyr': 1,\n",
       "             'associated': 25973,\n",
       "             'rescues': 152,\n",
       "             'taken': 3079,\n",
       "             'together': 5544,\n",
       "             'data': 38801,\n",
       "             'role': 13250,\n",
       "             'localization': 4361,\n",
       "             'sain': 1,\n",
       "             'domains': 12253,\n",
       "             'craparo': 2,\n",
       "             '1995': 2312,\n",
       "             'postulated': 358,\n",
       "             'functions': 4945,\n",
       "             'bridging': 151,\n",
       "             'molecule': 3465,\n",
       "             'work': 2444,\n",
       "             'evidence': 8389,\n",
       "             'demonstrate': 3104,\n",
       "             'significantly': 10683,\n",
       "             'impacts': 260,\n",
       "             'formation': 7934,\n",
       "             'activation': 32351,\n",
       "             'addition': 12029,\n",
       "             'use': 5992,\n",
       "             'mass': 1781,\n",
       "             'spectrometry': 618,\n",
       "             'show': 9296,\n",
       "             'hek293': 1380,\n",
       "             'family': 13306,\n",
       "             'members': 3864,\n",
       "             'irs4': 21,\n",
       "             'irs2': 48,\n",
       "             'dramatically': 683,\n",
       "             'reduces': 590,\n",
       "             'interactions': 7330,\n",
       "             'highlights': 631,\n",
       "             'comprising': 599,\n",
       "             'suggests': 5429,\n",
       "             'mechanism': 7811,\n",
       "             'explain': 1651,\n",
       "             'results': 29092,\n",
       "             'discussion': 4227,\n",
       "             'confirm': 2588,\n",
       "             'used': 24835,\n",
       "             'il': 4635,\n",
       "             'murine': 2165,\n",
       "             '32d': 976,\n",
       "             'line': 14910,\n",
       "             'expresses': 327,\n",
       "             'lacks': 813,\n",
       "             'sun': 483,\n",
       "             'baserga': 9,\n",
       "             'fail': 620,\n",
       "             'survive': 196,\n",
       "             'media': 1506,\n",
       "             'lacking': 2082,\n",
       "             'unless': 312,\n",
       "             'engineered': 1045,\n",
       "             'overexpress': 279,\n",
       "             'secrete': 22,\n",
       "             'alone': 4008,\n",
       "             'confer': 3262,\n",
       "             'independent': 12137,\n",
       "             'figure': 63191,\n",
       "             '1a': 4527,\n",
       "             'bottom': 1518,\n",
       "             'added': 2868,\n",
       "             'medium': 5088,\n",
       "             'supplementary': 18605,\n",
       "             'figures': 4162,\n",
       "             's1a': 477,\n",
       "             'b': 36263,\n",
       "             'co': 7720,\n",
       "             'along': 2143,\n",
       "             'conferred': 1247,\n",
       "             'proliferation': 12007,\n",
       "             'properties': 3375,\n",
       "             'absence': 7202,\n",
       "             'requires': 1790,\n",
       "             'since': 4491,\n",
       "             'overexpression': 4918,\n",
       "             'enhances': 709,\n",
       "             'transforming': 3471,\n",
       "             'ability': 9943,\n",
       "             'tested': 9212,\n",
       "             'could': 13473,\n",
       "             'support': 3454,\n",
       "             'null': 3216,\n",
       "             'transduced': 2662,\n",
       "             '1b': 4023,\n",
       "             'failed': 2427,\n",
       "             'overcome': 1043,\n",
       "             'requirement': 819,\n",
       "             'indicate': 6193,\n",
       "             'unfortunately': 1603,\n",
       "             'unable': 2691,\n",
       "             'accessible': 1642,\n",
       "             'alternative': 4262,\n",
       "             'text': 2047,\n",
       "             'require': 3030,\n",
       "             'assistance': 1113,\n",
       "             'access': 1883,\n",
       "             'image': 4122,\n",
       "             'please': 1134,\n",
       "             'contact': 2874,\n",
       "             'help': 2128,\n",
       "             'nature': 3103,\n",
       "             'com': 1667,\n",
       "             'author': 4859,\n",
       "             'promote': 3769,\n",
       "             'deficient': 2856,\n",
       "             'represent': 4042,\n",
       "             'average': 3194,\n",
       "             'three': 20901,\n",
       "             'experiments': 9246,\n",
       "             'western': 6875,\n",
       "             'blot': 4230,\n",
       "             'analysis': 38157,\n",
       "             'e': 14073,\n",
       "             'presence': 12883,\n",
       "             'soft': 1824,\n",
       "             'agar': 1822,\n",
       "             'colony': 2875,\n",
       "             'assays': 16480,\n",
       "             'cannot': 2567,\n",
       "             'lack': 3735,\n",
       "             'igfir': 4,\n",
       "             'icmyr': 11,\n",
       "             'positive': 11167,\n",
       "             'myr': 420,\n",
       "             'intracellular': 1914,\n",
       "             'yf': 58,\n",
       "             'y950f': 2,\n",
       "             'ka': 145,\n",
       "             'k1003a': 9,\n",
       "             'dead': 1088,\n",
       "             'ic': 460,\n",
       "             'non': 10546,\n",
       "             'grown': 2194,\n",
       "             'anchorage': 1169,\n",
       "             'conditions': 4558,\n",
       "             '24': 12358,\n",
       "             'h': 13295,\n",
       "             'dmem': 1528,\n",
       "             'containing': 11680,\n",
       "             '0': 67725,\n",
       "             '25': 11559,\n",
       "             'serum': 5852,\n",
       "             'lysed': 1853,\n",
       "             'subjected': 2757,\n",
       "             'blotting': 2573,\n",
       "             'eight': 3548,\n",
       "             'lines': 22444,\n",
       "             'plated': 1387,\n",
       "             'assessed': 4994,\n",
       "             'n': 23010,\n",
       "             'represents': 2105,\n",
       "             'p': 39612,\n",
       "             '0001': 1545,\n",
       "             'student': 884,\n",
       "             'test': 8462,\n",
       "             'full': 7960,\n",
       "             'legend': 1128,\n",
       "             '129k': 1,\n",
       "             'activated': 9768,\n",
       "             'latter': 1560,\n",
       "             'whether': 11083,\n",
       "             'necessary': 2500,\n",
       "             'defective': 2724,\n",
       "             'mutants': 33795,\n",
       "             'extracellular': 3190,\n",
       "             'ligand': 6536,\n",
       "             'tethered': 135,\n",
       "             'using': 40873,\n",
       "             'myristoylation': 97,\n",
       "             'tag': 1117,\n",
       "             'myrisotylated': 1,\n",
       "             's2': 4468,\n",
       "             'constructs': 5469,\n",
       "             'signal': 6311,\n",
       "             'designated': 525,\n",
       "             'expressed': 12515,\n",
       "             'antibodies': 7280,\n",
       "             '1c': 2555,\n",
       "             'panel': 5732,\n",
       "             'shown': 28184,\n",
       "             'panels': 1255,\n",
       "             'retaining': 197,\n",
       "             'gsk3': 391,\n",
       "             'phosphorylation': 19442,\n",
       "             'furthermore': 5274,\n",
       "             'expressing': 15595,\n",
       "             'exhibited': 3381,\n",
       "             '1d': 1525,\n",
       "             'strongly': 3377,\n",
       "             'retains': 443,\n",
       "             'interacts': 1979,\n",
       "             'demonstrated': 6871,\n",
       "             'forms': 4151,\n",
       "             'moreover': 4051,\n",
       "             'lysates': 4737,\n",
       "             'fast': 429,\n",
       "             'liquid': 1022,\n",
       "             'chromatography': 795,\n",
       "             'fplc': 16,\n",
       "             'ranging': 875,\n",
       "             '45': 4363,\n",
       "             '2': 100749,\n",
       "             '106': 1638,\n",
       "             'kda': 1963,\n",
       "             'identify': 5386,\n",
       "             'potential': 8470,\n",
       "             'interacting': 1528,\n",
       "             'immunoprecipitated': 1925,\n",
       "             'different': 19463,\n",
       "             'performed': 15645,\n",
       "             'phosphotyrosine': 1122,\n",
       "             'similar': 15281,\n",
       "             'previous': 7348,\n",
       "             '150': 2121,\n",
       "             '2a': 4624,\n",
       "             'additional': 9865,\n",
       "             '200': 2874,\n",
       "             '180': 514,\n",
       "             'precipitated': 790,\n",
       "             'species': 2386,\n",
       "             'vector': 8666,\n",
       "             'control': 17452,\n",
       "             'k380n': 21,\n",
       "             's3': 3009,\n",
       "             'reprobe': 1,\n",
       "             'band': 1633,\n",
       "             'reciprocal': 394,\n",
       "             'immunoprecipitation': 2365,\n",
       "             'confirmed': 6944,\n",
       "             'association': 7100,\n",
       "             '2b': 3852,\n",
       "             'associates': 473,\n",
       "             'large': 9359,\n",
       "             'size': 4703,\n",
       "             'buffer': 5451,\n",
       "             '10': 36230,\n",
       "             'mm': 11061,\n",
       "             'tris': 2119,\n",
       "             'nacl': 1667,\n",
       "             '5': 67678,\n",
       "             'mgcl2': 695,\n",
       "             '4': 55060,\n",
       "             'np': 698,\n",
       "             '40': 8792,\n",
       "             'superose': 9,\n",
       "             '6hr10': 1,\n",
       "             '30': 13284,\n",
       "             'gel': 3441,\n",
       "             'filtration': 591,\n",
       "             'column': 736,\n",
       "             'ge': 474,\n",
       "             'healthcare': 395,\n",
       "             'mississauga': 19,\n",
       "             'canada': 203,\n",
       "             'ml': 9331,\n",
       "             'collected': 2335,\n",
       "             'anti': 14325,\n",
       "             'immunoprecipitations': 310,\n",
       "             'fraction': 2680,\n",
       "             'numbers': 2763,\n",
       "             'weights': 107,\n",
       "             'indicated': 11576,\n",
       "             '4g10': 181,\n",
       "             'antibody': 9130,\n",
       "             'probe': 2501,\n",
       "             'majority': 4373,\n",
       "             'elutes': 18,\n",
       "             'corresponding': 5413,\n",
       "             '1000': 1049,\n",
       "             'range': 5060,\n",
       "             'immunoprecipitate': 179,\n",
       "             'unidentified': 244,\n",
       "             'identified': 25250,\n",
       "             'doublet': 149,\n",
       "             'stably': 2479,\n",
       "             'ha': 3916,\n",
       "             'tagged': 3890,\n",
       "             'conjugated': 1277,\n",
       "             'beads': 1853,\n",
       "             '20': 22394,\n",
       "             'santa': 2332,\n",
       "             'cruz': 2171,\n",
       "             'biotechnology': 1495,\n",
       "             'ca': 5045,\n",
       "             'usa': 2430,\n",
       "             'roche': 1451,\n",
       "             'laval': 1,\n",
       "             'detect': 2426,\n",
       "             'immunoprecipitating': 34,\n",
       "             '69k': 12,\n",
       "             'reported': 18631,\n",
       "             'localized': 1800,\n",
       "             'repeated': 915,\n",
       "             'attempts': 224,\n",
       "             'precipitate': 86,\n",
       "             'unsuccessful': 46,\n",
       "             'transient': 1478,\n",
       "             'myers': 55,\n",
       "             '1993': 1455,\n",
       "             'visualize': 170,\n",
       "             'bimolecular': 15,\n",
       "             'fluorescence': 2462,\n",
       "             'complementation': 1145,\n",
       "             'putative': 2516,\n",
       "             'interactors': 51,\n",
       "             'individually': 363,\n",
       "             'fused': 1359,\n",
       "             'separate': 1255,\n",
       "             'halves': 45,\n",
       "             'fluorescent': 1155,\n",
       "             'venus': 13,\n",
       "             'live': 266,\n",
       "             'interact': 3034,\n",
       "             'recombines': 1,\n",
       "             'reconstituted': 392,\n",
       "             'kerppola': 5,\n",
       "             '2009': 3348,\n",
       "             'terminal': 11806,\n",
       "             'v1': 276,\n",
       "             'second': 7630,\n",
       "             'copy': 6197,\n",
       "             'v2': 290,\n",
       "             'demonstrating': 1197,\n",
       "             'self': 1102,\n",
       "             'primarily': 1603,\n",
       "             'cytoplasm': 1335,\n",
       "             '3a': 3748,\n",
       "             'left': 2902,\n",
       "             'dimerizing': 38,\n",
       "             'v80e': 1,\n",
       "             'mutant': 47154,\n",
       "             'negligible': 134,\n",
       "             'background': 1795,\n",
       "             'robust': 1288,\n",
       "             'right': 2899,\n",
       "             'indicates': 3333,\n",
       "             'colocalize': 50,\n",
       "             'wondered': 47,\n",
       "             'acts': 894,\n",
       "             'bridge': 672,\n",
       "             'mediate': 1366,\n",
       "             'mentioned': 564,\n",
       "             'readily': 878,\n",
       "             'effects': 12409,\n",
       "             'treated': 10377,\n",
       "             'chemical': 1110,\n",
       "             'nm': 5918,\n",
       "             'potency': 1506,\n",
       "             'wittman': 2,\n",
       "             'fractionation': 276,\n",
       "             'concentrations': 5045,\n",
       "             'studies': 20406,\n",
       "             'selectively': 918,\n",
       "             'inhibit': 4146,\n",
       "             'kinases': 6821,\n",
       "             'although': 15896,\n",
       "             'remained': 1281,\n",
       "             '3b': 3152,\n",
       "             'dramatic': 877,\n",
       "             'redistribution': 117,\n",
       "             'low': 9907,\n",
       "             'treatment': 20031,\n",
       "             'middle': 971,\n",
       "             'dimeric': 384,\n",
       "             'monomeric': 265,\n",
       "             'indicating': 4753,\n",
       "             'maintain': 784,\n",
       "             'showing': 4064,\n",
       "             'colocalization': 237,\n",
       "             'venus1': 1,\n",
       "             'venus2': 2,\n",
       "             'transiently': 2105,\n",
       "             'transfected': 12107,\n",
       "             'whereas': 12390,\n",
       "             'shifts': 310,\n",
       "             'larger': 3029,\n",
       "             'ir': 471,\n",
       "             '53624': 1,\n",
       "             '100': 11798,\n",
       "             'min': 6034,\n",
       "             'fractionated': 172,\n",
       "             'described': 17177,\n",
       "             'mid': 156,\n",
       "             'upon': 4469,\n",
       "             'shifted': 364,\n",
       "             'suggesting': 8044,\n",
       "             'myrirs1': 7,\n",
       "             'enable': 365,\n",
       "             'construct': 2960,\n",
       "             'made': 2583,\n",
       "             'isolated': 3411,\n",
       "             'protocol': 2109,\n",
       "             'found': 24356,\n",
       "             'level': 9623,\n",
       "             'membranes': 1661,\n",
       "             'total': 10906,\n",
       "             'caveolin': 29,\n",
       "             'counterpart': 275,\n",
       "             'significant': 11998,\n",
       "             'difference': 3785,\n",
       "             'compared': 16913,\n",
       "             '346': 96,\n",
       "             'statistical': 3163,\n",
       "             'differences': 5638,\n",
       "             '00063': 1,\n",
       "             '01842': 1,\n",
       "             'tailed': 537,\n",
       "             '147k': 2,\n",
       "             'one': 26177,\n",
       "             'possibility': 3538,\n",
       "             'target': 10507,\n",
       "             '3c': 2180,\n",
       "             'levels': 22388,\n",
       "             'distribution': 3533,\n",
       "             'restore': 643,\n",
       "             'fact': 3701,\n",
       "             'appears': 2645,\n",
       "             'decrease': 2608,\n",
       "             '3d': 1797,\n",
       "             'functional': 21146,\n",
       "             'axis': 953,\n",
       "             'adherent': 114,\n",
       "             'cultures': 1205,\n",
       "             's4a': 340,\n",
       "             'reversed': 256,\n",
       "             'poly': 473,\n",
       "             'adp': 310,\n",
       "             'ribose': 124,\n",
       "             'polymerase': 2256,\n",
       "             'parp': 955,\n",
       "             'caspase': 1890,\n",
       "             'cleavage': 1205,\n",
       "             's4b': 196,\n",
       "             'reduced': 8499,\n",
       "             's4c': 172,\n",
       "             'supported': 1771,\n",
       "             'recent': 5561,\n",
       "             'report': 4182,\n",
       "             'epithelial': 2804,\n",
       "             'vitro': 13372,\n",
       "             'related': 8129,\n",
       "             'compound': 2023,\n",
       "             '754807': 8,\n",
       "             'vivo': 8823,\n",
       "             'consistent': 9622,\n",
       "             'model': 10524,\n",
       "             'whereby': 395,\n",
       "             'maintained': 1720,\n",
       "             'independently': 1380,\n",
       "             'evaluate': 1761,\n",
       "             'tandem': 1214,\n",
       "             'compare': 1355,\n",
       "             'untreated': 1520,\n",
       "             'abundance': 789,\n",
       "             'well': 18531,\n",
       "             'interactor': 32,\n",
       "             'plc': 659,\n",
       "             'strikingly': 588,\n",
       "             'abundant': 489,\n",
       "             'another': 5284,\n",
       "             'member': 1164,\n",
       "             'among': 8989,\n",
       "             'express': 2708,\n",
       "             'predominantly': 1365,\n",
       "             '4a': 2948,\n",
       "             'contains': 3817,\n",
       "             'immunoprecipitates': 934,\n",
       "             'coverage': 909,\n",
       "             '35': 6565,\n",
       "             '23': 8251,\n",
       "             'mascot': 24,\n",
       "             'scores': 1620,\n",
       "             '991': 10,\n",
       "             '392': 49,\n",
       "             'averaged': 179,\n",
       "             'peptide': 4467,\n",
       "             'areas': 1047,\n",
       "             'five': 6426,\n",
       "             'peptides': 1660,\n",
       "             'decreased': 4512,\n",
       "             'samples': 17532,\n",
       "             '4b': 2488,\n",
       "             'rapidly': 1047,\n",
       "             'destabilizes': 228,\n",
       "             'representative': 2866,\n",
       "             'ion': 450,\n",
       "             'chromatograms': 163,\n",
       "             's5': 1216,\n",
       "             'validated': 1735,\n",
       "             'markedly': 1357,\n",
       "             'precipitating': 39,\n",
       "             '4c': 2192,\n",
       "             'finally': 3403,\n",
       "             'precipitation': 317,\n",
       "             '4d': 1230,\n",
       "             'underscore': 232,\n",
       "             'disrupting': 668,\n",
       "             'disrupts': 665,\n",
       "             'isoform': 1839,\n",
       "             'highly': 7511,\n",
       "             'ratios': 1334,\n",
       "             'peak': 705,\n",
       "             'area': 1616,\n",
       "             'values': 6506,\n",
       "             'see': 6367,\n",
       "             'table': 33873,\n",
       "             's1': 5115,\n",
       "             'determine': 6420,\n",
       "             'significance': 4692,\n",
       "             '2e': 739,\n",
       "             '6': 35224,\n",
       "             'decreases': 548,\n",
       "             'amount': 2226,\n",
       "             'inhibits': 2430,\n",
       "             'mscv': 381,\n",
       "             '60': 5689,\n",
       "             'probed': 845,\n",
       "             'actin': 2023,\n",
       "             'ensure': 567,\n",
       "             'equal': 1759,\n",
       "             'input': 708,\n",
       "             '81k': 8,\n",
       "             'conclusions': 1057,\n",
       "             'presented': 2700,\n",
       "             'several': 12523,\n",
       "             'clues': 179,\n",
       "             'underlying': 2035,\n",
       "             'first': 12309,\n",
       "             'resulted': 4553,\n",
       "             'rapid': 1614,\n",
       "             'promoted': 760,\n",
       "             'release': 918,\n",
       "             'adaptor': 875,\n",
       "             'post': 1970,\n",
       "             'translational': 1011,\n",
       "             'modifications': 1044,\n",
       "             'occur': 4952,\n",
       "             'stabilize': 1030,\n",
       "             'optimize': 124,\n",
       "             'linkage': 723,\n",
       "             'modify': 278,\n",
       "             'drive': 985,\n",
       "             'overall': 7841,\n",
       "             'deliver': 30,\n",
       "             'associate': 725,\n",
       "             'currently': 2446,\n",
       "             'exploring': 86,\n",
       "             'possibilities': 386,\n",
       "             'vast': 621,\n",
       "             'oncoproteins': 348,\n",
       "             'including': 16030,\n",
       "             'src': 3626,\n",
       "             'sv40': 342,\n",
       "             'antigen': 1443,\n",
       "             'ews': 1288,\n",
       "             'fli1': 615,\n",
       "             'pax3': 26,\n",
       "             'fkhr': 22,\n",
       "             'provides': 2370,\n",
       "             'plausible': 299,\n",
       "             'phenomenon': 445,\n",
       "             'inhibitors': 14382,\n",
       "             'disrupt': 2238,\n",
       "             'possess': 667,\n",
       "             'clinical': 23037,\n",
       "             'efficacy': 2937,\n",
       "             'tumors': 35533,\n",
       "             'suggest': 8972,\n",
       "             'subset': 3822,\n",
       "             'enhanced': 4050,\n",
       "             'sensitivity': 9747,\n",
       "             'important': 10020,\n",
       "             'extends': 232,\n",
       "             'particularly': 2297,\n",
       "             'better': 2356,\n",
       "             'understand': 1347,\n",
       "             'origin': 1613,\n",
       "             'cancer': 63259,\n",
       "             'developed': 4322,\n",
       "             'recapitulates': 89,\n",
       "             'product': 2598,\n",
       "             'translocation': 2455,\n",
       "             'characteristic': 1236,\n",
       "             'human': 24307,\n",
       "             'tissues': 4183,\n",
       "             'wap': 59,\n",
       "             'cre': 1911,\n",
       "             'fully': 2368,\n",
       "             'penetrant': 174,\n",
       "             'multifocal': 111,\n",
       "             'malignant': 3489,\n",
       "             'short': 2547,\n",
       "             'latency': 537,\n",
       "             'genetic': 12311,\n",
       "             'nulliparous': 17,\n",
       "             'females': 787,\n",
       "             'committed': 131,\n",
       "             'alveolar': 111,\n",
       "             'bipotent': 15,\n",
       "             'cd61': 37,\n",
       "             'luminal': 475,\n",
       "             'progenitors': 792,\n",
       "             'targets': 4609,\n",
       "             'tumorigenesis': 3230,\n",
       "             'transforms': 122,\n",
       "             'otherwise': 727,\n",
       "             'ap1': 211,\n",
       "             'increasing': 2392,\n",
       "             'relevance': 1814,\n",
       "             'chromosomal': 2680,\n",
       "             'cancers': 13647,\n",
       "             'serve': 1061,\n",
       "             'paradigm': 356,\n",
       "             'pathogenesis': 1753,\n",
       "             'origins': 234,\n",
       "             'generation': 2921,\n",
       "             'novel': 6521,\n",
       "             'preclinical': 1556,\n",
       "             'models': 6622,\n",
       "             'largest': 613,\n",
       "             'class': 6653,\n",
       "             'little': 2013,\n",
       "             'initiating': 504,\n",
       "             'hits': 324,\n",
       "             'tissue': 8240,\n",
       "             'stem': 3892,\n",
       "             'uncertain': 1357,\n",
       "             'system': 6866,\n",
       "             'initial': 2978,\n",
       "             'event': 2138,\n",
       "             'frank': 88,\n",
       "             'malignancy': 911,\n",
       "             'based': 15432,\n",
       "             'marking': 69,\n",
       "             'rather': 2854,\n",
       "             'focus': 1577,\n",
       "             'transcriptional': 9701,\n",
       "             'warranted': 281,\n",
       "             'go': 6035,\n",
       "             'malignancies': 2281,\n",
       "             'neither': 1238,\n",
       "             'heterogeneous': 1008,\n",
       "             'composition': 373,\n",
       "             'organized': 131,\n",
       "             'hierarchy': 41,\n",
       "             'includes': 1622,\n",
       "             'competent': 465,\n",
       "             'recreate': 21,\n",
       "             'transplantation': 608,\n",
       "             'ics': 9,\n",
       "             'bulk': 158,\n",
       "             'reya': 14,\n",
       "             'following': 7616,\n",
       "             'hit': 370,\n",
       "             'occurs': 2505,\n",
       "             'within': 14863,\n",
       "             'differentiated': 1169,\n",
       "             'secondary': 6984,\n",
       "             'events': 4053,\n",
       "             'epigenetic': 1293,\n",
       "             'contribute': 3655,\n",
       "             'evolution': 730,\n",
       "             'early': 4549,\n",
       "             'steps': 657,\n",
       "             'impossible': 91,\n",
       "             'patients': 57977,\n",
       "             'animal': 1154,\n",
       "             'window': 1284,\n",
       "             'phase': 5800,\n",
       "             'development': 9237,\n",
       "             'relevant': 3289,\n",
       "             'biology': 1317,\n",
       "             'far': 2392,\n",
       "             'mirror': 81,\n",
       "             'occurring': 1859,\n",
       "             'considering': 908,\n",
       "             'issues': 427,\n",
       "             'sought': 970,\n",
       "             'apply': 342,\n",
       "             'principles': 205,\n",
       "             'proved': 543,\n",
       "             'successful': 587,\n",
       "             'hematopoietic': 1722,\n",
       "             'leukemias': 638,\n",
       "             'childhood': 701,\n",
       "             'sarcomas': 442,\n",
       "             'leading': 3717,\n",
       "             'production': 1561,\n",
       "             'rowley': 9,\n",
       "             'rearrangements': 2943,\n",
       "             'rare': 4850,\n",
       "             'infrequent': 236,\n",
       "             'defined': 4997,\n",
       "             'factors': 6670,\n",
       "             'normal': 12958,\n",
       "             'differentiation': 3968,\n",
       "             'generally': 1897,\n",
       "             'perturbed': 155,\n",
       "             'lately': 7,\n",
       "             'contribution': 1461,\n",
       "             'viewed': 137,\n",
       "             'minor': 1065,\n",
       "             'involving': 3523,\n",
       "             'ets': 1615,\n",
       "             '50': 13843,\n",
       "             'cases': 19048,\n",
       "             'prostate': 4462,\n",
       "             'tomlins': 63,\n",
       "             '2007': 3618,\n",
       "             'eml4': 1651,\n",
       "             'alk': 17282,\n",
       "             '7': 31138,\n",
       "             'small': 11684,\n",
       "             'lung': 14222,\n",
       "             'soda': 37,\n",
       "             'observations': 3445,\n",
       "             'prompt': 49,\n",
       "             'reassessment': 47,\n",
       "             'regarding': 940,\n",
       "             'involvement': 1319,\n",
       "             'unique': 3354,\n",
       "             'recurrent': 4152,\n",
       "             'germ': 1549,\n",
       "             'layers': 128,\n",
       "             'sbc': 376,\n",
       "             '1998b': 80,\n",
       "             '1998a': 86,\n",
       "             'rubin': 35,\n",
       "             'myelogenous': 503,\n",
       "             'produces': 335,\n",
       "             'oncogene': 3333,\n",
       "             'encodes': 1704,\n",
       "             'oligomerization': 790,\n",
       "             'tel': 533,\n",
       "             'ptk': 168,\n",
       "             'trkc': 132,\n",
       "             'trk': 614,\n",
       "             'strong': 4729,\n",
       "             'epidemiological': 95,\n",
       "             'initiation': 1701,\n",
       "             'heterogeneity': 1620,\n",
       "             'genetically': 850,\n",
       "             'clinically': 2842,\n",
       "             'simpson': 13,\n",
       "             'part': 4221,\n",
       "             ...})"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Count_number_of_each_word_in_TEXT(X_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of unique words in train data : 53295\n"
     ]
    }
   ],
   "source": [
    "# one-hot encoding of TEXT feature.\n",
    "\n",
    "# building a CountVectorizer with all the words that occured minimum 3 times in train data.\n",
    "text_vectorizer = CountVectorizer(min_df=3)\n",
    "X_train_text_onehotCoding = text_vectorizer.fit_transform(X_train['TEXT'])\n",
    "X_test_text_onehotCoding = text_vectorizer.transform(X_test['TEXT'])\n",
    "X_cv_text_onehotCoding = text_vectorizer.transform(X_cv['TEXT'])\n",
    "\n",
    "# Normalization\n",
    "X_train_text_onehotCoding = normalize(X_train_text_onehotCoding, axis=0)\n",
    "X_test_text_onehotCoding = normalize(X_test_text_onehotCoding, axis=0)\n",
    "X_cv_text_onehotCoding = normalize(X_cv_text_onehotCoding, axis=0)\n",
    "\n",
    "# Check Number of Unique words in TEXT feature\n",
    "print(\"Total number of unique words in train data :\", X_train_text_onehotCoding.shape[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Response encoding of TEXT feature. (https://stackoverflow.com/a/1602964)\n",
    "\n",
    "def get_text_responsecoding(df, dict_list):\n",
    "    text_feature_responseCoding = np.zeros((df.shape[0],9))\n",
    "    for i in range(0,9):\n",
    "        row_index = 0\n",
    "        for index, row in df.iterrows():\n",
    "            sum_prob = 0\n",
    "            for word in row['TEXT'].split():\n",
    "                sum_prob += math.log(((dict_list[i].get(word,0)+10 )/(total_dict.get(word,0)+90)))\n",
    "            text_feature_responseCoding[row_index][i] = math.exp(sum_prob/len(row['TEXT'].split()))\n",
    "            row_index += 1\n",
    "    return text_feature_responseCoding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dict_list: contains 9 dictoinaries each contains number of occurrence of each word in a class\n",
    "dict_list = []\n",
    "for i in range(1,10):\n",
    "    cls_text = X_train[X_train['Class']==i]\n",
    "    dict_list.append(Count_number_of_each_word_in_TEXT(cls_text))\n",
    "\n",
    "# dict_list[i] is build on i'th  class text data\n",
    "# total number of occurrence of each word in training text data\n",
    "total_dict = Count_number_of_each_word_in_TEXT(X_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train_text_responseCoding  = get_text_responsecoding(X_train, dict_list)\n",
    "X_test_text_responseCoding  = get_text_responsecoding(X_test, dict_list)\n",
    "X_cv_text_responseCoding  = get_text_responsecoding(X_cv, dict_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id='62'>B. Evaluating Transformed Features </a> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After converting each categorical feature into a format that our ML model can understand, we can investigate the importance of each feature.\n",
    "\n",
    "In this Kaggle competition, log-loss evaluation metric is suggested to evaluate ML models. Therefore, we use this metric to investigate the imortance of each feature. (Refer to following URL for more details about log-loss:https://scikit-learn.org/stable/modules/generated/sklearn.metrics.log_loss.html)\n",
    "\n",
    "log-loss value range is \\[0, $\\infty$) and the smaller the logg-loss value the better the model performance. Therefore, to specify log-loss value, whichis acceptable for us, we consider a random model. The log-loss is calculated for this random model and is considered as the worse log-loss value. Then, our goal is going to be designing a model, which provids lower log-loss value than that of our random model. In other words, the column with log-loss smaller than random model is considered as an important feature.\n",
    "\n",
    "Note: Log-loss ia a complete concept on probability theory. Therfore, y_pred needs to be in terms of probability since we are using log-loss to evaluate our model. To handel this issue when y_pred is not probability, we use callibirated classifier."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Random Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This functioin gives the predicted label using random Moldel:\n",
    "\n",
    "# Initialize y_pred with the same number of rows as input and 9 columns.\n",
    "# for each row:\n",
    "# Construct an array of size 9 and sum of probability equal to one.\n",
    "def log_loss_random_model(X_df,y_df, dataset = \"dataset\"):\n",
    "    y_row_no = X_df.shape[0]\n",
    "    y_pred = np.zeros((y_row_no,9))\n",
    "    for i in range (y_row_no):\n",
    "        y_pred[i] = ((np.random.rand(1,9)/sum(sum(np.random.rand(1,9))))[0])\n",
    "        log_loss_metric = log_loss(y_df,y_pred, eps=1e-15)\n",
    "    print(\"log-loss on \" + dataset + \" set using Random Model: {}\".format(log_loss_metric))    \n",
    "    return log_loss_metric,y_pred"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "log-loss on cv set using Random Model: 2.362665779332182\n",
      "log-loss on test set using Random Model: 2.4641943016502075\n"
     ]
    }
   ],
   "source": [
    "# Fine log-loss value for test and cv sets using random model\n",
    "cv_log_loss_metric_random_model, y_pred_cv = log_loss_random_model(X_cv, y_cv, dataset = \"cv\")\n",
    "test_log_loss_metric_random_model, y_pred_test_p = log_loss_random_model(X_test, y_test, dataset = \"test\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# y_pred_test_p"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Let's get the index of max probablity\n",
    "# y_pred_test =np.argmax(y_pred_test, axis=1) + 1 # one is added here to have y_pred in range 1-9\n",
    "# y_pred_test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### One - Hot - encoding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def feature_evaluation_log_loss(clf, X_train, y_train, X_cv, y_cv, X_test, y_test):\n",
    "    clf.fit(X_train, y_train)\n",
    "    y_pred_cv_proba = clf.predict_proba(X_cv)\n",
    "    y_pred_test_proba = clf.predict_proba(X_test)\n",
    "    cv_log_loss = log_loss(y_cv,y_pred_cv_proba, eps=1e-15)\n",
    "    test_log_loss= log_loss(y_test,y_pred_test_proba, eps=1e-15)\n",
    "    return (cv_log_loss, test_log_loss)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_cv</th>\n",
       "      <th>Gene_test</th>\n",
       "      <th>Variation_cv</th>\n",
       "      <th>Variation_test</th>\n",
       "      <th>Text_cv</th>\n",
       "      <th>Text_test</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Response Encoding</th>\n",
       "      <td>1.37921</td>\n",
       "      <td>1.42842</td>\n",
       "      <td>1.74223</td>\n",
       "      <td>1.76845</td>\n",
       "      <td>1.44548</td>\n",
       "      <td>1.47657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>One-Hot-Encoding</th>\n",
       "      <td>1.20091</td>\n",
       "      <td>1.26632</td>\n",
       "      <td>1.70393</td>\n",
       "      <td>1.72812</td>\n",
       "      <td>1.47769</td>\n",
       "      <td>1.63797</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Gene_cv Gene_test Variation_cv Variation_test  Text_cv  \\\n",
       "Response Encoding  1.37921   1.42842      1.74223        1.76845  1.44548   \n",
       "One-Hot-Encoding   1.20091   1.26632      1.70393        1.72812  1.47769   \n",
       "\n",
       "                  Text_test  \n",
       "Response Encoding   1.47657  \n",
       "One-Hot-Encoding    1.63797  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clf = LogisticRegression()\n",
    "\n",
    "Metrics_df = pd.DataFrame(\n",
    "columns = ['Gene_cv','Gene_test','Variation_cv','Variation_test','Text_cv', 'Text_test'],\n",
    "index = ['Response Encoding', 'One-Hot-Encoding'])\n",
    "Metrics_df.loc['Response Encoding'][0:2] = feature_evaluation_log_loss(clf, X_train_gene_responseCoding, y_train, \n",
    "                                                X_cv_gene_responseCoding, y_cv, X_test_gene_responseCoding, y_test)\n",
    "Metrics_df.loc['One-Hot-Encoding'][0:2] = feature_evaluation_log_loss(clf, X_train_gene_onehotCoding, y_train, \n",
    "                                                X_cv_gene_onehotCoding, y_cv, X_test_gene_onehotCoding, y_test)\n",
    "Metrics_df.loc['Response Encoding'][2:4] = feature_evaluation_log_loss(clf, X_train_variation_responseCoding, y_train, \n",
    "                                                X_cv_variation_responseCoding, y_cv, X_test_variation_responseCoding, y_test)\n",
    "Metrics_df.loc['One-Hot-Encoding'][2:4] = feature_evaluation_log_loss(clf, X_train_variation_onehotCoding, y_train, \n",
    "                                                X_cv_variation_onehotCoding, y_cv, X_test_variation_onehotCoding, y_test)\n",
    "Metrics_df.loc['Response Encoding'][4:6] = feature_evaluation_log_loss(clf, X_train_text_responseCoding, y_train, \n",
    "                                                X_cv_text_responseCoding, y_cv, X_test_text_responseCoding, y_test)\n",
    "Metrics_df.loc['One-Hot-Encoding'][4:6] = feature_evaluation_log_loss(clf, X_train_text_onehotCoding, y_train, \n",
    "                                                X_cv_text_onehotCoding, y_cv, X_test_text_onehotCoding, y_test)\n",
    "Metrics_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All features show smaller log_loss than Random model even without tuning the classifier."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id='63'>C. Merging Important Features </a> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merging converted gene, variance and text features\n",
    "\n",
    "# building train, test and cross validation data sets\n",
    "# a = [[1, 2], \n",
    "#      [3, 4]]\n",
    "# b = [[4, 5], \n",
    "#      [6, 7]]\n",
    "# hstack(a, b) = [[1, 2, 4, 5],\n",
    "#                [ 3, 4, 6, 7]]\n",
    "\n",
    "y_train = np.array(list(X_train['Class']))\n",
    "y_cv = np.array(list(X_cv['Class']))\n",
    "y_test = np.array(list(X_test['Class']))\n",
    "\n",
    "# One-Hot_Encoding\n",
    "train_gene_var_onehotCoding = hstack((X_train_gene_onehotCoding,X_train_variation_onehotCoding))\n",
    "test_gene_var_onehotCoding = hstack((X_test_gene_onehotCoding,X_test_variation_onehotCoding))\n",
    "cv_gene_var_onehotCoding = hstack((X_cv_gene_onehotCoding,X_cv_variation_onehotCoding))\n",
    "\n",
    "X_train_onehotCoding = hstack((train_gene_var_onehotCoding, X_train_text_onehotCoding)).tocsr()\n",
    "X_test_onehotCoding = hstack((test_gene_var_onehotCoding, X_test_text_onehotCoding)).tocsr()\n",
    "X_cv_onehotCoding = hstack((cv_gene_var_onehotCoding, X_cv_text_onehotCoding)).tocsr()\n",
    "\n",
    "\n",
    "# Response Encoding\n",
    "train_gene_var_responseCoding = np.hstack((X_train_gene_responseCoding, X_train_variation_responseCoding))\n",
    "test_gene_var_responseCoding = np.hstack((X_test_gene_responseCoding, X_test_variation_responseCoding))\n",
    "cv_gene_var_responseCoding = np.hstack((X_cv_gene_responseCoding, X_cv_variation_responseCoding))\n",
    "\n",
    "X_train_responseCoding = np.hstack((train_gene_var_responseCoding, X_train_text_responseCoding))\n",
    "X_test_responseCoding = np.hstack((test_gene_var_responseCoding, X_test_text_responseCoding))\n",
    "X_cv_responseCoding = np.hstack((cv_gene_var_responseCoding, X_cv_text_responseCoding))\n",
    "\n",
    "\n",
    "# # gene & variation (One-Hot_Encoding) - text (Response Encoding) \n",
    "# train_gene_var_onehotCoding = hstack((X_train_gene_onehotCoding,X_train_variation_onehotCoding))\n",
    "# test_gene_var_onehotCoding = hstack((X_test_gene_onehotCoding,X_test_variation_onehotCoding))\n",
    "# cv_gene_var_onehotCoding = hstack((X_cv_gene_onehotCoding,X_cv_variation_onehotCoding))\n",
    "\n",
    "# X_train_hot_res = hstack((train_gene_var_onehotCoding, X_train_text_responseCoding))\n",
    "# X_test_hot_res = hstack((test_gene_var_onehotCoding, X_test_text_responseCoding))\n",
    "# X_cv_hot_res = hstack((cv_gene_var_onehotCoding, X_cv_text_responseCoding))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "One hot encoding features :\n",
      "(number of data points, number of features) in train data =  (2124, 55483)\n",
      "(number of data points, number of features) in test data =  (665, 55483)\n",
      "(number of data points, number of features) in cross validation data = (532, 55483)\n"
     ]
    }
   ],
   "source": [
    "print(\"One hot encoding features :\")\n",
    "print(\"(number of data points, number of features) in train data = \", X_train_onehotCoding.shape)\n",
    "print(\"(number of data points, number of features) in test data = \", X_test_onehotCoding.shape)\n",
    "print(\"(number of data points, number of features) in cross validation data =\", X_cv_onehotCoding.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Response encoding features :\n",
      "(number of data points, number of features) in train data =  (2124, 27)\n",
      "(number of data points, number of features) in test data =  (665, 27)\n",
      "(number of data points, number of features) in cross validation data = (532, 27)\n"
     ]
    }
   ],
   "source": [
    "print(\" Response encoding features :\")\n",
    "print(\"(number of data points, number of features) in train data = \", X_train_responseCoding.shape)\n",
    "print(\"(number of data points, number of features) in test data = \", X_test_responseCoding.shape)\n",
    "print(\"(number of data points, number of features) in cross validation data =\", X_cv_responseCoding.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(\" Hot & Response features :\")\n",
    "# print(\"(number of data points, number of features) in train data = \", X_train_hot_res.shape)\n",
    "# print(\"(number of data points, number of features) in test data = \", X_test_hot_res.shape)\n",
    "# print(\"(number of data points, number of features) in cross validation data =\", X_cv_hot_res.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id=7 > 7. Modeling Workflow"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this Section, different classifiers such as LR, NB, KNN, SVM, and RF are used to compare the result of different encoding tech. Then, we conclude which encoding technique and classifire leads to better result. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id='71'>A. Model 1: Compare Different Classifiers and Different Encoding Technique</a> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Model_evaluation(classifier, X_train, y_train, X_cv, y_cv, X_test, y_test):\n",
    "    metrics_dict = {}\n",
    "    \n",
    "    classifier.fit(X_train, y_train)    \n",
    "    sig_classifier = CalibratedClassifierCV(classifier, method=\"sigmoid\")\n",
    "    sig_classifier.fit(X_train, y_train)\n",
    "    \n",
    "    y_pred_train_proba = sig_classifier.predict_proba(X_train)\n",
    "    y_pred_cv_proba = sig_classifier.predict_proba(X_cv)\n",
    "    y_pred_test_proba = sig_classifier.predict_proba(X_test)\n",
    "\n",
    "#     y_pred_cv_proba = classifier.predict_proba(X_cv)\n",
    "#     y_pred_test_proba = classifier.predict_proba(X_test)\n",
    "    \n",
    "    # log-loss\n",
    "    metrics_dict['train_log_loss'] = log_loss(y_train,y_pred_train_proba, eps=1e-15)\n",
    "    metrics_dict['cv_log_loss'] = log_loss(y_cv,y_pred_cv_proba, eps=1e-15)\n",
    "    metrics_dict['test_log_loss'] = log_loss(y_test,y_pred_test_proba, eps=1e-15)\n",
    "    \n",
    "    # Accuracy\n",
    "    metrics_dict['train_Acc (%)'] = round(sig_classifier.score(X_train, y_train),2) * 100 \n",
    "    metrics_dict['cv_Acc (%)'] = round(sig_classifier.score(X_cv, y_cv),2) * 100 \n",
    "    metrics_dict['test_Acc (%)'] = round(sig_classifier.score(X_test, y_test),2) * 100\n",
    "\n",
    "    return metrics_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Logistic Regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compare Response Encoding and One-Hot-Encoding using LR:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>train_log_loss</th>\n",
       "      <th>cv_log_loss</th>\n",
       "      <th>test_log_loss</th>\n",
       "      <th>cv_Acc (%)</th>\n",
       "      <th>test_Acc (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Response Encoding</th>\n",
       "      <td>1.07905</td>\n",
       "      <td>1.17073</td>\n",
       "      <td>1.25189</td>\n",
       "      <td>57</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>One-Hot-Encoding</th>\n",
       "      <td>0.531786</td>\n",
       "      <td>1.01216</td>\n",
       "      <td>1.13506</td>\n",
       "      <td>66</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  train_log_loss cv_log_loss test_log_loss cv_Acc (%)  \\\n",
       "Response Encoding        1.07905     1.17073       1.25189         57   \n",
       "One-Hot-Encoding        0.531786     1.01216       1.13506         66   \n",
       "\n",
       "                  test_Acc (%)  \n",
       "Response Encoding           57  \n",
       "One-Hot-Encoding            62  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Initialize a classifier object with default params\n",
    "LR_clf = LogisticRegression()\n",
    "# Use Model_evaluation function to evaluate the model\n",
    "LR_responseCoding_result = Model_evaluation(LR_clf, X_train_responseCoding, y_train, X_cv_responseCoding, y_cv, X_test_responseCoding, y_test)\n",
    "LR_onehotCoding_result = Model_evaluation(LR_clf, X_train_onehotCoding, y_train, X_cv_onehotCoding, y_cv, X_test_onehotCoding, y_test)\n",
    "\n",
    "LR_metrics_df = pd.DataFrame(\n",
    "columns = ['train_log_loss','cv_log_loss','test_log_loss','cv_Acc (%)','test_Acc (%)'],\n",
    "index = ['Response Encoding', 'One-Hot-Encoding'])\n",
    "LR_metrics_df.loc['Response Encoding'] = LR_responseCoding_result\n",
    "LR_metrics_df.loc['One-Hot-Encoding'] = LR_onehotCoding_result\n",
    "print(\"Compare Response Encoding and One-Hot-Encoding using LR:\")\n",
    "LR_metrics_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Naive Bayes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compare Response Encoding and One-Hot-Encoding using NB:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>train_log_loss</th>\n",
       "      <th>cv_log_loss</th>\n",
       "      <th>test_log_loss</th>\n",
       "      <th>cv_Acc (%)</th>\n",
       "      <th>test_Acc (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Response Encoding</th>\n",
       "      <td>0.666964</td>\n",
       "      <td>1.93124</td>\n",
       "      <td>2.00406</td>\n",
       "      <td>46</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>One-Hot-Encoding</th>\n",
       "      <td>0.761873</td>\n",
       "      <td>1.40219</td>\n",
       "      <td>1.45601</td>\n",
       "      <td>55</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  train_log_loss cv_log_loss test_log_loss cv_Acc (%)  \\\n",
       "Response Encoding       0.666964     1.93124       2.00406         46   \n",
       "One-Hot-Encoding        0.761873     1.40219       1.45601         55   \n",
       "\n",
       "                  test_Acc (%)  \n",
       "Response Encoding           45  \n",
       "One-Hot-Encoding            52  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Initialize a classifier object with default params\n",
    "NB_clf = GaussianNB()\n",
    "# Use Model_evaluation function to evaluate the model\n",
    "NB_responseCoding_result = Model_evaluation(NB_clf, X_train_responseCoding, y_train, X_cv_responseCoding, y_cv,\n",
    "                                            X_test_responseCoding, y_test)\n",
    "NB_onehotCoding_result = Model_evaluation(NB_clf, X_train_onehotCoding.toarray(), y_train, X_cv_onehotCoding.toarray(), y_cv, \n",
    "                                          X_test_onehotCoding.toarray(), y_test)\n",
    "\n",
    "NB_metrics_df = pd.DataFrame(\n",
    "columns = ['train_log_loss','cv_log_loss','test_log_loss','cv_Acc (%)','test_Acc (%)'],\n",
    "index = ['Response Encoding', 'One-Hot-Encoding'])\n",
    "NB_metrics_df.loc['Response Encoding'] = NB_responseCoding_result\n",
    "NB_metrics_df.loc['One-Hot-Encoding'] = NB_onehotCoding_result\n",
    "\n",
    "print(\"Compare Response Encoding and One-Hot-Encoding using NB:\")\n",
    "NB_metrics_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### K-Nearest Neighbour"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compare Response Encoding and One-Hot-Encoding using KNN:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>train_log_loss</th>\n",
       "      <th>cv_log_loss</th>\n",
       "      <th>test_log_loss</th>\n",
       "      <th>cv_Acc (%)</th>\n",
       "      <th>test_Acc (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Response Encoding</th>\n",
       "      <td>0.433242</td>\n",
       "      <td>1.05383</td>\n",
       "      <td>1.21585</td>\n",
       "      <td>66</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>One-Hot-Encoding</th>\n",
       "      <td>0.930286</td>\n",
       "      <td>1.21185</td>\n",
       "      <td>1.33733</td>\n",
       "      <td>59</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  train_log_loss cv_log_loss test_log_loss cv_Acc (%)  \\\n",
       "Response Encoding       0.433242     1.05383       1.21585         66   \n",
       "One-Hot-Encoding        0.930286     1.21185       1.33733         59   \n",
       "\n",
       "                  test_Acc (%)  \n",
       "Response Encoding           59  \n",
       "One-Hot-Encoding            53  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Initialize a classifier object with default params\n",
    "KNN_clf = KNeighborsClassifier()\n",
    "# Use Model_evaluation function to evaluate the model\n",
    "KNN_responseCoding_result = Model_evaluation(KNN_clf, X_train_responseCoding, y_train, X_cv_responseCoding, y_cv,\n",
    "                                            X_test_responseCoding, y_test)\n",
    "KNN_onehotCoding_result = Model_evaluation(KNN_clf, X_train_onehotCoding.toarray(), y_train, X_cv_onehotCoding.toarray(), y_cv, \n",
    "                                          X_test_onehotCoding.toarray(), y_test)\n",
    "\n",
    "KNN_metrics_df = pd.DataFrame(\n",
    "columns = ['train_log_loss','cv_log_loss','test_log_loss','cv_Acc (%)','test_Acc (%)'],\n",
    "index = ['Response Encoding', 'One-Hot-Encoding'])\n",
    "KNN_metrics_df.loc['Response Encoding'] = KNN_responseCoding_result\n",
    "KNN_metrics_df.loc['One-Hot-Encoding'] = KNN_onehotCoding_result\n",
    "\n",
    "print(\"Compare Response Encoding and One-Hot-Encoding using KNN:\")\n",
    "KNN_metrics_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Support Vector Machine (SVM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compare Response Encoding and One-Hot-Encoding using SVM:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>train_log_loss</th>\n",
       "      <th>cv_log_loss</th>\n",
       "      <th>test_log_loss</th>\n",
       "      <th>cv_Acc (%)</th>\n",
       "      <th>test_Acc (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Response Encoding</th>\n",
       "      <td>0.886483</td>\n",
       "      <td>1.18646</td>\n",
       "      <td>1.28481</td>\n",
       "      <td>62</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>One-Hot-Encoding</th>\n",
       "      <td>0.725338</td>\n",
       "      <td>1.45769</td>\n",
       "      <td>1.47813</td>\n",
       "      <td>53</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  train_log_loss cv_log_loss test_log_loss cv_Acc (%)  \\\n",
       "Response Encoding       0.886483     1.18646       1.28481         62   \n",
       "One-Hot-Encoding        0.725338     1.45769       1.47813         53   \n",
       "\n",
       "                  test_Acc (%)  \n",
       "Response Encoding           58  \n",
       "One-Hot-Encoding            53  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Initialize a classifier object with default params\n",
    "SVM_clf = SVC()\n",
    "# Use Model_evaluation function to evaluate the model\n",
    "SVM_responseCoding_result = Model_evaluation(SVM_clf, X_train_responseCoding, y_train, X_cv_responseCoding, y_cv,\n",
    "                                            X_test_responseCoding, y_test)\n",
    "SVM_onehotCoding_result = Model_evaluation(SVM_clf, X_train_onehotCoding.toarray(), y_train, X_cv_onehotCoding.toarray(), y_cv, \n",
    "                                          X_test_onehotCoding.toarray(), y_test)\n",
    "\n",
    "SVM_metrics_df = pd.DataFrame(\n",
    "columns = ['train_log_loss','cv_log_loss','test_log_loss','cv_Acc (%)','test_Acc (%)'],\n",
    "index = ['Response Encoding', 'One-Hot-Encoding'])\n",
    "SVM_metrics_df.loc['Response Encoding'] = SVM_responseCoding_result\n",
    "SVM_metrics_df.loc['One-Hot-Encoding'] = SVM_onehotCoding_result\n",
    "\n",
    "print(\"Compare Response Encoding and One-Hot-Encoding using SVM:\")\n",
    "SVM_metrics_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Random Forest (RF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compare Response Encoding and One-Hot-Encoding using RF:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>train_log_loss</th>\n",
       "      <th>cv_log_loss</th>\n",
       "      <th>test_log_loss</th>\n",
       "      <th>cv_Acc (%)</th>\n",
       "      <th>test_Acc (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Response Encoding</th>\n",
       "      <td>0.0364591</td>\n",
       "      <td>1.7945</td>\n",
       "      <td>1.83311</td>\n",
       "      <td>29</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>One-Hot-Encoding</th>\n",
       "      <td>0.374148</td>\n",
       "      <td>1.17851</td>\n",
       "      <td>1.2541</td>\n",
       "      <td>59</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  train_log_loss cv_log_loss test_log_loss cv_Acc (%)  \\\n",
       "Response Encoding      0.0364591      1.7945       1.83311         29   \n",
       "One-Hot-Encoding        0.374148     1.17851        1.2541         59   \n",
       "\n",
       "                  test_Acc (%)  \n",
       "Response Encoding           29  \n",
       "One-Hot-Encoding            58  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Initialize a classifier object with default params\n",
    "RF_clf = RandomForestClassifier()\n",
    "# Use Model_evaluation function to evaluate the model\n",
    "RF_responseCoding_result = Model_evaluation(RF_clf, X_train_responseCoding, y_train, X_cv_responseCoding, y_cv,\n",
    "                                            X_test_responseCoding, y_test)\n",
    "RF_onehotCoding_result = Model_evaluation(RF_clf, X_train_onehotCoding.toarray(), y_train, X_cv_onehotCoding.toarray(), y_cv, \n",
    "                                          X_test_onehotCoding.toarray(), y_test)\n",
    "\n",
    "RF_metrics_df = pd.DataFrame(\n",
    "columns = ['train_log_loss','cv_log_loss','test_log_loss','cv_Acc (%)','test_Acc (%)'],\n",
    "index = ['Response Encoding', 'One-Hot-Encoding'])\n",
    "RF_metrics_df.loc['Response Encoding'] = RF_responseCoding_result\n",
    "RF_metrics_df.loc['One-Hot-Encoding'] = RF_onehotCoding_result\n",
    "\n",
    "print(\"Compare Response Encoding and One-Hot-Encoding using RF:\")\n",
    "RF_metrics_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Based on the results, LR and RF classifiers using One-Hot-Encoding technique provide the first  and the second best results, respectively. For the rest of our experiments, we use LR, RF, and One-Hot-Encoding technique."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id='72'>B. Model 2: Hyperparameter Tuning</a> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id='721'>a. Logistic Regression using Hyperparameter Tuning</a> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log Loss for alpha = 1e-06 is:  1.3116609680690392\n",
      "Log Loss for alpha = 1e-05 is:  1.2766644613377498\n",
      "Log Loss for alpha = 0.0001 is:  1.1134633800277676\n",
      "Log Loss for alpha = 0.001 is:  1.0800139337017636\n",
      "Log Loss for alpha = 0.01 is:  1.1440416225610062\n",
      "Log Loss for alpha = 0.1 is:  1.3540064124351994\n",
      "Log Loss for alpha = 1 is:  1.583683573338141\n",
      "Log Loss for alpha = 10 is:  1.6210306919877266\n",
      "Log Loss for alpha = 100 is:  1.625306916265727\n"
     ]
    }
   ],
   "source": [
    "# Initialize a classifier object with default params\n",
    "# LR_onehotCoding_result = Model_evaluation(LR_clf, X_train_onehotCoding, y_train, X_cv_onehotCoding, y_cv, X_test_onehotCoding, y_test)\n",
    "\n",
    "alpha = [10 ** x for x in range(-6, 3)]\n",
    "cv_log_loss_array = []\n",
    "for i in alpha:\n",
    "    classifier = SGDClassifier(alpha=i, penalty='l2', loss='log', random_state=42)\n",
    "    classifier.fit(X_train_onehotCoding, y_train)    \n",
    "    sig_classifier = CalibratedClassifierCV(classifier, method=\"sigmoid\")\n",
    "    sig_classifier.fit(X_train_onehotCoding, y_train)\n",
    "    \n",
    "    y_pred_cv_proba = sig_classifier.predict_proba(X_cv_onehotCoding)\n",
    "    \n",
    "    cv_log_loss_array.append(log_loss(y_cv,y_pred_cv_proba))\n",
    "    print(\"Log Loss for alpha =\", i, \"is: \",(log_loss(y_cv,y_pred_cv_proba, eps=1e-15)))\n",
    "   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For LR, the best value of alpha = 0.001 , provides log-loss value of 1.0800139337017636\n"
     ]
    }
   ],
   "source": [
    "# The alpha that give the minimum value of log-loss\n",
    "best_alpha = np.argmin(cv_log_loss_array)\n",
    "\n",
    "print('For LR, the best value of alpha =', alpha[best_alpha],', provides log-loss value of', cv_log_loss_array[best_alpha])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "By using Tuned LR classifier we have:\n",
      "The train log-loss is: 0.505912541080467\n",
      "The cross validation log-loss is: 1.0800139337017636\n",
      "The test log-loss is: 1.2494786119544001\n"
     ]
    }
   ],
   "source": [
    "# Use the best value of alpha to train the classifier\n",
    "classifier = SGDClassifier(alpha=alpha[best_alpha], penalty='l2', loss='log', random_state=42)\n",
    "classifier.fit(X_train_onehotCoding, y_train)    \n",
    "sig_classifier = CalibratedClassifierCV(classifier, method=\"sigmoid\")\n",
    "sig_classifier.fit(X_train_onehotCoding, y_train)\n",
    "\n",
    "log_loss_LR_train = log_loss(y_train, sig_classifier.predict_proba(X_train_onehotCoding))\n",
    "log_loss_LR_cv = log_loss(y_cv, sig_classifier.predict_proba(X_cv_onehotCoding))                                           \n",
    "log_loss_LR_test = log_loss(y_test, sig_classifier.predict_proba(X_test_onehotCoding))\n",
    "\n",
    "print(\"By using Tuned LR classifier we have:\")\n",
    "print(\"The train log-loss is:\",log_loss_LR_train)\n",
    "print(\"The cross validation log-loss is:\",log_loss_LR_cv)\n",
    "print(\"The test log-loss is:\",log_loss_LR_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id='722'>b. Random Forest using Hyperparameter Tuning</a> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log Loss for alpha = 100 and max depth = 5 is:  1.2157149722499943\n",
      "Log Loss for alpha = 100 and max depth = 10 is:  1.1443183162870751\n",
      "Log Loss for alpha = 200 and max depth = 5 is:  1.2063788319833724\n",
      "Log Loss for alpha = 200 and max depth = 10 is:  1.1336617283524495\n",
      "Log Loss for alpha = 500 and max depth = 5 is:  1.2014400753611916\n",
      "Log Loss for alpha = 500 and max depth = 10 is:  1.1335467872348575\n",
      "Log Loss for alpha = 1000 and max depth = 5 is:  1.201793822531273\n",
      "Log Loss for alpha = 1000 and max depth = 10 is:  1.13665359767696\n",
      "Log Loss for alpha = 2000 and max depth = 5 is:  1.203703234772255\n",
      "Log Loss for alpha = 2000 and max depth = 10 is:  1.1373173405500907\n"
     ]
    }
   ],
   "source": [
    "# Initialize a classifier object with default params\n",
    "alpha = [100,200,500,1000,2000]\n",
    "max_depth = [5, 10]\n",
    "cv_log_loss_array = []\n",
    "for i in alpha:\n",
    "    for j in max_depth:\n",
    "        classifier = RandomForestClassifier(n_estimators=i, criterion='gini', max_depth=j, random_state=42, n_jobs=-1)\n",
    "        classifier.fit(X_train_onehotCoding, y_train)    \n",
    "        sig_classifier = CalibratedClassifierCV(classifier, method=\"sigmoid\")\n",
    "        sig_classifier.fit(X_train_onehotCoding, y_train)\n",
    "\n",
    "        y_pred_cv_proba = sig_classifier.predict_proba(X_cv_onehotCoding)\n",
    "\n",
    "        cv_log_loss_array.append(log_loss(y_cv,y_pred_cv_proba))\n",
    "        print(\"Log Loss for alpha =\", i, \"and max depth =\", j, \"is: \",(log_loss(y_cv,y_pred_cv_proba, eps=1e-15)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For RF, the best value of alpha =  500 and max depth = 10 , provides log-loss value of 1.1335467872348575\n"
     ]
    }
   ],
   "source": [
    "# The alpha that give the minimum value of log-loss\n",
    "best_alpha = np.argmin(cv_log_loss_array)\n",
    "\n",
    "print('For RF, the best value of alpha = ', alpha[int(best_alpha/2)], 'and max depth =', \n",
    "      max_depth[int(best_alpha%2)],', provides log-loss value of', cv_log_loss_array[best_alpha])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "By using Tuned RF classifier we have:\n",
      "The train log-loss is: 0.6757988808182556\n",
      "The cross validation log-loss is: 1.1335467872348575\n",
      "The test log-loss is: 1.2251900663727597\n"
     ]
    }
   ],
   "source": [
    "classifier = RandomForestClassifier(n_estimators=alpha[int(best_alpha/2)], criterion='gini', \n",
    "                                    max_depth=max_depth[int(best_alpha%2)], random_state=42, n_jobs=-1)\n",
    "classifier.fit(X_train_onehotCoding, y_train)    \n",
    "sig_classifier = CalibratedClassifierCV(classifier, method=\"sigmoid\")\n",
    "sig_classifier.fit(X_train_onehotCoding, y_train)\n",
    "\n",
    "log_loss_RF_train = log_loss(y_train, sig_classifier.predict_proba(X_train_onehotCoding))\n",
    "log_loss_RF_cv = log_loss(y_cv, sig_classifier.predict_proba(X_cv_onehotCoding))                                           \n",
    "log_loss_RF_test = log_loss(y_test, sig_classifier.predict_proba(X_test_onehotCoding))\n",
    "\n",
    "print(\"By using Tuned RF classifier we have:\")\n",
    "print(\"The train log-loss is:\",log_loss_RF_train)\n",
    "print(\"The cross validation log-loss is:\",log_loss_RF_cv)\n",
    "print(\"The test log-loss is:\",log_loss_RF_test)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id='73'>C. Model 3: Stacking Model</a> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Logistic Regression:  Log Loss: 1.08 =\n",
      "Random Forest: Log Loss: 1.13 = \n",
      "----------------------------------------------------------------------------------------------------\n",
      "Stacking Classifer: for the value of alpha = 0.000100 Log Loss: 1.816 is\n",
      "Stacking Classifer: for the value of alpha = 0.001000 Log Loss: 1.707 is\n",
      "Stacking Classifer: for the value of alpha = 0.010000 Log Loss: 1.308 is\n",
      "Stacking Classifer: for the value of alpha = 0.100000 Log Loss: 1.122 is\n",
      "Stacking Classifer: for the value of alpha = 1.000000 Log Loss: 1.356 is\n",
      "Stacking Classifer: for the value of alpha = 10.000000 Log Loss: 1.734 is\n"
     ]
    }
   ],
   "source": [
    "# Logistic Regression\n",
    "classifierLR = SGDClassifier(alpha=0.001, penalty='l2', loss='log', random_state=42)\n",
    "classifierLR.fit(X_train_onehotCoding, y_train) \n",
    "sig_classifierLR = CalibratedClassifierCV(classifierLR, method=\"sigmoid\")\n",
    "sig_classifierLR.fit(X_train_onehotCoding, y_train)\n",
    "y_LRpred_cv_proba = sig_classifierLR.predict_proba(X_cv_onehotCoding)\n",
    "\n",
    "# Random Forest\n",
    "classifierRF = RandomForestClassifier(n_estimators=500, criterion='gini', max_depth=10, random_state=42, n_jobs=-1)\n",
    "classifierRF.fit(X_train_onehotCoding, y_train)    \n",
    "sig_classifierRF = CalibratedClassifierCV(classifierRF, method=\"sigmoid\")\n",
    "sig_classifierRF.fit(X_train_onehotCoding, y_train)\n",
    "y_RFpred_cv_proba = sig_classifierRF.predict_proba(X_cv_onehotCoding)\n",
    "\n",
    "print(\"Logistic Regression:  Log Loss: %0.2f =\" % (log_loss(y_cv, y_LRpred_cv_proba)))\n",
    "print(\"Random Forest: Log Loss: %0.2f = \" % (log_loss(y_cv, y_RFpred_cv_proba)))\n",
    "print(\"-\"*100)\n",
    "alpha = [0.0001,0.001,0.01,0.1,1,10] \n",
    "cv_log_loss_array = []\n",
    "best_alpha = 999\n",
    "for i in alpha:\n",
    "    lr = LogisticRegression(C=i)\n",
    "    StackingClassifer = StackingClassifier(classifiers=[sig_classifierLR, sig_classifierRF], \n",
    "                                           meta_classifier=lr, use_probas=True)\n",
    "    StackingClassifer.fit(X_train_onehotCoding, y_train)\n",
    "    log_loss_StackingClassifer = log_loss(y_cv, StackingClassifer.predict_proba(X_cv_onehotCoding))\n",
    "    print(\"Stacking Classifer: for the value of alpha = %f Log Loss: %0.3f is\" % (i, log_loss_StackingClassifer))\n",
    "    cv_log_loss_array.append(log_loss_StackingClassifer)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For Stacking Classifer, the best value of alpha is 0.1\n"
     ]
    }
   ],
   "source": [
    "# The alpha that give the minimum value of log-loss\n",
    "best_alpha = np.argmin(cv_log_loss_array)\n",
    "\n",
    "print('For Stacking Classifer, the best value of alpha is', alpha[best_alpha])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "By using Stacking Classifer we have:\n",
      "The train log-loss is: 0.9129089410128405\n",
      "The cross validation log-loss is: 1.279622378823506\n",
      "The test log-loss is: 1.3572488220174919\n"
     ]
    }
   ],
   "source": [
    "lr = LogisticRegression(C = alpha[best_alpha])\n",
    "Classifer = StackingClassifier(classifiers=[sig_classifierLR, sig_classifierRF], meta_classifier=lr, use_probas=True)\n",
    "classifier.fit(X_train_onehotCoding, y_train)    \n",
    "\n",
    "log_loss_StackingClassifer_train = log_loss(y_train, classifier.predict_proba(X_train_onehotCoding))\n",
    "log_loss_StackingClassifer_cv = log_loss(y_cv, classifier.predict_proba(X_cv_onehotCoding))                                           \n",
    "log_loss_StackingClassifer_test = log_loss(y_test, classifier.predict_proba(X_test_onehotCoding))\n",
    "\n",
    "print(\"By using Stacking Classifer we have:\")\n",
    "print(\"The train log-loss is:\",log_loss_StackingClassifer_train)#log_loss(y_train, y_pred_train_proba, labels=classifier.classes_, eps=1e-15))\n",
    "print(\"The cross validation log-loss is:\",log_loss_StackingClassifer_cv)#log_loss(y_cv, y_pred_cv_proba, labels=classifier.classes_, eps=1e-15))\n",
    "print(\"The test log-loss is:\",log_loss_StackingClassifer_test)#log_loss(y_test, y_pred_test_proba, labels=classifier.classes_, eps=1e-15))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <a id='74'>D. Model 4: Voting Classifier</a> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "By using Voting Classifer we have:\n",
      "The train log-loss is: 0.5706206852330493\n",
      "The cross validation log-loss is: 1.0046195701950729\n",
      "The test log-loss is: 1.1301653120307023\n"
     ]
    }
   ],
   "source": [
    "vclassifier = VotingClassifier(estimators=[('lr', sig_classifierLR), ('RF', sig_classifierRF)], voting='soft')\n",
    "vclassifier.fit(X_train_onehotCoding, y_train) \n",
    "                      \n",
    "log_loss_vclassifier_train = log_loss(y_train, vclassifier.predict_proba(X_train_onehotCoding))\n",
    "log_loss_vclassifier_cv = log_loss(y_cv, vclassifier.predict_proba(X_cv_onehotCoding))                                           \n",
    "log_loss_vclassifier_test = log_loss(y_test, vclassifier.predict_proba(X_test_onehotCoding))\n",
    "\n",
    "print(\"By using Voting Classifer we have:\")\n",
    "print(\"The train log-loss is:\", log_loss_vclassifier_train)\n",
    "print(\"The cross validation log-loss is:\", log_loss_vclassifier_cv)\n",
    "print(\"The test log-loss is:\", log_loss_vclassifier_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <a id = 8 > 8. Conclusion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model 1</th>\n",
       "      <th>Model 2 (LR)</th>\n",
       "      <th>Model 2 (RF)</th>\n",
       "      <th>Model 3</th>\n",
       "      <th>Model 4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>log-loss (Test set)</th>\n",
       "      <td>1.135065</td>\n",
       "      <td>1.249479</td>\n",
       "      <td>1.22519</td>\n",
       "      <td>1.357249</td>\n",
       "      <td>1.130165</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Model 1  Model 2 (LR)  Model 2 (RF)   Model 3   Model 4\n",
       "log-loss (Test set)  1.135065      1.249479       1.22519  1.357249  1.130165"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metrics_df_conc = pd.DataFrame(\n",
    "columns = ['Model 1','Model 2 (LR)','Model 2 (RF)', 'Model 3','Model 4'],\n",
    "index = ['log-loss (Test set)'])\n",
    "metrics_df_conc['Model 1'] = LR_onehotCoding_result['test_log_loss']\n",
    "metrics_df_conc['Model 2 (LR)'] = log_loss_LR_test\n",
    "metrics_df_conc['Model 2 (RF)'] = log_loss_RF_test\n",
    "metrics_df_conc['Model 3'] = log_loss_StackingClassifer_test\n",
    "metrics_df_conc['Model 4'] = log_loss_vclassifier_test\n",
    "metrics_df_conc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### In summary, the results show that the Model 4  provides the best log-loss compared to other models. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "# confusion_matrix_test = confusion_matrix(y_test, y_pred_test)\n",
    "# ModelSpe = 'Random Model'\n",
    "\n",
    "# g = plt.figure(figsize=(15,5))\n",
    "# labels = [1,2,3,4,5,6,7,8,9]\n",
    "# ax = sns.heatmap(confusion_matrix_test,cmap=\"YlGnBu\", annot=True, annot_kws={\"size\": 16},xticklabels=labels, yticklabels=labels) # font size\n",
    "# plt.title('Confusion Matrix - '+ ModelSpe, fontsize = 20) # title with fontsize 20\n",
    "# plt.xlabel('Predicted Label', fontsize = 15) # x-axis label with fontsize 15\n",
    "# plt.ylabel('True Label', fontsize = 15) # y-axis label with fontsize 15\n",
    "# ax.set_xticklabels(ax.get_xticklabels(), rotation=45, horizontalalignment='right')\n",
    "# ax.set_yticklabels(ax.get_yticklabels(), rotation=45, horizontalalignment='right')\n",
    "# g.show()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
